// @nestjs/graphql/dist/tests/tests/schema-builder/type-definitions-generator.register-in.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const testing_1 = require("@nestjs/testing");
const lib_1 = require("../../lib");
const type_definitions_generator_1 = require("../../lib/schema-builder/type-definitions.generator");
const type_definitions_storage_1 = require("../../lib/schema-builder/storages/type-definitions.storage");
const lazy_metadata_storage_1 = require("../../lib/schema-builder/storages/lazy-metadata.storage");
// Mock modules for testing
class UsersModule {
}
class ProductsModule {
}
// Types for UsersModule
let User = class User {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], User.prototype, "id", void 0);
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], User.prototype, "name", void 0);
User = tslib_1.__decorate([
    (0, lib_1.ObjectType)({ registerIn: () => UsersModule })
], User);
let CreateUserInput = class CreateUserInput {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], CreateUserInput.prototype, "name", void 0);
CreateUserInput = tslib_1.__decorate([
    (0, lib_1.InputType)({ registerIn: () => UsersModule })
], CreateUserInput);
let UserInterface = class UserInterface {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], UserInterface.prototype, "id", void 0);
UserInterface = tslib_1.__decorate([
    (0, lib_1.InterfaceType)({ registerIn: () => UsersModule })
], UserInterface);
// Types for ProductsModule
let Product = class Product {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], Product.prototype, "id", void 0);
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], Product.prototype, "name", void 0);
Product = tslib_1.__decorate([
    (0, lib_1.ObjectType)({ registerIn: () => ProductsModule })
], Product);
let CreateProductInput = class CreateProductInput {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], CreateProductInput.prototype, "name", void 0);
CreateProductInput = tslib_1.__decorate([
    (0, lib_1.InputType)({ registerIn: () => ProductsModule })
], CreateProductInput);
let ProductInterface = class ProductInterface {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], ProductInterface.prototype, "id", void 0);
ProductInterface = tslib_1.__decorate([
    (0, lib_1.InterfaceType)({ registerIn: () => ProductsModule })
], ProductInterface);
// Global types (no registerIn)
let GlobalEntity = class GlobalEntity {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], GlobalEntity.prototype, "id", void 0);
GlobalEntity = tslib_1.__decorate([
    (0, lib_1.ObjectType)()
], GlobalEntity);
let GlobalInput = class GlobalInput {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], GlobalInput.prototype, "value", void 0);
GlobalInput = tslib_1.__decorate([
    (0, lib_1.InputType)()
], GlobalInput);
// Enums
var UserStatus;
(function (UserStatus) {
    UserStatus["ACTIVE"] = "ACTIVE";
    UserStatus["INACTIVE"] = "INACTIVE";
})(UserStatus || (UserStatus = {}));
var ProductStatus;
(function (ProductStatus) {
    ProductStatus["AVAILABLE"] = "AVAILABLE";
    ProductStatus["SOLD_OUT"] = "SOLD_OUT";
})(ProductStatus || (ProductStatus = {}));
var GlobalStatus;
(function (GlobalStatus) {
    GlobalStatus["YES"] = "YES";
    GlobalStatus["NO"] = "NO";
})(GlobalStatus || (GlobalStatus = {}));
(0, lib_1.registerEnumType)(UserStatus, {
    name: 'UserStatus',
    registerIn: () => UsersModule,
});
(0, lib_1.registerEnumType)(ProductStatus, {
    name: 'ProductStatus',
    registerIn: () => ProductsModule,
});
(0, lib_1.registerEnumType)(GlobalStatus, {
    name: 'GlobalStatus',
});
// Unions - will be created in beforeAll after ObjectTypes are loaded
let UserUnion;
let ProductUnion;
describe('TypeDefinitionsGenerator with includeModules', () => {
    let generator;
    let storage;
    beforeAll(async () => {
        // Load all decorated class metadata
        lazy_metadata_storage_1.LazyMetadataStorage.load([
            User,
            CreateUserInput,
            UserInterface,
            Product,
            CreateProductInput,
            ProductInterface,
            GlobalEntity,
            GlobalInput,
        ]);
        // Create unions after ObjectTypes are loaded
        UserUnion = (0, lib_1.createUnionType)({
            name: 'UserUnion',
            types: () => [User, GlobalEntity],
            registerIn: () => UsersModule,
        });
        ProductUnion = (0, lib_1.createUnionType)({
            name: 'ProductUnion',
            types: () => [Product, GlobalEntity],
            registerIn: () => ProductsModule,
        });
        // Load remaining metadata
        lazy_metadata_storage_1.LazyMetadataStorage.load([]);
        lib_1.TypeMetadataStorage.compile();
        const moduleRef = await testing_1.Test.createTestingModule({
            imports: [lib_1.GraphQLSchemaBuilderModule],
        }).compile();
        generator = moduleRef.get(type_definitions_generator_1.TypeDefinitionsGenerator);
        storage = moduleRef.get(type_definitions_storage_1.TypeDefinitionsStorage);
    });
    afterAll(() => {
        lib_1.TypeMetadataStorage.clear();
    });
    describe('when includeModules is specified', () => {
        describe('UsersModule only', () => {
            beforeEach(() => {
                generator.generate({}, [UsersModule]);
            });
            it('should include ObjectTypes registered in UsersModule', () => {
                const userType = storage.getObjectTypeByTarget(User);
                expect(userType).toBeDefined();
            });
            it('should include global ObjectTypes (no registerIn)', () => {
                const globalType = storage.getObjectTypeByTarget(GlobalEntity);
                expect(globalType).toBeDefined();
            });
            it('should NOT include ObjectTypes registered in ProductsModule', () => {
                const productType = storage.getObjectTypeByTarget(Product);
                expect(productType).toBeUndefined();
            });
            it('should include InputTypes registered in UsersModule', () => {
                const inputType = storage.getInputTypeByTarget(CreateUserInput);
                expect(inputType).toBeDefined();
            });
            it('should NOT include InputTypes registered in ProductsModule', () => {
                const inputType = storage.getInputTypeByTarget(CreateProductInput);
                expect(inputType).toBeUndefined();
            });
            it('should include InterfaceTypes registered in UsersModule', () => {
                const interfaceType = storage.getInterfaceByTarget(UserInterface);
                expect(interfaceType).toBeDefined();
            });
            it('should NOT include InterfaceTypes registered in ProductsModule', () => {
                const interfaceType = storage.getInterfaceByTarget(ProductInterface);
                expect(interfaceType).toBeUndefined();
            });
            it('should include enums registered in UsersModule', () => {
                const enumDef = storage.getEnumByObject(UserStatus);
                expect(enumDef).toBeDefined();
            });
            it('should include global enums', () => {
                const enumDef = storage.getEnumByObject(GlobalStatus);
                expect(enumDef).toBeDefined();
            });
            it('should NOT include enums registered in ProductsModule', () => {
                const enumDef = storage.getEnumByObject(ProductStatus);
                expect(enumDef).toBeUndefined();
            });
        });
        describe('ProductsModule only', () => {
            beforeEach(() => {
                generator.generate({}, [ProductsModule]);
            });
            it('should include ObjectTypes registered in ProductsModule', () => {
                const productType = storage.getObjectTypeByTarget(Product);
                expect(productType).toBeDefined();
            });
            it('should NOT include ObjectTypes registered in UsersModule', () => {
                const userType = storage.getObjectTypeByTarget(User);
                expect(userType).toBeUndefined();
            });
            it('should include global ObjectTypes', () => {
                const globalType = storage.getObjectTypeByTarget(GlobalEntity);
                expect(globalType).toBeDefined();
            });
            it('should include InputTypes registered in ProductsModule', () => {
                const inputType = storage.getInputTypeByTarget(CreateProductInput);
                expect(inputType).toBeDefined();
            });
            it('should include global InputTypes', () => {
                const inputType = storage.getInputTypeByTarget(GlobalInput);
                expect(inputType).toBeDefined();
            });
        });
        describe('Both modules', () => {
            beforeEach(() => {
                generator.generate({}, [UsersModule, ProductsModule]);
            });
            it('should include ObjectTypes from both modules', () => {
                const userType = storage.getObjectTypeByTarget(User);
                const productType = storage.getObjectTypeByTarget(Product);
                expect(userType).toBeDefined();
                expect(productType).toBeDefined();
            });
            it('should include InputTypes from both modules', () => {
                const userInput = storage.getInputTypeByTarget(CreateUserInput);
                const productInput = storage.getInputTypeByTarget(CreateProductInput);
                expect(userInput).toBeDefined();
                expect(productInput).toBeDefined();
            });
            it('should include global types', () => {
                const globalType = storage.getObjectTypeByTarget(GlobalEntity);
                const globalInput = storage.getInputTypeByTarget(GlobalInput);
                expect(globalType).toBeDefined();
                expect(globalInput).toBeDefined();
            });
            it('should include enums from both modules and global', () => {
                const userEnum = storage.getEnumByObject(UserStatus);
                const productEnum = storage.getEnumByObject(ProductStatus);
                const globalEnum = storage.getEnumByObject(GlobalStatus);
                expect(userEnum).toBeDefined();
                expect(productEnum).toBeDefined();
                expect(globalEnum).toBeDefined();
            });
        });
    });
    describe('when includeModules is not specified (all types included)', () => {
        beforeEach(() => {
            generator.generate({});
        });
        it('should include all ObjectTypes', () => {
            const userType = storage.getObjectTypeByTarget(User);
            const productType = storage.getObjectTypeByTarget(Product);
            const globalType = storage.getObjectTypeByTarget(GlobalEntity);
            expect(userType).toBeDefined();
            expect(productType).toBeDefined();
            expect(globalType).toBeDefined();
        });
        it('should include all InputTypes', () => {
            const userInput = storage.getInputTypeByTarget(CreateUserInput);
            const productInput = storage.getInputTypeByTarget(CreateProductInput);
            const globalInput = storage.getInputTypeByTarget(GlobalInput);
            expect(userInput).toBeDefined();
            expect(productInput).toBeDefined();
            expect(globalInput).toBeDefined();
        });
        it('should include all enums', () => {
            const userEnum = storage.getEnumByObject(UserStatus);
            const productEnum = storage.getEnumByObject(ProductStatus);
            const globalEnum = storage.getEnumByObject(GlobalStatus);
            expect(userEnum).toBeDefined();
            expect(productEnum).toBeDefined();
            expect(globalEnum).toBeDefined();
        });
    });
    describe('TypeDefinitionsStorage.clear() functionality', () => {
        it('should clear previous definitions when generating new schema', () => {
            // First generate with UsersModule
            generator.generate({}, [UsersModule]);
            const userTypeFirst = storage.getObjectTypeByTarget(User);
            expect(userTypeFirst).toBeDefined();
            // Then generate with ProductsModule only
            generator.generate({}, [ProductsModule]);
            // User should no longer be present
            const userTypeSecond = storage.getObjectTypeByTarget(User);
            expect(userTypeSecond).toBeUndefined();
            // Product should now be present
            const productType = storage.getObjectTypeByTarget(Product);
            expect(productType).toBeDefined();
        });
    });
});
