// @nestjs/graphql/dist/tests/tests/utils/get-number-of-arguments.util.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../../lib/utils");
describe('getNumberOfArguments', () => {
    describe('when using function', () => {
        it('should return 0 for a 0-argument function', () => {
            function zeroArgFunction() { }
            expect((0, utils_1.getNumberOfArguments)(zeroArgFunction)).toBe(0);
        });
        it('should return 0 for a 0-argument function with body', () => {
            function zeroArgFunction() {
                if (1 == +'2') {
                    return false;
                }
                return true;
            }
            expect((0, utils_1.getNumberOfArguments)(zeroArgFunction)).toBe(0);
        });
        it('should return 1 for a 1-argument function', () => {
            function oneArgFunction(_arg1) { }
            expect((0, utils_1.getNumberOfArguments)(oneArgFunction)).toBe(1);
        });
        it('should return 1 for a 1-argument function with body', () => {
            function oneArgFunction(_arg1) {
                if (1 == +'2') {
                    return false;
                }
                return true;
            }
            expect((0, utils_1.getNumberOfArguments)(oneArgFunction)).toBe(1);
        });
        it('should return 2 for a 2-argument function', () => {
            function twoArgFunction(_arg1, _arg2) { }
            expect((0, utils_1.getNumberOfArguments)(twoArgFunction)).toBe(2);
        });
        it('should return 2 for a 2-argument function, even if it has a default parameter', () => {
            function twoArgFunction(_arg1, _arg2 = 'text') { }
            expect((0, utils_1.getNumberOfArguments)(twoArgFunction)).toBe(2);
        });
    });
    describe('when using arrow function', () => {
        it('should return 0 for a 0-argument function', () => {
            const zeroArgFunction = () => { };
            expect((0, utils_1.getNumberOfArguments)(zeroArgFunction)).toBe(0);
        });
        it('should return 1 for a 1-argument function', () => {
            const oneArgFunction = (_arg1) => { };
            expect((0, utils_1.getNumberOfArguments)(oneArgFunction)).toBe(1);
        });
        it('should return 2 for a 2-argument function', () => {
            const twoArgFunction = (_arg1, _arg2) => { };
            expect((0, utils_1.getNumberOfArguments)(twoArgFunction)).toBe(2);
        });
        it('should return 2 for a 2-argument function, even if it has a default parameter', () => {
            const twoArgFunction = (_arg1, _arg2 = 1) => { };
            expect((0, utils_1.getNumberOfArguments)(twoArgFunction)).toBe(2);
        });
    });
    describe('when using class method', () => {
        class TestClass {
            methodZeroArguments() { }
            methodOneArgument(_arg) { }
            methodTwoArguments(_arg1, _arg2) { }
            methodTwoArgumentsOneOptional(_arg1, _arg2 = ['raw']) { }
        }
        const instance = new TestClass();
        it('should return 0 for a 0-argument function', () => {
            expect((0, utils_1.getNumberOfArguments)(instance.methodZeroArguments)).toBe(0);
        });
        it('should return 1 for a 1-argument function', () => {
            expect((0, utils_1.getNumberOfArguments)(instance.methodOneArgument)).toBe(1);
        });
        it('should return 2 for a 2-argument function', () => {
            expect((0, utils_1.getNumberOfArguments)(instance.methodTwoArguments)).toBe(2);
        });
        it('should return 2 for a 2-argument function, even if it has a default parameter', () => {
            expect((0, utils_1.getNumberOfArguments)(instance.methodTwoArgumentsOneOptional)).toBe(2);
        });
    });
    describe('when using class static method', () => {
        class TestStaticClass {
            static methodZeroArguments() { }
            static methodZeroArgumentsWithBody() {
                if (1 == +'2') {
                    return false;
                }
                return true;
            }
            static methodOneArgument(_arg) { }
            static methodOneArgumentWithBody(_arg) {
                if (1 == +'2') {
                    return false;
                }
                return true;
            }
            static methodTwoArguments(_arg1, _arg2) { }
            static methodTwoArgumentsOneOptional(_arg1, _arg2 = ['raw']) { }
        }
        it('should return 0 for a 0-argument function', () => {
            expect((0, utils_1.getNumberOfArguments)(TestStaticClass.methodZeroArguments)).toBe(0);
        });
        it('should return 0 for a 0-argument function with body', () => {
            expect((0, utils_1.getNumberOfArguments)(TestStaticClass.methodZeroArgumentsWithBody)).toBe(0);
        });
        it('should return 1 for a 1-argument function', () => {
            expect((0, utils_1.getNumberOfArguments)(TestStaticClass.methodOneArgument)).toBe(1);
        });
        it('should return 1 for a 1-argument function with body', () => {
            expect((0, utils_1.getNumberOfArguments)(TestStaticClass.methodOneArgumentWithBody)).toBe(1);
        });
        it('should return 2 for a 2-argument function', () => {
            expect((0, utils_1.getNumberOfArguments)(TestStaticClass.methodTwoArguments)).toBe(2);
        });
        it('should return 2 for a 2-argument function, even if it has a default parameter', () => {
            expect((0, utils_1.getNumberOfArguments)(TestStaticClass.methodTwoArgumentsOneOptional)).toBe(2);
        });
    });
    describe('edge cases', () => {
        it('should not count an "empty" argument when the function ends with a comma', () => {
            // prettier-ignore
            function functionEndingWithComma(_arg1, _arg2) { }
            expect((0, utils_1.getNumberOfArguments)(functionEndingWithComma)).toBe(2);
        });
        it('should count correctly for multi-lined functions', () => {
            // prettier-ignore
            function multiLineFunction(_arg1 = 1, _arg2 = 2, _arg3 = 3) { }
            expect((0, utils_1.getNumberOfArguments)(multiLineFunction)).toBe(3);
        });
        it('should count correctly for functions containing arrays as default values', () => {
            function functionWithArray(_arg1 = 1, _arg2 = [1, 2, 3]) { }
            expect((0, utils_1.getNumberOfArguments)(functionWithArray)).toBe(2);
        });
        it('should count correctly for functions containing objects as default values', () => {
            function functionWithArray(_arg1 = 1, _arg2 = { a: 1, b: 2 }) { }
            expect((0, utils_1.getNumberOfArguments)(functionWithArray)).toBe(2);
        });
        it('should count correctly for functions containing both objects and arrays as default values', () => {
            function functionWithArray(_arg1 = [1, 2, 3], _arg2 = { a: 1, b: 2 }) { }
            expect((0, utils_1.getNumberOfArguments)(functionWithArray)).toBe(2);
        });
        it('should count correctly for arrow functions without parenthesis', () => {
            // prettier-ignore
            const leanArrowFunction = x => x + 2;
            expect((0, utils_1.getNumberOfArguments)(leanArrowFunction)).toBe(1);
        });
        it('should count correctly for async arrow functions without parenthesis', () => {
            // prettier-ignore
            const asyncLeanArrowFunction = async (x) => x + 2;
            expect((0, utils_1.getNumberOfArguments)(asyncLeanArrowFunction)).toBe(1);
        });
    });
});
describe('stringifyWithoutQuotes', () => {
    it('should stringify object correctly', () => {
        const obj = {
            name: '@tag',
            as: '@mytag',
        };
        expect((0, utils_1.stringifyWithoutQuotes)(obj)).toBe('{ name: "@tag", as: "@mytag" }');
    });
});
