// @nestjs/apollo/dist/tests/tests/subscriptions/app/app.module.js
"use strict";
var AppModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppModule = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const graphql_1 = require("@nestjs/graphql");
const drivers_1 = require("../../../lib/drivers");
const notification_module_1 = require("./notification.module");
let AppModule = AppModule_1 = class AppModule {
    static forRoot(options) {
        return {
            module: AppModule_1,
            imports: [
                notification_module_1.NotificationModule,
                graphql_1.GraphQLModule.forRoot({
                    driver: drivers_1.ApolloDriver,
                    includeStacktraceInErrorResponses: false,
                    context: options?.context,
                    autoSchemaFile: true,
                    subscriptions: options?.subscriptions,
                }),
            ],
        };
    }
};
exports.AppModule = AppModule;
exports.AppModule = AppModule = AppModule_1 = tslib_1.__decorate([
    (0, common_1.Module)({})
], AppModule);
