// @nestjs/apollo/dist/tests/tests/subscriptions/app/notification.resolver.js
"use strict";
var NotificationResolver_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotificationResolver = exports.pubSub = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const graphql_1 = require("@nestjs/graphql");
const graphql_subscriptions_1 = require("graphql-subscriptions");
const auth_guard_1 = require("./auth.guard");
const notification_1 = require("./notification");
exports.pubSub = new graphql_subscriptions_1.PubSub();
let NotificationResolver = NotificationResolver_1 = class NotificationResolver {
    constructor() {
        this.logger = new common_1.Logger(NotificationResolver_1.name);
    }
    getNotification() {
        return {
            message: 'Hello!',
        };
    }
    newNotification(id) {
        this.logger.log('User subscribed to newNotification');
        return exports.pubSub.asyncIterableIterator('newNotification');
    }
};
exports.NotificationResolver = NotificationResolver;
tslib_1.__decorate([
    (0, graphql_1.Query)(() => notification_1.Notification),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", void 0)
], NotificationResolver.prototype, "getNotification", null);
tslib_1.__decorate([
    (0, common_1.UseGuards)(auth_guard_1.AuthGuard),
    (0, graphql_1.Subscription)(() => notification_1.Notification, {
        filter(payload, variables, context) {
            return (context.user === payload.newNotification.recipient &&
                payload.newNotification.id === variables.id);
        },
    }),
    tslib_1.__param(0, (0, graphql_1.Args)('id', {
        nullable: false,
    })),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [String]),
    tslib_1.__metadata("design:returntype", void 0)
], NotificationResolver.prototype, "newNotification", null);
exports.NotificationResolver = NotificationResolver = NotificationResolver_1 = tslib_1.__decorate([
    (0, graphql_1.Resolver)(() => notification_1.Notification)
], NotificationResolver);
