// @nestjs/graphql/dist/tests/tests/type-factories/enum-union-options-validation.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const lib_1 = require("../../lib");
const type_factories_1 = require("../../lib/type-factories");
let MemberA = class MemberA {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => lib_1.ID),
    tslib_1.__metadata("design:type", String)
], MemberA.prototype, "id", void 0);
MemberA = tslib_1.__decorate([
    (0, lib_1.ObjectType)()
], MemberA);
let MemberB = class MemberB {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => lib_1.ID),
    tslib_1.__metadata("design:type", String)
], MemberB.prototype, "id", void 0);
MemberB = tslib_1.__decorate([
    (0, lib_1.ObjectType)()
], MemberB);
var SampleEnum;
(function (SampleEnum) {
    SampleEnum["A"] = "A";
    SampleEnum["B"] = "B";
})(SampleEnum || (SampleEnum = {}));
describe('registerEnumType / createUnionType options validation', () => {
    it('should throw a descriptive error when registerEnumType is called without options', () => {
        expect(() => type_factories_1.registerEnumType(SampleEnum)).toThrowError(/registerEnumType requires an "options" object/);
    });
    it('should throw a descriptive error when registerEnumType is called without a name', () => {
        expect(() => (0, type_factories_1.registerEnumType)(SampleEnum, {})).toThrowError(/non-empty "name"/);
    });
    it('should throw a descriptive error when createUnionType is called without options', () => {
        expect(() => type_factories_1.createUnionType()).toThrowError(/createUnionType requires an "options" object/);
    });
    it('should throw a descriptive error when createUnionType is called without a name', () => {
        expect(() => (0, type_factories_1.createUnionType)({
            types: () => [MemberA, MemberB],
        })).toThrowError(/non-empty "name"/);
    });
    it('should throw a descriptive error when createUnionType is called without a types function', () => {
        expect(() => (0, type_factories_1.createUnionType)({
            name: 'BadUnion',
        })).toThrowError(/"options\.types" to be a function/);
    });
    it('should accept a fully formed options object', () => {
        expect(() => (0, type_factories_1.registerEnumType)(SampleEnum, { name: 'SampleEnum' })).not.toThrow();
        expect(() => (0, type_factories_1.createUnionType)({
            name: 'SampleUnion',
            types: () => [MemberA, MemberB],
        })).not.toThrow();
    });
});
