// @nestjs/apollo/dist/tests/tests/code-first/recipes/models/ingredient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ingredient = void 0;
const tslib_1 = require("tslib");
const graphql_1 = require("@nestjs/graphql");
class NodeIngredient {
}
tslib_1.__decorate([
    (0, graphql_1.Field)((type) => graphql_1.ID, {
        defaultValue: 'Not applied',
        description: 'Not applied',
        deprecationReason: 'Not applied',
    }),
    tslib_1.__metadata("design:type", String)
], NodeIngredient.prototype, "id", void 0);
let BaseIngredient = class BaseIngredient extends NodeIngredient {
};
tslib_1.__decorate([
    (0, graphql_1.Field)({
        defaultValue: 'default',
        deprecationReason: 'is deprecated',
        description: 'ingredient base name',
        nullable: true,
    }),
    tslib_1.__metadata("design:type", String)
], BaseIngredient.prototype, "baseName", void 0);
BaseIngredient = tslib_1.__decorate([
    (0, graphql_1.ObjectType)({ isAbstract: true })
], BaseIngredient);
let Ingredient = class Ingredient extends BaseIngredient {
    constructor(ingredient) {
        super();
        Object.assign(this, ingredient);
    }
};
exports.Ingredient = Ingredient;
tslib_1.__decorate([
    (0, graphql_1.Field)({
        defaultValue: 'default',
        deprecationReason: 'is deprecated',
        description: 'ingredient name',
        nullable: true,
    }),
    tslib_1.__metadata("design:type", String)
], Ingredient.prototype, "name", void 0);
exports.Ingredient = Ingredient = tslib_1.__decorate([
    (0, graphql_1.ObjectType)(),
    tslib_1.__metadata("design:paramtypes", [Object])
], Ingredient);
