// @nestjs/apollo/dist/tests/tests/code-first/recipes/models/recipe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Recipe = exports.IRecipe = exports.Base = void 0;
const tslib_1 = require("tslib");
const graphql_1 = require("@nestjs/graphql");
const plugin_constants_1 = require("@nestjs/graphql/plugin/plugin-constants");
let Base = class Base {
};
exports.Base = Base;
tslib_1.__decorate([
    (0, graphql_1.Field)((type) => graphql_1.ID),
    tslib_1.__metadata("design:type", String)
], Base.prototype, "id", void 0);
exports.Base = Base = tslib_1.__decorate([
    (0, graphql_1.InterfaceType)()
], Base);
let IRecipe = class IRecipe extends Base {
};
exports.IRecipe = IRecipe;
tslib_1.__decorate([
    (0, graphql_1.Field)(),
    tslib_1.__metadata("design:type", String)
], IRecipe.prototype, "title", void 0);
exports.IRecipe = IRecipe = tslib_1.__decorate([
    (0, graphql_1.InterfaceType)({
        description: 'example interface',
        resolveType: (value) => {
            return Recipe;
        },
    })
], IRecipe);
let Recipe = class Recipe extends IRecipe {
    get averageRating() {
        return 0.5;
    }
    constructor(recipe) {
        super();
        Object.assign(this, recipe);
    }
    static [plugin_constants_1.METADATA_FACTORY_NAME]() {
        return {
            lastRate: {
                nullable: true,
                type: () => Number,
                description: 'last rate description',
            },
            tags: { nullable: false, type: () => [String] },
            averageRating: {
                description: "The recipe's average rating",
            },
            creationDate: {},
        };
    }
};
exports.Recipe = Recipe;
tslib_1.__decorate([
    (0, graphql_1.Field)({
        nullable: true,
        middleware: [
            async (ctx, next) => {
                const value = await next();
                return value ? 'Description: ' + value : 'Placeholder';
            },
        ],
    }),
    tslib_1.__metadata("design:type", String)
], Recipe.prototype, "description", void 0);
tslib_1.__decorate([
    (0, graphql_1.Field)(),
    tslib_1.__metadata("design:type", Date)
], Recipe.prototype, "creationDate", void 0);
tslib_1.__decorate([
    (0, graphql_1.Field)(),
    tslib_1.__metadata("design:type", Number),
    tslib_1.__metadata("design:paramtypes", [])
], Recipe.prototype, "averageRating", null);
exports.Recipe = Recipe = tslib_1.__decorate([
    (0, graphql_1.ObjectType)({ implements: IRecipe, description: 'recipe object type' }),
    tslib_1.__metadata("design:paramtypes", [Object])
], Recipe);
