// @nestjs/apollo/dist/tests/tests/code-first-federation/app.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplicationModule = void 0;
const tslib_1 = require("tslib");
const disabled_1 = require("@apollo/server/plugin/disabled");
const common_1 = require("@nestjs/common");
const graphql_1 = require("@nestjs/graphql");
const drivers_1 = require("../../lib/drivers");
const post_module_1 = require("./post/post.module");
const recipe_module_1 = require("./recipe/recipe.module");
const user_entity_1 = require("./user/user.entity");
const user_module_1 = require("./user/user.module");
const human_module_1 = require("./human/human.module");
let ApplicationModule = class ApplicationModule {
};
exports.ApplicationModule = ApplicationModule;
exports.ApplicationModule = ApplicationModule = tslib_1.__decorate([
    (0, common_1.Module)({
        imports: [
            user_module_1.UserModule,
            post_module_1.PostModule,
            recipe_module_1.RecipeModule,
            human_module_1.HumanModule,
            graphql_1.GraphQLModule.forRoot({
                inheritResolversFromInterfaces: true,
                driver: drivers_1.ApolloFederationDriver,
                includeStacktraceInErrorResponses: false,
                autoSchemaFile: true,
                buildSchemaOptions: {
                    orphanedTypes: [user_entity_1.User],
                },
                plugins: [(0, disabled_1.ApolloServerPluginInlineTraceDisabled)()],
            }),
        ],
    })
], ApplicationModule);
