// @nestjs/apollo/dist/tests/tests/code-first-federation/caching.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CachingApplicationModule = void 0;
const tslib_1 = require("tslib");
const server_plugin_response_cache_1 = tslib_1.__importDefault(require("@apollo/server-plugin-response-cache"));
const disabled_1 = require("@apollo/server/plugin/disabled");
const common_1 = require("@nestjs/common");
const graphql_1 = require("@nestjs/graphql");
const graphql_2 = require("graphql");
const drivers_1 = require("../../lib/drivers");
const post_module_1 = require("./post/post.module");
const recipe_module_1 = require("./recipe/recipe.module");
const user_entity_1 = require("./user/user.entity");
const user_module_1 = require("./user/user.module");
let CachingApplicationModule = class CachingApplicationModule {
};
exports.CachingApplicationModule = CachingApplicationModule;
exports.CachingApplicationModule = CachingApplicationModule = tslib_1.__decorate([
    (0, common_1.Module)({
        imports: [
            user_module_1.UserModule,
            post_module_1.PostModule,
            recipe_module_1.RecipeModule,
            graphql_1.GraphQLModule.forRoot({
                driver: drivers_1.ApolloFederationDriver,
                includeStacktraceInErrorResponses: false,
                autoSchemaFile: true,
                buildSchemaOptions: {
                    orphanedTypes: [user_entity_1.User],
                    directives: [
                        new graphql_2.GraphQLDirective({
                            name: 'cacheControl',
                            locations: [
                                graphql_2.DirectiveLocation.FIELD_DEFINITION,
                                graphql_2.DirectiveLocation.OBJECT,
                                graphql_2.DirectiveLocation.INTERFACE,
                                graphql_2.DirectiveLocation.UNION,
                            ],
                            args: {
                                maxAge: { type: graphql_2.GraphQLInt },
                                scope: {
                                    type: new graphql_2.GraphQLEnumType({
                                        name: 'CacheControlScope',
                                        values: {
                                            PUBLIC: {},
                                            PRIVATE: {},
                                        },
                                    }),
                                },
                                inheritMaxAge: { type: graphql_2.GraphQLBoolean },
                            },
                        }),
                    ],
                },
                plugins: [(0, server_plugin_response_cache_1.default)(), (0, disabled_1.ApolloServerPluginInlineTraceDisabled)()],
            }),
        ],
    })
], CachingApplicationModule);
