// @nestjs/graphql/dist/tests/tests/plugin/decorators/field.decorator.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const decorators_1 = require("../../../lib/decorators");
class Inner {
}
// It should expect the right primitive as defaultValue
class WrongPrimitive {
}
tslib_1.__decorate([
    (0, decorators_1.Field)(() => Boolean, { defaultValue: 'true' }),
    tslib_1.__metadata("design:type", Boolean)
], WrongPrimitive.prototype, "bool", void 0);
class CorrectPrimitive {
}
tslib_1.__decorate([
    (0, decorators_1.Field)(() => Boolean, { defaultValue: true }),
    tslib_1.__metadata("design:type", Boolean)
], CorrectPrimitive.prototype, "bool", void 0);
// It should expect the right object as defaultValue
class WrongObject {
}
tslib_1.__decorate([
    (0, decorators_1.Field)(() => Inner, { defaultValue: { success: true } }),
    tslib_1.__metadata("design:type", Inner)
], WrongObject.prototype, "inner", void 0);
class CorrectObject {
}
tslib_1.__decorate([
    (0, decorators_1.Field)(() => Inner, { defaultValue: { test: 'hello' } }),
    tslib_1.__metadata("design:type", Inner)
], CorrectObject.prototype, "inner", void 0);
// It should expect an Array as defaultValue
class WrongArray {
}
tslib_1.__decorate([
    (0, decorators_1.Field)(() => [Inner], { defaultValue: { test: 'test' } }),
    tslib_1.__metadata("design:type", Array)
], WrongArray.prototype, "inners", void 0);
class CorrectArray {
}
tslib_1.__decorate([
    (0, decorators_1.Field)(() => [Inner], { defaultValue: [{ test: 'test' }] }),
    tslib_1.__metadata("design:type", Inner)
], CorrectArray.prototype, "inner", void 0);
describe('Field decorator (defaultValue)', () => {
    it('TypeScript should not complain about the type mismatch', () => {
        expect(true).toBeTruthy();
    });
});
