// @nestjs/apollo/dist/tests/tests/code-first/recipes/recipes.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecipesService = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const recipe_1 = require("./models/recipe");
let RecipesService = class RecipesService {
    /**
     * MOCK
     * Put some real business logic here
     * Left for demonstration purposes
     */
    async create(data) {
        return {
            id: 3,
            ...data,
        };
    }
    async findOneById(id) {
        return {};
    }
    async findAll(recipesArgs) {
        return [
            new recipe_1.Recipe({
                id: '1',
                title: 'Pizza',
                description: 'Calzone',
                creationDate: new Date(),
            }),
            new recipe_1.Recipe({
                id: '2',
                title: 'Spaghetti',
                creationDate: new Date(),
            }),
        ];
    }
    async remove(id) {
        return true;
    }
};
exports.RecipesService = RecipesService;
exports.RecipesService = RecipesService = tslib_1.__decorate([
    (0, common_1.Injectable)()
], RecipesService);
