// @nestjs/apollo/dist/tests/tests/code-first/recipes/recipes.resolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecipesResolver = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const graphql_1 = require("@nestjs/graphql");
const graphql_subscriptions_1 = require("graphql-subscriptions");
const auth_guard_1 = require("../common/guards/auth.guard");
const filter_recipes_count_args_1 = require("./dto/filter-recipes-count.args");
const new_recipe_input_1 = require("./dto/new-recipe.input");
const recipes_args_1 = require("./dto/recipes.args");
const category_1 = require("./models/category");
const ingredient_1 = require("./models/ingredient");
const recipe_1 = require("./models/recipe");
const recipes_service_1 = require("./recipes.service");
const search_result_union_1 = require("./unions/search-result.union");
const pubSub = new graphql_subscriptions_1.PubSub();
let RecipesResolver = class RecipesResolver {
    constructor(recipesService) {
        this.recipesService = recipesService;
    }
    async recipe(id) {
        const recipe = await this.recipesService.findOneById(id);
        if (!recipe) {
            throw new common_1.NotFoundException(id);
        }
        return recipe;
    }
    async search() {
        return [
            new recipe_1.Recipe({ title: 'recipe' }),
            new ingredient_1.Ingredient({
                name: 'test',
            }),
        ];
    }
    categories() {
        return [new category_1.Category({ name: 'Category #1' })];
    }
    recipes(recipesArgs) {
        return this.recipesService.findAll(recipesArgs);
    }
    async addRecipe(newRecipeData) {
        const recipe = await this.recipesService.create(newRecipeData);
        pubSub.publish('recipeAdded', { recipeAdded: recipe });
        return recipe;
    }
    getIngredients(root) {
        return [new ingredient_1.Ingredient({ name: 'cherry' })];
    }
    count(filters) {
        return 10;
    }
    rating() {
        return 10;
    }
    async removeRecipe(id) {
        return this.recipesService.remove(id);
    }
    recipeAdded() {
        return pubSub.asyncIterableIterator('recipeAdded');
    }
};
exports.RecipesResolver = RecipesResolver;
tslib_1.__decorate([
    (0, common_1.UseGuards)(auth_guard_1.AuthGuard),
    (0, graphql_1.Query)((returns) => recipe_1.IRecipe, { description: 'get recipe by id' }),
    tslib_1.__param(0, (0, graphql_1.Args)('id', {
        defaultValue: '1',
        description: 'recipe id',
        nullable: true,
    })),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [String]),
    tslib_1.__metadata("design:returntype", Promise)
], RecipesResolver.prototype, "recipe", null);
tslib_1.__decorate([
    (0, graphql_1.Query)((returns) => [search_result_union_1.SearchResultUnion], { deprecationReason: 'test' }),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", Promise)
], RecipesResolver.prototype, "search", null);
tslib_1.__decorate([
    (0, graphql_1.Query)((returns) => [category_1.Category]),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", void 0)
], RecipesResolver.prototype, "categories", null);
tslib_1.__decorate([
    (0, graphql_1.Query)((returns) => [recipe_1.Recipe]),
    tslib_1.__param(0, (0, graphql_1.Args)()),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [recipes_args_1.RecipesArgs]),
    tslib_1.__metadata("design:returntype", Promise)
], RecipesResolver.prototype, "recipes", null);
tslib_1.__decorate([
    (0, graphql_1.Mutation)((returns) => recipe_1.Recipe),
    tslib_1.__param(0, (0, graphql_1.Args)('newRecipeData')),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [new_recipe_input_1.NewRecipeInput]),
    tslib_1.__metadata("design:returntype", Promise)
], RecipesResolver.prototype, "addRecipe", null);
tslib_1.__decorate([
    (0, graphql_1.ResolveField)('ingredients', () => [ingredient_1.Ingredient]),
    tslib_1.__param(0, (0, graphql_1.Parent)()),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", void 0)
], RecipesResolver.prototype, "getIngredients", null);
tslib_1.__decorate([
    (0, graphql_1.ResolveField)((type) => Number),
    tslib_1.__param(0, (0, graphql_1.Args)()),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [filter_recipes_count_args_1.FilterRecipesCountArgs]),
    tslib_1.__metadata("design:returntype", void 0)
], RecipesResolver.prototype, "count", null);
tslib_1.__decorate([
    (0, graphql_1.ResolveField)(),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", Number)
], RecipesResolver.prototype, "rating", null);
tslib_1.__decorate([
    (0, graphql_1.Mutation)((returns) => Boolean),
    tslib_1.__param(0, (0, graphql_1.Args)('id')),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [String]),
    tslib_1.__metadata("design:returntype", Promise)
], RecipesResolver.prototype, "removeRecipe", null);
tslib_1.__decorate([
    (0, graphql_1.Subscription)((returns) => recipe_1.Recipe, {
        description: 'subscription description',
    }),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", void 0)
], RecipesResolver.prototype, "recipeAdded", null);
exports.RecipesResolver = RecipesResolver = tslib_1.__decorate([
    (0, graphql_1.Resolver)((of) => recipe_1.Recipe),
    tslib_1.__metadata("design:paramtypes", [recipes_service_1.RecipesService])
], RecipesResolver);
