// @nestjs/graphql/dist/tests/tests/decorators/register-in.decorator.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const lib_1 = require("../../lib");
const lazy_metadata_storage_1 = require("../../lib/schema-builder/storages/lazy-metadata.storage");
/**
 * Integration tests for the registerIn decorator option.
 *
 * These tests verify that:
 * 1. The registerIn option is correctly passed through decorators to metadata storage
 * 2. Both direct module references and arrow functions work correctly
 * 3. Arrow functions properly resolve to avoid circular dependency issues
 */
// Mock modules
class AppModule {
}
class FeatureModule {
}
describe('registerIn decorator option', () => {
    afterEach(() => {
        lib_1.TypeMetadataStorage.clear();
    });
    describe('@ObjectType decorator', () => {
        it('should store registerIn as arrow function', () => {
            let TestType = class TestType {
            };
            tslib_1.__decorate([
                (0, lib_1.Field)(() => String),
                tslib_1.__metadata("design:type", String)
            ], TestType.prototype, "name", void 0);
            TestType = tslib_1.__decorate([
                (0, lib_1.ObjectType)({ registerIn: () => FeatureModule })
            ], TestType);
            lazy_metadata_storage_1.LazyMetadataStorage.load([TestType]);
            lib_1.TypeMetadataStorage.compile();
            const metadata = lib_1.TypeMetadataStorage.getObjectTypeMetadataByTarget(TestType);
            expect(metadata).toBeDefined();
            expect(metadata.registerIn).toBeDefined();
            // Arrow function should be stored
            expect(typeof metadata.registerIn).toBe('function');
            // When called, it should return the module
            const resolvedModule = metadata.registerIn();
            expect(resolvedModule).toBe(FeatureModule);
        });
        it('should store registerIn as direct module reference', () => {
            let TestType = class TestType {
            };
            tslib_1.__decorate([
                (0, lib_1.Field)(() => String),
                tslib_1.__metadata("design:type", String)
            ], TestType.prototype, "name", void 0);
            TestType = tslib_1.__decorate([
                (0, lib_1.ObjectType)({ registerIn: AppModule })
            ], TestType);
            lazy_metadata_storage_1.LazyMetadataStorage.load([TestType]);
            lib_1.TypeMetadataStorage.compile();
            const metadata = lib_1.TypeMetadataStorage.getObjectTypeMetadataByTarget(TestType);
            expect(metadata).toBeDefined();
            expect(metadata.registerIn).toBe(AppModule);
        });
        it('should work without registerIn option (included in all schemas)', () => {
            let TestType = class TestType {
            };
            tslib_1.__decorate([
                (0, lib_1.Field)(() => String),
                tslib_1.__metadata("design:type", String)
            ], TestType.prototype, "name", void 0);
            TestType = tslib_1.__decorate([
                (0, lib_1.ObjectType)()
            ], TestType);
            lazy_metadata_storage_1.LazyMetadataStorage.load([TestType]);
            lib_1.TypeMetadataStorage.compile();
            const metadata = lib_1.TypeMetadataStorage.getObjectTypeMetadataByTarget(TestType);
            expect(metadata).toBeDefined();
            expect(metadata.registerIn).toBeUndefined();
        });
        it('should work with name and registerIn option', () => {
            let TestType = class TestType {
            };
            tslib_1.__decorate([
                (0, lib_1.Field)(() => String),
                tslib_1.__metadata("design:type", String)
            ], TestType.prototype, "name", void 0);
            TestType = tslib_1.__decorate([
                (0, lib_1.ObjectType)('CustomName', { registerIn: () => FeatureModule })
            ], TestType);
            lazy_metadata_storage_1.LazyMetadataStorage.load([TestType]);
            lib_1.TypeMetadataStorage.compile();
            const metadata = lib_1.TypeMetadataStorage.getObjectTypeMetadataByTarget(TestType);
            expect(metadata).toBeDefined();
            expect(metadata.name).toBe('CustomName');
            expect(metadata.registerIn()).toBe(FeatureModule);
        });
    });
    describe('@InputType decorator', () => {
        it('should store registerIn option', () => {
            let TestInput = class TestInput {
            };
            tslib_1.__decorate([
                (0, lib_1.Field)(() => String),
                tslib_1.__metadata("design:type", String)
            ], TestInput.prototype, "name", void 0);
            TestInput = tslib_1.__decorate([
                (0, lib_1.InputType)({ registerIn: () => FeatureModule })
            ], TestInput);
            lazy_metadata_storage_1.LazyMetadataStorage.load([TestInput]);
            lib_1.TypeMetadataStorage.compile();
            const metadata = lib_1.TypeMetadataStorage.getInputTypeMetadataByTarget(TestInput);
            expect(metadata).toBeDefined();
            expect(metadata.registerIn()).toBe(FeatureModule);
        });
        it('should work with name and registerIn option', () => {
            let TestInput = class TestInput {
            };
            tslib_1.__decorate([
                (0, lib_1.Field)(() => String),
                tslib_1.__metadata("design:type", String)
            ], TestInput.prototype, "value", void 0);
            TestInput = tslib_1.__decorate([
                (0, lib_1.InputType)('CustomInput', { registerIn: () => AppModule })
            ], TestInput);
            lazy_metadata_storage_1.LazyMetadataStorage.load([TestInput]);
            lib_1.TypeMetadataStorage.compile();
            const metadata = lib_1.TypeMetadataStorage.getInputTypeMetadataByTarget(TestInput);
            expect(metadata).toBeDefined();
            expect(metadata.name).toBe('CustomInput');
            expect(metadata.registerIn()).toBe(AppModule);
        });
    });
    describe('@InterfaceType decorator', () => {
        it('should store registerIn option', () => {
            let TestInterface = class TestInterface {
            };
            tslib_1.__decorate([
                (0, lib_1.Field)(() => String),
                tslib_1.__metadata("design:type", String)
            ], TestInterface.prototype, "id", void 0);
            TestInterface = tslib_1.__decorate([
                (0, lib_1.InterfaceType)({ registerIn: () => FeatureModule })
            ], TestInterface);
            lazy_metadata_storage_1.LazyMetadataStorage.load([TestInterface]);
            lib_1.TypeMetadataStorage.compile();
            const metadata = lib_1.TypeMetadataStorage.getInterfaceMetadataByTarget(TestInterface);
            expect(metadata).toBeDefined();
            expect(metadata.registerIn()).toBe(FeatureModule);
        });
    });
    describe('@ArgsType decorator', () => {
        it('should store registerIn option', () => {
            let TestArgs = class TestArgs {
            };
            tslib_1.__decorate([
                (0, lib_1.Field)(() => String),
                tslib_1.__metadata("design:type", String)
            ], TestArgs.prototype, "query", void 0);
            TestArgs = tslib_1.__decorate([
                (0, lib_1.ArgsType)({ registerIn: () => AppModule })
            ], TestArgs);
            lazy_metadata_storage_1.LazyMetadataStorage.load([TestArgs]);
            lib_1.TypeMetadataStorage.compile();
            const metadata = lib_1.TypeMetadataStorage.getArgumentsMetadataByTarget(TestArgs);
            expect(metadata).toBeDefined();
            expect(metadata.registerIn()).toBe(AppModule);
        });
    });
    describe('registerEnumType function', () => {
        it('should store registerIn option', () => {
            let TestStatus;
            (function (TestStatus) {
                TestStatus["ACTIVE"] = "ACTIVE";
                TestStatus["INACTIVE"] = "INACTIVE";
            })(TestStatus || (TestStatus = {}));
            (0, lib_1.registerEnumType)(TestStatus, {
                name: 'TestStatus',
                registerIn: () => FeatureModule,
            });
            lazy_metadata_storage_1.LazyMetadataStorage.load([]);
            lib_1.TypeMetadataStorage.compile();
            const allEnums = lib_1.TypeMetadataStorage.getEnumsMetadata();
            const metadata = allEnums.find((e) => e.name === 'TestStatus');
            expect(metadata).toBeDefined();
            expect(metadata.registerIn()).toBe(FeatureModule);
        });
        it('should work with direct module reference', () => {
            let DirectRefStatus;
            (function (DirectRefStatus) {
                DirectRefStatus["YES"] = "YES";
                DirectRefStatus["NO"] = "NO";
            })(DirectRefStatus || (DirectRefStatus = {}));
            (0, lib_1.registerEnumType)(DirectRefStatus, {
                name: 'DirectRefStatus',
                registerIn: AppModule,
            });
            lazy_metadata_storage_1.LazyMetadataStorage.load([]);
            lib_1.TypeMetadataStorage.compile();
            const allEnums = lib_1.TypeMetadataStorage.getEnumsMetadata();
            const metadata = allEnums.find((e) => e.name === 'DirectRefStatus');
            expect(metadata).toBeDefined();
            expect(metadata.registerIn).toBe(AppModule);
        });
    });
    describe('createUnionType function', () => {
        it('should store registerIn option', () => {
            let UnionMemberA = class UnionMemberA {
            };
            tslib_1.__decorate([
                (0, lib_1.Field)(() => String),
                tslib_1.__metadata("design:type", String)
            ], UnionMemberA.prototype, "a", void 0);
            UnionMemberA = tslib_1.__decorate([
                (0, lib_1.ObjectType)()
            ], UnionMemberA);
            let UnionMemberB = class UnionMemberB {
            };
            tslib_1.__decorate([
                (0, lib_1.Field)(() => String),
                tslib_1.__metadata("design:type", String)
            ], UnionMemberB.prototype, "b", void 0);
            UnionMemberB = tslib_1.__decorate([
                (0, lib_1.ObjectType)()
            ], UnionMemberB);
            lazy_metadata_storage_1.LazyMetadataStorage.load([UnionMemberA, UnionMemberB]);
            const TestUnion = (0, lib_1.createUnionType)({
                name: 'TestUnion',
                types: () => [UnionMemberA, UnionMemberB],
                registerIn: () => FeatureModule,
            });
            lazy_metadata_storage_1.LazyMetadataStorage.load([]);
            lib_1.TypeMetadataStorage.compile();
            const allUnions = lib_1.TypeMetadataStorage.getUnionsMetadata();
            const metadata = allUnions.find((u) => u.name === 'TestUnion');
            expect(metadata).toBeDefined();
            expect(metadata.registerIn()).toBe(FeatureModule);
        });
    });
    describe('circular dependency handling', () => {
        it('should handle deferred module resolution via arrow function', () => {
            // Simulate circular dependency scenario where module is not yet defined
            let DeferredModule;
            let DeferredType = class DeferredType {
            };
            tslib_1.__decorate([
                (0, lib_1.Field)(() => String),
                tslib_1.__metadata("design:type", String)
            ], DeferredType.prototype, "name", void 0);
            DeferredType = tslib_1.__decorate([
                (0, lib_1.ObjectType)({ registerIn: () => DeferredModule })
            ], DeferredType);
            // Define module after decorator is applied
            DeferredModule = class DeferredModuleClass {
            };
            lazy_metadata_storage_1.LazyMetadataStorage.load([DeferredType]);
            lib_1.TypeMetadataStorage.compile();
            const metadata = lib_1.TypeMetadataStorage.getObjectTypeMetadataByTarget(DeferredType);
            expect(metadata).toBeDefined();
            // Arrow function should correctly resolve the module
            const resolvedModule = metadata.registerIn();
            expect(resolvedModule).toBe(DeferredModule);
        });
    });
});
