// @nestjs/apollo/dist/tests/tests/graphql/graphiql-playground.module.js
"use strict";
var GraphiQLPlaygroundModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphiQLPlaygroundModule = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const graphql_1 = require("@nestjs/graphql");
const path_1 = require("path");
const drivers_1 = require("../../lib/drivers");
const cats_module_1 = require("./cats/cats.module");
let GraphiQLPlaygroundModule = GraphiQLPlaygroundModule_1 = class GraphiQLPlaygroundModule {
    static withEnabled() {
        return {
            module: GraphiQLPlaygroundModule_1,
            imports: [
                graphql_1.GraphQLModule.forRoot({
                    driver: drivers_1.ApolloDriver,
                    csrfPrevention: false,
                    graphiql: true,
                    typePaths: [(0, path_1.join)(__dirname, '**', '*.graphql')],
                }),
            ],
        };
    }
    static withEnabledAndCustomized(options) {
        return {
            module: GraphiQLPlaygroundModule_1,
            imports: [
                graphql_1.GraphQLModule.forRoot({
                    driver: drivers_1.ApolloDriver,
                    csrfPrevention: false,
                    typePaths: [(0, path_1.join)(__dirname, '**', '*.graphql')],
                    graphiql: options,
                }),
            ],
        };
    }
};
exports.GraphiQLPlaygroundModule = GraphiQLPlaygroundModule;
exports.GraphiQLPlaygroundModule = GraphiQLPlaygroundModule = GraphiQLPlaygroundModule_1 = tslib_1.__decorate([
    (0, common_1.Module)({
        imports: [cats_module_1.CatsModule],
    })
], GraphiQLPlaygroundModule);
