// @nestjs/apollo/dist/tests/tests/graphql/transform-auto-schema-file.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransformAutoSchemaFileModule = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const graphql_1 = require("@nestjs/graphql");
const graphql_2 = require("graphql");
const drivers_1 = require("../../lib/drivers");
const directions_module_1 = require("../code-first/directions/directions.module");
const recipes_module_1 = require("../code-first/recipes/recipes.module");
let TransformAutoSchemaFileModule = class TransformAutoSchemaFileModule {
};
exports.TransformAutoSchemaFileModule = TransformAutoSchemaFileModule;
exports.TransformAutoSchemaFileModule = TransformAutoSchemaFileModule = tslib_1.__decorate([
    (0, common_1.Module)({
        imports: [
            recipes_module_1.RecipesModule,
            directions_module_1.DirectionsModule,
            graphql_1.GraphQLModule.forRoot({
                driver: drivers_1.ApolloDriver,
                autoSchemaFile: 'schema.graphql',
                transformSchema: (schema) => (0, graphql_2.lexicographicSortSchema)(schema),
                transformAutoSchemaFile: true,
            }),
        ],
    })
], TransformAutoSchemaFileModule);
