// @nestjs/apollo/dist/tests/tests/e2e/code-first.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("@nestjs/graphql");
const testing_1 = require("@nestjs/testing");
const graphql_tag_1 = require("graphql-tag");
const app_module_1 = require("../code-first/app.module");
const assertion_utils_1 = require("../utils/assertion-utils");
describe('Code-first', () => {
    let app;
    let apolloClient;
    beforeEach(async () => {
        const module = await testing_1.Test.createTestingModule({
            imports: [app_module_1.ApplicationModule],
        }).compile();
        app = module.createNestApplication();
        await app.init();
        const graphqlModule = app.get(graphql_1.GraphQLModule);
        apolloClient = graphqlModule.graphQlAdapter?.instance;
    });
    it('should return the categories result', async () => {
        const response = await apolloClient.executeOperation({
            query: (0, graphql_tag_1.gql) `
        {
          categories {
            name
            description
            tags
          }
        }
      `,
        });
        (0, assertion_utils_1.expectSingleResult)(response).toEqual({
            categories: [
                {
                    name: 'Category #1',
                    description: 'default value',
                    tags: [],
                },
            ],
        });
    });
    it('should return the search result', async () => {
        const response = await apolloClient.executeOperation({
            query: (0, graphql_tag_1.gql) `
        {
          search {
            ... on Recipe {
              title
            }
            ... on Ingredient {
              name
            }
          }
        }
      `,
        });
        (0, assertion_utils_1.expectSingleResult)(response).toEqual({
            search: [
                {
                    title: 'recipe',
                },
                {
                    name: 'test',
                },
            ],
        });
    });
    it(`should return query result`, async () => {
        const response = await apolloClient.executeOperation({
            query: (0, graphql_tag_1.gql) `
        {
          recipes {
            id
            description
            ingredients {
              name
            }
            rating
            interfaceResolver
            averageRating
          }
        }
      `,
        });
        (0, assertion_utils_1.expectSingleResult)(response).toEqual({
            recipes: [
                {
                    id: '1',
                    description: 'Description: Calzone',
                    ingredients: [
                        {
                            name: 'cherry',
                        },
                    ],
                    rating: 10,
                    interfaceResolver: true,
                    averageRating: 0.5,
                },
                {
                    id: '2',
                    description: 'Placeholder',
                    ingredients: [
                        {
                            name: 'cherry',
                        },
                    ],
                    rating: 10,
                    interfaceResolver: true,
                    averageRating: 0.5,
                },
            ],
        });
    });
    it(`should return query result`, async () => {
        const response = await apolloClient.executeOperation({
            query: (0, graphql_tag_1.gql) `
        {
          recipes {
            id
            ingredients {
              name
            }
            rating
            averageRating
          }
        }
      `,
        });
        (0, assertion_utils_1.expectSingleResult)(response).toEqual({
            recipes: [
                {
                    id: '1',
                    ingredients: [
                        {
                            name: 'cherry',
                        },
                    ],
                    rating: 10,
                    averageRating: 0.5,
                },
                {
                    id: '2',
                    ingredients: [
                        {
                            name: 'cherry',
                        },
                    ],
                    rating: 10,
                    averageRating: 0.5,
                },
            ],
        });
    });
    afterEach(async () => {
        await app.close();
    });
});
