// @nestjs/apollo/dist/tests/tests/e2e/code-first-federation-caching.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("@nestjs/graphql");
const graphql_constants_1 = require("@nestjs/graphql/graphql.constants");
const testing_1 = require("@nestjs/testing");
const graphql_2 = require("graphql");
const graphql_tag_1 = require("graphql-tag");
const caching_module_1 = require("../code-first-federation/caching.module");
const post_entity_1 = require("../code-first-federation/post/post.entity");
const post_resolver_1 = require("../code-first-federation/post/post.resolver");
const post_service_1 = require("../code-first-federation/post/post.service");
const irecipe_resolver_1 = require("../code-first-federation/recipe/irecipe.resolver");
const user_resolver_1 = require("../code-first-federation/user/user.resolver");
const printed_schema_with_cache_control_snapshot_1 = require("../utils/printed-schema-with-cache-control.snapshot");
describe.skip('Code-first - Federation with caching', () => {
    describe('generated schema', () => {
        let schema;
        beforeAll(async () => {
            const moduleRef = await testing_1.Test.createTestingModule({
                imports: [graphql_1.GraphQLSchemaBuilderModule],
            }).compile();
            const schemaFactory = moduleRef.get(graphql_1.GraphQLSchemaFactory);
            schema = await schemaFactory.create([post_resolver_1.PostResolver, irecipe_resolver_1.IRecipeResolver, user_resolver_1.UserResolver], {
                directives: [
                    new graphql_2.GraphQLDirective({
                        name: 'cacheControl',
                        locations: [
                            graphql_2.DirectiveLocation.FIELD_DEFINITION,
                            graphql_2.DirectiveLocation.OBJECT,
                            graphql_2.DirectiveLocation.INTERFACE,
                        ],
                        args: {
                            maxAge: { type: graphql_2.GraphQLInt },
                            scope: {
                                type: new graphql_2.GraphQLEnumType({
                                    name: 'CacheControlScope',
                                    values: {
                                        PUBLIC: {},
                                        PRIVATE: {},
                                    },
                                }),
                            },
                            inheritMaxAge: { type: graphql_2.GraphQLBoolean },
                        },
                    }),
                ],
            });
            let introspectionSchema = await (await (0, graphql_2.graphql)({ schema, source: (0, graphql_2.getIntrospectionQuery)() })).data.__schema;
        });
        it('should be valid', async () => {
            expect(schema).toBeInstanceOf(graphql_2.GraphQLSchema);
        });
        it('should match schema snapshot', () => {
            expect(graphql_constants_1.GRAPHQL_SDL_FILE_HEADER + (0, graphql_2.printSchema)(schema)).toEqual(printed_schema_with_cache_control_snapshot_1.printedSchemaSnapshot);
        });
    });
    describe('enabled cache', () => {
        let app;
        let apolloClient;
        let postService;
        beforeEach(async () => {
            const module = await testing_1.Test.createTestingModule({
                imports: [caching_module_1.CachingApplicationModule],
                providers: [
                    {
                        provide: post_service_1.PostService,
                        useValue: {
                            findOne: () => new post_entity_1.Post({ id: 1, title: 'HELLO WORLD' }),
                        },
                    },
                ],
            }).compile();
            app = module.createNestApplication();
            await app.init();
            const graphqlModule = app.get(graphql_1.GraphQLModule);
            apolloClient = graphqlModule.graphQlAdapter?.instance;
            postService = module.get(post_service_1.PostService);
        });
        it('findOne should called once', async () => {
            const spy = vi.spyOn(postService, 'findOne');
            const request = {
                query: (0, graphql_tag_1.gql) `
          {
            findPost(id: 1) {
              id
              title
            }
          }
        `,
            };
            await apolloClient.executeOperation(request);
            await apolloClient.executeOperation(request);
            expect(spy).toHaveBeenCalledTimes(1);
        });
    });
});
