// @nestjs/apollo/dist/tests/tests/e2e/code-first-inheritance.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const graphql_1 = require("@nestjs/graphql");
const testing_1 = require("@nestjs/testing");
const graphql_tag_1 = require("graphql-tag");
const lib_1 = require("../../lib");
const assertion_utils_1 = require("../utils/assertion-utils");
let MyArgs = class MyArgs {
};
tslib_1.__decorate([
    (0, graphql_1.Field)(),
    tslib_1.__metadata("design:type", Number)
], MyArgs.prototype, "myInput", void 0);
MyArgs = tslib_1.__decorate([
    (0, graphql_1.InputType)()
], MyArgs);
let MyBaseResolver = class MyBaseResolver {
    myQuery() {
        return 10;
    }
    myMutation(args) {
        return args.myInput;
    }
};
tslib_1.__decorate([
    (0, graphql_1.Query)(() => graphql_1.Int),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", Number)
], MyBaseResolver.prototype, "myQuery", null);
tslib_1.__decorate([
    (0, graphql_1.Mutation)(() => graphql_1.Int),
    tslib_1.__param(0, (0, graphql_1.Args)({ type: () => MyArgs, name: 'input' })),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [MyArgs]),
    tslib_1.__metadata("design:returntype", Number)
], MyBaseResolver.prototype, "myMutation", null);
MyBaseResolver = tslib_1.__decorate([
    (0, graphql_1.Resolver)({ isAbstract: true })
], MyBaseResolver);
let MyResolver = class MyResolver extends MyBaseResolver {
};
MyResolver = tslib_1.__decorate([
    (0, graphql_1.Resolver)()
], MyResolver);
let MyModule = class MyModule {
};
MyModule = tslib_1.__decorate([
    (0, common_1.Module)({
        imports: [
            graphql_1.GraphQLModule.forRoot({
                driver: lib_1.ApolloDriver,
                autoSchemaFile: true,
            }),
        ],
        providers: [MyResolver],
    })
], MyModule);
describe('code-first with a resolver that uses inheritance', () => {
    let app;
    let apolloClient;
    beforeEach(async () => {
        const module = await testing_1.Test.createTestingModule({
            imports: [MyModule],
        }).compile();
        app = module.createNestApplication();
        await app.init();
        const graphqlModule = app.get(graphql_1.GraphQLModule);
        apolloClient = graphqlModule.graphQlAdapter?.instance;
    });
    afterEach(async () => {
        await app.close();
    });
    function createTest() {
        it("resolver inherits schema from its' parent class", async () => {
            const mutation = (0, graphql_tag_1.gql) `
        mutation MyTest {
          myMutation(input: { myInput: 5 })
        }
      `;
            const response = await apolloClient.executeOperation({
                query: mutation,
            });
            (0, assertion_utils_1.expectSingleResult)(response).toEqual({
                myMutation: 5,
            });
        });
    }
    createTest();
    createTest();
});
