// @nestjs/apollo/dist/tests/tests/e2e/code-first-federation.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("@nestjs/graphql");
const testing_1 = require("@nestjs/testing");
const graphql_tag_1 = require("graphql-tag");
const app_module_1 = require("../code-first-federation/app.module");
const assertion_utils_1 = require("../utils/assertion-utils");
describe('Code-first - Federation', () => {
    let app;
    let apolloClient;
    beforeEach(async () => {
        const module = await testing_1.Test.createTestingModule({
            imports: [app_module_1.ApplicationModule],
        }).compile();
        app = module.createNestApplication();
        await app.init();
        const graphqlModule = app.get(graphql_1.GraphQLModule);
        apolloClient = graphqlModule.graphQlAdapter?.instance;
    });
    it(`should return query result`, async () => {
        const response = await apolloClient.executeOperation({
            query: (0, graphql_tag_1.gql) `
        {
          _service {
            sdl
          }
        }
      `,
        });
        (0, assertion_utils_1.expectSingleResult)(response).toMatchSnapshot();
    });
    it('should return the search result', async () => {
        const response = await apolloClient.executeOperation({
            query: (0, graphql_tag_1.gql) `
        {
          search {
            ... on Post {
              title
            }
            ... on User {
              id
            }
            __typename
          }
        }
      `,
        });
        (0, assertion_utils_1.expectSingleResult)(response).toEqual({
            search: [
                {
                    id: '1',
                    __typename: 'User',
                },
                {
                    title: 'lorem ipsum',
                    __typename: 'Post',
                },
            ],
        });
    });
    it(`should return query result`, async () => {
        const response = await apolloClient.executeOperation({
            query: (0, graphql_tag_1.gql) `
        {
          recipe {
            id
            title
            ... on Recipe {
              description
            }
          }
        }
      `,
        });
        (0, assertion_utils_1.expectSingleResult)(response).toEqual({
            recipe: {
                id: '1',
                title: 'Recipe',
                description: 'Interface description',
            },
        });
    });
    it('should return the human result with resolve interface', async () => {
        const response = await apolloClient.executeOperation({
            query: (0, graphql_tag_1.gql) `
        {
          humans {
            id
            name
            friends {
              id
              name
            }
          }
        }
      `,
        });
        (0, assertion_utils_1.expectSingleResult)(response).toEqual({
            humans: [
                {
                    id: '1',
                    name: 'Bob',
                    friends: [{ id: '3', name: 'Peter' }],
                },
                {
                    id: '2',
                    name: 'Alice',
                    friends: [{ id: '3', name: 'Peter' }],
                },
            ],
        });
    });
    afterEach(async () => {
        await app.close();
    });
});
