// @nestjs/apollo/dist/tests/tests/e2e/code-first-register-in.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("@nestjs/graphql");
const testing_1 = require("@nestjs/testing");
const graphql_tag_1 = require("graphql-tag");
const lib_1 = require("../../lib");
const products_module_1 = require("../code-first-register-in/products/products.module");
const users_module_1 = require("../code-first-register-in/users/users.module");
const assertion_utils_1 = require("../utils/assertion-utils");
const buildApp = async (include) => {
    const moduleRef = await testing_1.Test.createTestingModule({
        imports: [
            users_module_1.UsersModule,
            products_module_1.ProductsModule,
            graphql_1.GraphQLModule.forRoot({
                driver: lib_1.ApolloDriver,
                autoSchemaFile: true,
                includeStacktraceInErrorResponses: false,
                ...(include ? { include } : {}),
            }),
        ],
    }).compile();
    const app = moduleRef.createNestApplication();
    await app.init();
    const apolloClient = app.get(graphql_1.GraphQLModule)
        .graphQlAdapter?.instance;
    return { app, apolloClient };
};
const queryTypeName = async (apolloClient, name) => {
    const response = await apolloClient.executeOperation({
        query: (0, graphql_tag_1.gql) `
      query TypeByName($name: String!) {
        __type(name: $name) {
          name
        }
      }
    `,
        variables: { name },
    });
    return (0, assertion_utils_1.expectSingleResult)(response);
};
describe('Code-first - registerIn', () => {
    describe('include: [UsersModule]', () => {
        let app;
        let apolloClient;
        beforeAll(async () => {
            ({ app, apolloClient } = await buildApp([users_module_1.UsersModule]));
        });
        afterAll(async () => {
            await app.close();
        });
        it('should expose User type', async () => {
            (await queryTypeName(apolloClient, 'User')).toEqual({
                __type: { name: 'User' },
            });
        });
        it('should expose CreateUserInput type', async () => {
            (await queryTypeName(apolloClient, 'CreateUserInput')).toEqual({
                __type: { name: 'CreateUserInput' },
            });
        });
        it('should expose AppInfo (no registerIn) referenced by User', async () => {
            (await queryTypeName(apolloClient, 'AppInfo')).toEqual({
                __type: { name: 'AppInfo' },
            });
        });
        it('should NOT expose Product type', async () => {
            (await queryTypeName(apolloClient, 'Product')).toEqual({ __type: null });
        });
        it('should NOT expose products query field', async () => {
            const response = await apolloClient.executeOperation({
                query: (0, graphql_tag_1.gql) `
          {
            __schema {
              queryType {
                fields {
                  name
                }
              }
            }
          }
        `,
            });
            const data = response.body.singleResult.data;
            const fieldNames = data.__schema.queryType.fields.map((f) => f.name);
            expect(fieldNames).toContain('users');
            expect(fieldNames).not.toContain('products');
        });
        it('should resolve users query', async () => {
            const response = await apolloClient.executeOperation({
                query: (0, graphql_tag_1.gql) `
          {
            users {
              id
              name
              info {
                version
              }
            }
          }
        `,
            });
            (0, assertion_utils_1.expectSingleResult)(response).toEqual({
                users: [{ id: '1', name: 'Ada', info: { version: '1.0.0' } }],
            });
        });
        it('should resolve createUser mutation', async () => {
            const response = await apolloClient.executeOperation({
                query: (0, graphql_tag_1.gql) `
          mutation CreateUser($input: CreateUserInput!) {
            createUser(input: $input) {
              id
              name
            }
          }
        `,
                variables: { input: { name: 'Grace' } },
            });
            (0, assertion_utils_1.expectSingleResult)(response).toEqual({
                createUser: { id: '2', name: 'Grace' },
            });
        });
    });
    describe('include: [UsersModule, ProductsModule]', () => {
        let app;
        let apolloClient;
        beforeAll(async () => {
            ({ app, apolloClient } = await buildApp([users_module_1.UsersModule, products_module_1.ProductsModule]));
        });
        afterAll(async () => {
            await app.close();
        });
        it('should expose both User and Product types', async () => {
            (await queryTypeName(apolloClient, 'User')).toEqual({
                __type: { name: 'User' },
            });
            (await queryTypeName(apolloClient, 'Product')).toEqual({
                __type: { name: 'Product' },
            });
        });
        it('should resolve users and products queries', async () => {
            const response = await apolloClient.executeOperation({
                query: (0, graphql_tag_1.gql) `
          {
            users {
              id
              name
            }
            products {
              id
              name
            }
          }
        `,
            });
            (0, assertion_utils_1.expectSingleResult)(response).toEqual({
                users: [{ id: '1', name: 'Ada' }],
                products: [{ id: '1', name: 'Widget' }],
            });
        });
    });
    describe('include omitted', () => {
        let app;
        let apolloClient;
        beforeAll(async () => {
            ({ app, apolloClient } = await buildApp());
        });
        afterAll(async () => {
            await app.close();
        });
        it('should expose all types regardless of registerIn', async () => {
            (await queryTypeName(apolloClient, 'User')).toEqual({
                __type: { name: 'User' },
            });
            (await queryTypeName(apolloClient, 'Product')).toEqual({
                __type: { name: 'Product' },
            });
            (await queryTypeName(apolloClient, 'AppInfo')).toEqual({
                __type: { name: 'AppInfo' },
            });
        });
        it('should resolve queries from all modules', async () => {
            const response = await apolloClient.executeOperation({
                query: (0, graphql_tag_1.gql) `
          {
            users {
              id
              name
            }
            products {
              id
              name
            }
          }
        `,
            });
            (0, assertion_utils_1.expectSingleResult)(response).toEqual({
                users: [{ id: '1', name: 'Ada' }],
                products: [{ id: '1', name: 'Widget' }],
            });
        });
    });
});
