// @nestjs/apollo/dist/tests/tests/e2e/code-first-schema.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("@nestjs/graphql");
const graphql_constants_1 = require("@nestjs/graphql/graphql.constants");
const testing_1 = require("@nestjs/testing");
const graphql_2 = require("graphql");
const directions_resolver_1 = require("../code-first/directions/directions.resolver");
const sample_orphaned_enum_1 = require("../code-first/enums/sample-orphaned.enum");
const abstract_resolver_1 = require("../code-first/other/abstract.resolver");
const sample_orphaned_type_1 = require("../code-first/other/sample-orphaned.type");
const sample_scalar_1 = require("../code-first/other/sample-scalar");
const ingredients_resolver_1 = require("../code-first/recipes/ingredients.resolver");
const irecipes_resolver_1 = require("../code-first/recipes/irecipes.resolver");
const recipe_1 = require("../code-first/recipes/models/recipe");
const recipes_resolver_1 = require("../code-first/recipes/recipes.resolver");
const introspection_schema_utils_1 = require("../utils/introspection-schema.utils");
const printed_schema_snapshot_1 = require("../utils/printed-schema.snapshot");
describe('Code-first - schema factory', () => {
    let schemaFactory;
    beforeAll(async () => {
        const moduleRef = await testing_1.Test.createTestingModule({
            imports: [graphql_1.GraphQLSchemaBuilderModule],
        }).compile();
        schemaFactory = moduleRef.get(graphql_1.GraphQLSchemaFactory);
    });
    describe('generated schema', () => {
        let schema;
        let introspectionSchema;
        beforeAll(async () => {
            schema = await schemaFactory.create([
                ingredients_resolver_1.IngredientsResolver,
                recipes_resolver_1.RecipesResolver,
                directions_resolver_1.DirectionsResolver,
                abstract_resolver_1.AbstractResolver,
                irecipes_resolver_1.IRecipesResolver,
            ], [sample_scalar_1.SampleScalar], { orphanedTypes: [sample_orphaned_type_1.SampleOrphanedType, sample_orphaned_enum_1.SampleOrphanedEnum] });
            const introspectionResult = await (0, graphql_2.graphql)({
                schema,
                source: (0, graphql_2.getIntrospectionQuery)(),
            });
            if (!introspectionResult.data?.__schema) {
                throw new Error('Missing introspection schema data');
            }
            introspectionSchema = introspectionResult.data
                .__schema;
        });
        it('should be valid', async () => {
            expect(schema).toBeInstanceOf(graphql_2.GraphQLSchema);
        });
        it('should match schema snapshot', () => {
            expect(graphql_constants_1.GRAPHQL_SDL_FILE_HEADER + (0, graphql_2.printSchema)(schema)).toEqual(printed_schema_snapshot_1.printedSchemaSnapshot);
        });
        it('should add scalar to schema', () => {
            // makeExecutableSchema throws if a resolver is defined without being
            // present in the schema. We should include scalars even if they're not
            // used by any field (yet)
            const type = introspectionSchema.types.find(({ name }) => name === 'SampleScalar');
            expect(type).toEqual(expect.objectContaining({
                description: 'A sample scalar',
                kind: 'SCALAR',
                name: 'SampleScalar',
            }));
        });
        it('should define 5 queries', async () => {
            const type = (0, introspection_schema_utils_1.getQuery)(introspectionSchema);
            expect(type.fields.length).toEqual(5);
            expect(type.fields.map((item) => item.name)).toEqual(expect.arrayContaining([
                'recipes',
                'search',
                'categories',
                'move',
                'recipe',
            ]));
        });
        it('should define "addRecipe" and "removeRecipe", mutations', async () => {
            const type = (0, introspection_schema_utils_1.getMutation)(introspectionSchema);
            expect(type.fields.length).toEqual(2);
            expect(type.fields.map((item) => item.name)).toEqual(expect.arrayContaining(['addRecipe', 'removeRecipe']));
        });
        it('should define "recipeAdded" subscription', async () => {
            const type = (0, introspection_schema_utils_1.getSubscription)(introspectionSchema);
            expect(type.fields.length).toEqual(1);
            expect(type.fields.map((item) => item.name)).toEqual(expect.arrayContaining(['recipeAdded']));
        });
        it('should not define an abstract resolver', () => {
            const abstractQuery = (0, introspection_schema_utils_1.getQueryByName)(introspectionSchema, 'abstractRecipes');
            expect(abstractQuery).toBeUndefined();
        });
        it('should define "Direction" enum', () => {
            const type = introspectionSchema.types.find(({ name }) => name === 'Direction');
            expect(type).toEqual(expect.objectContaining({
                kind: graphql_2.TypeKind.ENUM,
                name: 'Direction',
                description: 'The basic directions',
                enumValues: [
                    {
                        deprecationReason: null,
                        description: 'The primary direction',
                        isDeprecated: false,
                        name: 'Up',
                    },
                    {
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'Down',
                    },
                    {
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'Left',
                    },
                    {
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'Right',
                    },
                    {
                        deprecationReason: 'Replaced with Left or Right',
                        description: null,
                        isDeprecated: true,
                        name: 'Sideways',
                    },
                ],
            }));
        });
        it('should define "SearchResultUnion" union', () => {
            const type = introspectionSchema.types.find(({ name }) => name === 'SearchResultUnion');
            expect(type).toEqual(expect.objectContaining({
                description: 'Search result description',
                kind: graphql_2.TypeKind.UNION,
                name: 'SearchResultUnion',
                possibleTypes: [
                    {
                        kind: graphql_2.TypeKind.OBJECT,
                        name: 'Ingredient',
                        ofType: null,
                    },
                    {
                        kind: graphql_2.TypeKind.OBJECT,
                        name: 'Recipe',
                        ofType: null,
                    },
                ],
            }));
        });
        it('should define "Ingredient" type', () => {
            const type = introspectionSchema.types.find(({ name }) => name === 'Ingredient');
            expect(type).toEqual(expect.objectContaining({
                name: 'Ingredient',
                kind: graphql_2.TypeKind.OBJECT,
                fields: expect.arrayContaining([
                    {
                        args: [],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'id',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: { kind: graphql_2.TypeKind.SCALAR, name: 'ID', ofType: null },
                        },
                    },
                    {
                        args: [],
                        deprecationReason: 'is deprecated',
                        description: 'ingredient name',
                        isDeprecated: true,
                        name: 'name',
                        type: {
                            kind: graphql_2.TypeKind.SCALAR,
                            name: 'String',
                            ofType: null,
                        },
                    },
                    {
                        args: [],
                        deprecationReason: 'is deprecated',
                        description: 'ingredient base name',
                        isDeprecated: true,
                        name: 'baseName',
                        type: {
                            kind: graphql_2.TypeKind.SCALAR,
                            name: 'String',
                            ofType: null,
                        },
                    },
                ]),
            }));
        });
        it('should define "Recipe" type', () => {
            const type = introspectionSchema.types.find(({ name }) => name === 'Recipe');
            expect(type).toEqual(expect.objectContaining({
                name: recipe_1.Recipe.name,
                kind: graphql_2.TypeKind.OBJECT,
                description: 'recipe object type',
                interfaces: [
                    { kind: graphql_2.TypeKind.INTERFACE, name: 'IRecipe', ofType: null },
                ],
                fields: [
                    {
                        args: [],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'id',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: { kind: graphql_2.TypeKind.SCALAR, name: 'ID', ofType: null },
                        },
                    },
                    {
                        args: [],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'title',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: { kind: graphql_2.TypeKind.SCALAR, name: 'String', ofType: null },
                        },
                    },
                    {
                        args: [
                            {
                                defaultValue: null,
                                description: null,
                                name: 'arg',
                                type: { kind: 'SCALAR', name: 'Float', ofType: null },
                            },
                        ],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'interfaceResolver',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: {
                                kind: graphql_2.TypeKind.SCALAR,
                                name: 'Boolean',
                                ofType: null,
                            },
                        },
                    },
                    {
                        args: [],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'description',
                        type: { kind: graphql_2.TypeKind.SCALAR, name: 'String', ofType: null },
                    },
                    {
                        args: [],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'creationDate',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: {
                                kind: graphql_2.TypeKind.SCALAR,
                                name: 'DateTime',
                                ofType: null,
                            },
                        },
                    },
                    {
                        args: [],
                        deprecationReason: null,
                        description: "The recipe's average rating",
                        isDeprecated: false,
                        name: 'averageRating',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: { kind: graphql_2.TypeKind.SCALAR, name: 'Float', ofType: null },
                        },
                    },
                    {
                        args: [],
                        deprecationReason: null,
                        description: 'last rate description',
                        isDeprecated: false,
                        name: 'lastRate',
                        type: { kind: graphql_2.TypeKind.SCALAR, name: 'Float', ofType: null },
                    },
                    {
                        args: [],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'tags',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: {
                                kind: graphql_2.TypeKind.LIST,
                                name: null,
                                ofType: {
                                    kind: graphql_2.TypeKind.NON_NULL,
                                    name: null,
                                    ofType: {
                                        kind: graphql_2.TypeKind.SCALAR,
                                        name: 'String',
                                        ofType: null,
                                    },
                                },
                            },
                        },
                    },
                    {
                        args: [],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'ingredients',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: {
                                kind: 'LIST',
                                name: null,
                                ofType: {
                                    kind: graphql_2.TypeKind.NON_NULL,
                                    name: null,
                                    ofType: {
                                        kind: 'OBJECT',
                                        name: 'Ingredient',
                                        ofType: null,
                                    },
                                },
                            },
                        },
                    },
                    {
                        args: [
                            {
                                defaultValue: null,
                                description: null,
                                name: 'type',
                                type: { kind: graphql_2.TypeKind.SCALAR, name: 'String', ofType: null },
                            },
                            {
                                defaultValue: null,
                                description: null,
                                name: 'status',
                                type: { kind: graphql_2.TypeKind.SCALAR, name: 'String', ofType: null },
                            },
                        ],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'count',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: { kind: graphql_2.TypeKind.SCALAR, name: 'Float', ofType: null },
                        },
                    },
                    {
                        args: [],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'rating',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: { kind: graphql_2.TypeKind.SCALAR, name: 'Float', ofType: null },
                        },
                    },
                ],
            }));
        });
        it('should define "SampleOrphanedType" orphaned type', () => {
            const type = introspectionSchema.types.find(({ name }) => name === 'SampleOrphanedType');
            expect(type).toEqual(expect.objectContaining({
                description: 'orphaned type',
                enumValues: null,
                fields: [
                    {
                        args: [],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'id',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: { kind: graphql_2.TypeKind.SCALAR, name: 'ID', ofType: null },
                        },
                    },
                    {
                        args: [],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'title',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: { kind: graphql_2.TypeKind.SCALAR, name: 'String', ofType: null },
                        },
                    },
                    {
                        args: [],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'description',
                        type: { kind: graphql_2.TypeKind.SCALAR, name: 'String', ofType: null },
                    },
                    {
                        args: [],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'creationDate',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: {
                                kind: graphql_2.TypeKind.SCALAR,
                                name: 'DateTime',
                                ofType: null,
                            },
                        },
                    },
                    {
                        args: [],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'averageRating',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: { kind: graphql_2.TypeKind.SCALAR, name: 'Float', ofType: null },
                        },
                    },
                ],
                inputFields: null,
                interfaces: [],
                kind: 'OBJECT',
                name: 'SampleOrphanedType',
                possibleTypes: null,
            }));
        });
        it('should define "SampleOrphanedEnum" orphaned type', () => {
            const type = introspectionSchema.types.find(({ name }) => name === 'SampleOrphanedEnum');
            expect(type).toEqual(expect.objectContaining({
                kind: graphql_2.TypeKind.ENUM,
                name: 'SampleOrphanedEnum',
                description: 'orphaned enum',
                enumValues: [
                    {
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'Red',
                    },
                    {
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'Blue',
                    },
                    {
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'Black',
                    },
                    {
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'White',
                    },
                ],
            }));
        });
        it('should define "IRecipe" interface', () => {
            const type = introspectionSchema.types.find(({ name }) => name === 'IRecipe');
            expect(type).toEqual(expect.objectContaining({
                name: 'IRecipe',
                kind: graphql_2.TypeKind.INTERFACE,
                description: 'example interface',
                possibleTypes: [
                    { kind: graphql_2.TypeKind.OBJECT, name: 'Recipe', ofType: null },
                ],
                fields: [
                    {
                        args: [],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'id',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: { kind: graphql_2.TypeKind.SCALAR, name: 'ID', ofType: null },
                        },
                    },
                    {
                        args: [],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'title',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: { kind: graphql_2.TypeKind.SCALAR, name: 'String', ofType: null },
                        },
                    },
                    {
                        args: [
                            {
                                defaultValue: null,
                                description: null,
                                name: 'arg',
                                type: { kind: 'SCALAR', name: 'Float', ofType: null },
                            },
                        ],
                        deprecationReason: null,
                        description: null,
                        isDeprecated: false,
                        name: 'interfaceResolver',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: {
                                kind: graphql_2.TypeKind.SCALAR,
                                name: 'Boolean',
                                ofType: null,
                            },
                        },
                    },
                ],
            }));
        });
        it('should define "NewRecipeInput" input type', () => {
            const type = introspectionSchema.types.find(({ name }) => name === 'NewRecipeInput');
            expect(type).toEqual(expect.objectContaining({
                name: 'NewRecipeInput',
                kind: graphql_2.TypeKind.INPUT_OBJECT,
                description: 'new recipe input',
                inputFields: [
                    {
                        defaultValue: null,
                        description: 'recipe title',
                        name: 'title',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: { kind: graphql_2.TypeKind.SCALAR, name: 'String', ofType: null },
                        },
                    },
                    {
                        defaultValue: null,
                        description: null,
                        name: 'description',
                        type: { kind: graphql_2.TypeKind.SCALAR, name: 'String', ofType: null },
                    },
                    {
                        defaultValue: '"published"',
                        description: null,
                        name: 'status',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: {
                                kind: graphql_2.TypeKind.SCALAR,
                                name: 'String',
                                ofType: null,
                            },
                        },
                    },
                    {
                        defaultValue: null,
                        description: null,
                        name: 'ingredients',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: {
                                kind: graphql_2.TypeKind.LIST,
                                name: null,
                                ofType: {
                                    kind: graphql_2.TypeKind.NON_NULL,
                                    name: null,
                                    ofType: {
                                        kind: graphql_2.TypeKind.SCALAR,
                                        name: 'String',
                                        ofType: null,
                                    },
                                },
                            },
                        },
                    },
                ],
            }));
        });
        describe('Query: "recipe"', () => {
            let recipeQuery;
            beforeAll(() => {
                recipeQuery = (0, introspection_schema_utils_1.getQueryByName)(introspectionSchema, 'recipe');
            });
            it('should define description', () => {
                expect(recipeQuery.description).toEqual('get recipe by id');
            });
            it('should set as not deprecated', () => {
                expect(recipeQuery.isDeprecated).toBeFalsy();
            });
            it('should return "IRecipe" interface', () => {
                expect(recipeQuery.type).toEqual({
                    kind: graphql_2.TypeKind.NON_NULL,
                    name: null,
                    ofType: { kind: graphql_2.TypeKind.INTERFACE, name: 'IRecipe', ofType: null },
                });
            });
            it('should take "id" as an input argument', () => {
                expect(recipeQuery.args.length).toEqual(1);
                expect(recipeQuery.args).toEqual(expect.arrayContaining([
                    {
                        defaultValue: '"1"',
                        description: 'recipe id',
                        name: 'id',
                        type: { kind: graphql_2.TypeKind.SCALAR, name: 'String', ofType: null },
                    },
                ]));
            });
        });
        describe('Query: "search"', () => {
            let searchQuery;
            beforeAll(() => {
                searchQuery = (0, introspection_schema_utils_1.getQueryByName)(introspectionSchema, 'search');
            });
            it('should not set description', () => {
                expect(searchQuery.description).toEqual(null);
            });
            it('should set as deprecated', () => {
                expect(searchQuery.isDeprecated).toBeTruthy();
                expect(searchQuery.deprecationReason).toEqual('test');
            });
            it('should return "SearchResultUnion" union', () => {
                expect(searchQuery.type).toEqual({
                    kind: graphql_2.TypeKind.NON_NULL,
                    name: null,
                    ofType: {
                        kind: graphql_2.TypeKind.LIST,
                        name: null,
                        ofType: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: {
                                kind: graphql_2.TypeKind.UNION,
                                name: 'SearchResultUnion',
                                ofType: null,
                            },
                        },
                    },
                });
            });
            it('should not take any input arguments', () => {
                expect(searchQuery.args.length).toEqual(0);
            });
        });
        describe('Query: "recipes"', () => {
            let recipesQuery;
            beforeAll(() => {
                recipesQuery = (0, introspection_schema_utils_1.getQueryByName)(introspectionSchema, 'recipes');
            });
            it('should not set description', () => {
                expect(recipesQuery.description).toEqual(null);
            });
            it('should not set as deprecated', () => {
                expect(recipesQuery.isDeprecated).toBeFalsy();
            });
            it('should return non nullable "Recipe[]" list', () => {
                expect(recipesQuery.type).toEqual({
                    kind: graphql_2.TypeKind.NON_NULL,
                    name: null,
                    ofType: {
                        kind: graphql_2.TypeKind.LIST,
                        name: null,
                        ofType: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: { kind: graphql_2.TypeKind.OBJECT, name: 'Recipe', ofType: null },
                        },
                    },
                });
            });
            it('should take 2 input arguments', () => {
                expect(recipesQuery.args.length).toEqual(2);
                expect(recipesQuery.args).toEqual(expect.arrayContaining([
                    {
                        defaultValue: '0',
                        description: 'number of items to skip',
                        name: 'skip',
                        type: { kind: graphql_2.TypeKind.SCALAR, name: 'Int', ofType: null },
                    },
                    {
                        defaultValue: '25',
                        description: null,
                        name: 'take',
                        type: { kind: graphql_2.TypeKind.SCALAR, name: 'Int', ofType: null },
                    },
                ]));
            });
        });
        describe('Mutation: "addRecipe"', () => {
            let addRecipeMutation;
            beforeAll(() => {
                addRecipeMutation = (0, introspection_schema_utils_1.getMutationByName)(introspectionSchema, 'addRecipe');
            });
            it('should not set description', () => {
                expect(addRecipeMutation.description).toEqual(null);
            });
            it('should not set as deprecated', () => {
                expect(addRecipeMutation.isDeprecated).toBeFalsy();
            });
            it('should return non nullable "Recipe" object', () => {
                expect(addRecipeMutation.type).toEqual({
                    kind: graphql_2.TypeKind.NON_NULL,
                    name: null,
                    ofType: { kind: graphql_2.TypeKind.OBJECT, name: 'Recipe', ofType: null },
                });
            });
            it('should take "NewRecipeInput" input type object as argument', () => {
                expect(addRecipeMutation.args.length).toEqual(1);
                expect(addRecipeMutation.args).toEqual([
                    {
                        defaultValue: null,
                        description: null,
                        name: 'newRecipeData',
                        type: {
                            kind: graphql_2.TypeKind.NON_NULL,
                            name: null,
                            ofType: {
                                kind: graphql_2.TypeKind.INPUT_OBJECT,
                                name: 'NewRecipeInput',
                                ofType: null,
                            },
                        },
                    },
                ]);
            });
        });
        describe('Subscription: "recipeAdded"', () => {
            let recipeAddedSub;
            beforeAll(() => {
                recipeAddedSub = (0, introspection_schema_utils_1.getSubscriptionByName)(introspectionSchema, 'recipeAdded');
            });
            it('should set description', () => {
                expect(recipeAddedSub.description).toEqual('subscription description');
            });
            it('should not set as deprecated', () => {
                expect(recipeAddedSub.isDeprecated).toBeFalsy();
            });
            it('should return non nullable "Recipe" object', () => {
                expect(recipeAddedSub.type).toEqual({
                    kind: graphql_2.TypeKind.NON_NULL,
                    name: null,
                    ofType: { kind: graphql_2.TypeKind.OBJECT, name: 'Recipe', ofType: null },
                });
            });
            it('should not take any arguments', () => {
                expect(recipeAddedSub.args.length).toEqual(0);
            });
        });
    });
});
