// @nestjs/apollo/dist/tests/tests/e2e/custom-context.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const platform_express_1 = require("@nestjs/platform-express");
const testing_1 = require("@nestjs/testing");
const supertest_1 = tslib_1.__importDefault(require("supertest"));
const custom_context_module_1 = require("../graphql/custom-context/custom-context.module");
const platform_fastify_1 = require("@nestjs/platform-fastify");
describe('GraphQL (custom context)', () => {
    let app;
    describe.each([
        ['Express', new platform_express_1.ExpressAdapter()],
        ['Fastify', new platform_fastify_1.FastifyAdapter()],
    ])('Custom context with %s', (_, adapter) => {
        beforeEach(async () => {
            const module = await testing_1.Test.createTestingModule({
                imports: [custom_context_module_1.CustomContextModule],
            }).compile();
            app = module.createNestApplication(adapter);
            await app.init();
            const instance = app.getHttpAdapter().getInstance();
            if ('ready' in instance && typeof instance.ready === 'function') {
                await instance.ready();
            }
        });
        it('should return query result', () => {
            return (0, supertest_1.default)(app.getHttpServer())
                .post('/graphql')
                .send({
                operationName: null,
                variables: {},
                query: `
          {
            fooFromContext
          }
          `,
            })
                .expect(200, {
                data: {
                    fooFromContext: 'bar',
                },
            });
        });
        afterEach(async () => {
            await app.close();
        });
    });
});
