// @nestjs/apollo/dist/tests/tests/e2e/enum-union-directives.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const disabled_1 = require("@apollo/server/plugin/disabled");
const common_1 = require("@nestjs/common");
const graphql_1 = require("@nestjs/graphql");
const testing_1 = require("@nestjs/testing");
const assert_1 = tslib_1.__importDefault(require("assert"));
const graphql_tag_1 = require("graphql-tag");
const lib_1 = require("../../lib");
var Visibility;
(function (Visibility) {
    Visibility["PUBLIC"] = "PUBLIC";
    Visibility["PRIVATE"] = "PRIVATE";
})(Visibility || (Visibility = {}));
(0, graphql_1.registerEnumType)(Visibility, {
    name: 'Visibility',
    description: 'Visibility of a resource',
    directives: ['@tag(name: "visibility")'],
    valuesMap: {
        PUBLIC: { directives: ['@tag(name: "public")'] },
    },
});
let Article = class Article {
};
tslib_1.__decorate([
    (0, graphql_1.Field)(() => graphql_1.ID),
    tslib_1.__metadata("design:type", String)
], Article.prototype, "id", void 0);
Article = tslib_1.__decorate([
    (0, graphql_1.ObjectType)(),
    (0, graphql_1.Directive)('@key(fields: "id")')
], Article);
let Video = class Video {
};
tslib_1.__decorate([
    (0, graphql_1.Field)(() => graphql_1.ID),
    tslib_1.__metadata("design:type", String)
], Video.prototype, "id", void 0);
Video = tslib_1.__decorate([
    (0, graphql_1.ObjectType)(),
    (0, graphql_1.Directive)('@key(fields: "id")')
], Video);
const Media = (0, graphql_1.createUnionType)({
    name: 'Media',
    types: () => [Article, Video],
    directives: ['@tag(name: "media")'],
});
let DirectivesResolver = class DirectivesResolver {
    visibility() {
        return Visibility.PUBLIC;
    }
    latest() {
        return { id: '1' };
    }
};
tslib_1.__decorate([
    (0, graphql_1.Query)(() => Visibility),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", String)
], DirectivesResolver.prototype, "visibility", null);
tslib_1.__decorate([
    (0, graphql_1.Query)(() => Media),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", Object)
], DirectivesResolver.prototype, "latest", null);
DirectivesResolver = tslib_1.__decorate([
    (0, graphql_1.Resolver)()
], DirectivesResolver);
let DirectivesAppModule = class DirectivesAppModule {
};
DirectivesAppModule = tslib_1.__decorate([
    (0, common_1.Module)({
        imports: [
            graphql_1.GraphQLModule.forRoot({
                driver: lib_1.ApolloFederationDriver,
                autoSchemaFile: true,
                buildSchemaOptions: {
                    orphanedTypes: [Article, Video],
                },
                plugins: [(0, disabled_1.ApolloServerPluginInlineTraceDisabled)()],
            }),
        ],
        providers: [DirectivesResolver],
    })
], DirectivesAppModule);
describe('Code-first - directives on enums and unions (e2e)', () => {
    let app;
    let apolloClient;
    beforeAll(async () => {
        const moduleRef = await testing_1.Test.createTestingModule({
            imports: [DirectivesAppModule],
        }).compile();
        app = moduleRef.createNestApplication();
        await app.init();
        const graphqlModule = app.get(graphql_1.GraphQLModule);
        apolloClient = graphqlModule.graphQlAdapter?.instance;
    });
    afterAll(async () => {
        await app.close();
    });
    it('should emit enum, enum-value and union directives in the subgraph SDL', async () => {
        const response = await apolloClient.executeOperation({
            query: (0, graphql_tag_1.gql) `
        {
          _service {
            sdl
          }
        }
      `,
        });
        (0, assert_1.default)(response.body.kind === 'single');
        expect(response.body.singleResult.errors).toBeUndefined();
        const sdl = response.body.singleResult.data._service.sdl;
        expect(sdl).toMatch(/enum Visibility[^{]*@tag\(name: "visibility"\)/);
        expect(sdl).toMatch(/PUBLIC @tag\(name: "public"\)/);
        expect(sdl).not.toMatch(/PRIVATE[^\n]*@tag/);
        expect(sdl).toMatch(/union Media[^=]*@tag\(name: "media"\)/);
    });
});
