// @nestjs/apollo/dist/tests/tests/e2e/federation-timestamp-scalar.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const disabled_1 = require("@apollo/server/plugin/disabled");
const assert_1 = tslib_1.__importDefault(require("assert"));
const common_1 = require("@nestjs/common");
const graphql_1 = require("@nestjs/graphql");
const testing_1 = require("@nestjs/testing");
const graphql_tag_1 = require("graphql-tag");
const lib_1 = require("../../lib");
let Reading = class Reading {
};
tslib_1.__decorate([
    (0, graphql_1.Field)(() => graphql_1.ID),
    tslib_1.__metadata("design:type", String)
], Reading.prototype, "id", void 0);
tslib_1.__decorate([
    (0, graphql_1.Field)(),
    tslib_1.__metadata("design:type", Date)
], Reading.prototype, "capturedAt", void 0);
Reading = tslib_1.__decorate([
    (0, graphql_1.ObjectType)()
], Reading);
let ReadingResolver = class ReadingResolver {
    reading() {
        const r = new Reading();
        r.id = '1';
        r.capturedAt = new Date(0);
        return r;
    }
};
tslib_1.__decorate([
    (0, graphql_1.Query)(() => Reading),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", Reading)
], ReadingResolver.prototype, "reading", null);
ReadingResolver = tslib_1.__decorate([
    (0, graphql_1.Resolver)(() => Reading)
], ReadingResolver);
let TimestampFederationModule = class TimestampFederationModule {
};
TimestampFederationModule = tslib_1.__decorate([
    (0, common_1.Module)({
        imports: [
            graphql_1.GraphQLModule.forRoot({
                driver: lib_1.ApolloFederationDriver,
                autoSchemaFile: true,
                buildSchemaOptions: { dateScalarMode: 'timestamp' },
                plugins: [(0, disabled_1.ApolloServerPluginInlineTraceDisabled)()],
            }),
        ],
        providers: [ReadingResolver],
    })
], TimestampFederationModule);
describe('Federation - Timestamp scalar (dateScalarMode: "timestamp")', () => {
    let app;
    let apolloClient;
    beforeAll(async () => {
        const moduleRef = await testing_1.Test.createTestingModule({
            imports: [TimestampFederationModule],
        }).compile();
        app = moduleRef.createNestApplication();
        await app.init();
        const graphqlModule = app.get(graphql_1.GraphQLModule);
        apolloClient = graphqlModule.graphQlAdapter?.instance;
    });
    afterAll(async () => {
        await app.close();
    });
    it('should serialize a Date field through the Timestamp scalar in the federated schema', async () => {
        const response = await apolloClient.executeOperation({
            query: (0, graphql_tag_1.gql) `
        {
          reading {
            id
            capturedAt
          }
        }
      `,
        });
        (0, assert_1.default)(response.body.kind === 'single');
        expect(response.body.singleResult.errors).toBeUndefined();
        const data = response.body.singleResult.data;
        expect(data.reading.id).toBe('1');
        expect(data.reading.capturedAt).toBe(0);
    });
});
