// @nestjs/apollo/dist/tests/tests/e2e/global-prefix.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@nestjs/core");
const platform_express_1 = require("@nestjs/platform-express");
const testing_1 = require("@nestjs/testing");
const supertest_1 = tslib_1.__importDefault(require("supertest"));
const global_prefix_async_options_class_module_1 = require("../graphql/global-prefix-async-options-class.module");
const global_prefix_async_options_module_1 = require("../graphql/global-prefix-async-options.module");
const global_prefix_module_1 = require("../graphql/global-prefix.module");
describe('GraphQL (global prefix)', () => {
    let app;
    describe('Global prefix with starting slash', () => {
        beforeEach(async () => {
            const module = await testing_1.Test.createTestingModule({
                imports: [global_prefix_module_1.GlobalPrefixModule],
            }).compile();
            app = module.createNestApplication(new platform_express_1.ExpressAdapter(), {
                abortOnError: false,
            });
            app.setGlobalPrefix('/api/v1');
            await app.init();
        });
        it('should return query result', () => {
            return (0, supertest_1.default)(app.getHttpServer())
                .post('/api/v1/graphql')
                .send({
                operationName: null,
                variables: {},
                query: `
          {
            getCats {
              id,
              color,
              weight
            }
          }`,
            })
                .expect(200, {
                data: {
                    getCats: [
                        {
                            id: 1,
                            color: 'black',
                            weight: 5,
                        },
                    ],
                },
            });
        });
        afterEach(async () => {
            await app.close();
        });
    });
    describe('Global prefix without starting slash', () => {
        beforeEach(async () => {
            const module = await testing_1.Test.createTestingModule({
                imports: [global_prefix_module_1.GlobalPrefixModule],
            }).compile();
            app = module.createNestApplication(new platform_express_1.ExpressAdapter(), {
                abortOnError: false,
            });
            app.setGlobalPrefix('api/v1');
            await app.init();
        });
        it('should return query result', () => {
            return (0, supertest_1.default)(app.getHttpServer())
                .post('/api/v1/graphql')
                .send({
                operationName: null,
                variables: {},
                query: `
          {
            getCats {
              id,
              color,
              weight
            }
          }`,
            })
                .expect(200, {
                data: {
                    getCats: [
                        {
                            id: 1,
                            color: 'black',
                            weight: 5,
                        },
                    ],
                },
            });
        });
        afterEach(async () => {
            await app.close();
        });
    });
    describe('Global prefix with ending slash', () => {
        beforeEach(async () => {
            const module = await testing_1.Test.createTestingModule({
                imports: [global_prefix_module_1.GlobalPrefixModule],
            }).compile();
            app = module.createNestApplication(new platform_express_1.ExpressAdapter(), {
                abortOnError: false,
            });
            app.setGlobalPrefix('/api/v1/');
            await app.init();
        });
        it('should return query result', () => {
            return (0, supertest_1.default)(app.getHttpServer())
                .post('/api/v1/graphql')
                .send({
                operationName: null,
                variables: {},
                query: `
          {
            getCats {
              id,
              color,
              weight
            }
          }`,
            })
                .expect(200, {
                data: {
                    getCats: [
                        {
                            id: 1,
                            color: 'black',
                            weight: 5,
                        },
                    ],
                },
            });
        });
        afterEach(async () => {
            await app.close();
        });
    });
    describe('Global prefix (async configuration)', () => {
        beforeEach(async () => {
            app = await core_1.NestFactory.create(global_prefix_async_options_module_1.GlobalPrefixAsyncOptionsModule, {
                logger: false,
                abortOnError: false,
            });
            app.setGlobalPrefix('/api/v1/');
            await app.init();
        });
        it('should return query result', () => {
            return (0, supertest_1.default)(app.getHttpServer())
                .post('/api/v1/graphql')
                .send({
                operationName: null,
                variables: {},
                query: `
          {
            getCats {
              id,
              color,
              weight
            }
          }`,
            })
                .expect(200, {
                data: {
                    getCats: [
                        {
                            id: 1,
                            color: 'black',
                            weight: 5,
                        },
                    ],
                },
            });
        });
        afterEach(async () => {
            await app.close();
        });
    });
    describe('Global prefix (async class)', () => {
        beforeEach(async () => {
            app = await core_1.NestFactory.create(global_prefix_async_options_class_module_1.GlobalPrefixAsyncOptionsClassModule, {
                logger: false,
                abortOnError: false,
            });
            app.setGlobalPrefix('/api/v1/');
            await app.init();
        });
        it('should return query result', () => {
            return (0, supertest_1.default)(app.getHttpServer())
                .post('/api/v1/graphql')
                .send({
                operationName: null,
                variables: {},
                query: `
          {
            getCats {
              id,
              color,
              weight
            }
          }`,
            })
                .expect(200, {
                data: {
                    getCats: [
                        {
                            id: 1,
                            color: 'black',
                            weight: 5,
                        },
                    ],
                },
            });
        });
        afterEach(async () => {
            await app.close();
        });
    });
});
