// @nestjs/apollo/dist/tests/tests/e2e/graphql-gateway-transform-schema.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const graphql_1 = require("@nestjs/graphql");
const testing_1 = require("@nestjs/testing");
const supertest_1 = tslib_1.__importDefault(require("supertest"));
const drivers_1 = require("../../lib/drivers");
const supergraph_sdl_1 = require("../graphql-federation/gateway/supergraph-sdl");
const federation_posts_module_1 = require("../graphql-federation/posts-service/federation-posts.module");
const federation_users_module_1 = require("../graphql-federation/users-service/federation-users.module");
describe('GraphQL Gateway transformSchema', () => {
    let postsApp;
    let usersApp;
    let gatewayApp;
    afterEach(async () => {
        await postsApp?.close();
        await usersApp?.close();
        await gatewayApp?.close();
    });
    it('invokes the gateway-level transformSchema callback with the composed supergraph schema', async () => {
        const usersModule = await testing_1.Test.createTestingModule({
            imports: [federation_users_module_1.AppModule],
        }).compile();
        usersApp = usersModule.createNestApplication();
        await usersApp.listen(0);
        const usersPort = usersApp.getHttpServer().address().port;
        const postsModule = await testing_1.Test.createTestingModule({
            imports: [federation_posts_module_1.AppModule],
        }).compile();
        postsApp = postsModule.createNestApplication();
        await postsApp.listen(0);
        const postsPort = postsApp.getHttpServer().address().port;
        let receivedSchema;
        let invocationCount = 0;
        const gatewayModule = await testing_1.Test.createTestingModule({
            imports: [
                graphql_1.GraphQLModule.forRoot({
                    driver: drivers_1.ApolloGatewayDriver,
                    gateway: {
                        includeStacktraceInErrorResponses: false,
                        supergraphSdl: (0, supergraph_sdl_1.getSupergraphSdl)(usersPort, postsPort),
                    },
                    transformSchema: (schema) => {
                        invocationCount += 1;
                        receivedSchema = schema;
                        return schema;
                    },
                }),
            ],
        }).compile();
        gatewayApp = gatewayModule.createNestApplication();
        await gatewayApp.init();
        expect(invocationCount).toBe(1);
        expect(receivedSchema).toBeDefined();
        // The composed supergraph must expose the federated `Post` type.
        expect(receivedSchema.getType('Post')).toBeDefined();
        expect(receivedSchema.getType('User')).toBeDefined();
    }, 15000);
    it('exposes the transformed schema to Apollo Server during query execution', async () => {
        const usersModule = await testing_1.Test.createTestingModule({
            imports: [federation_users_module_1.AppModule],
        }).compile();
        usersApp = usersModule.createNestApplication();
        await usersApp.listen(0);
        const usersPort = usersApp.getHttpServer().address().port;
        const postsModule = await testing_1.Test.createTestingModule({
            imports: [federation_posts_module_1.AppModule],
        }).compile();
        postsApp = postsModule.createNestApplication();
        await postsApp.listen(0);
        const postsPort = postsApp.getHttpServer().address().port;
        const gatewayModule = await testing_1.Test.createTestingModule({
            imports: [
                graphql_1.GraphQLModule.forRoot({
                    driver: drivers_1.ApolloGatewayDriver,
                    gateway: {
                        includeStacktraceInErrorResponses: false,
                        supergraphSdl: (0, supergraph_sdl_1.getSupergraphSdl)(usersPort, postsPort),
                    },
                    // Identity transform: confirms the wrapper does not break the
                    // standard query path while still routing the schema through
                    // the user-supplied callback.
                    transformSchema: (schema) => schema,
                }),
            ],
        }).compile();
        gatewayApp = gatewayModule.createNestApplication();
        await gatewayApp.init();
        return (0, supertest_1.default)(gatewayApp.getHttpServer())
            .post('/graphql')
            .send({
            operationName: null,
            variables: {},
            query: `
        {
          getPosts {
            id
            title
          }
        }`,
        })
            .expect(200)
            .expect((res) => {
            expect(res.body.errors).toBeUndefined();
            expect(res.body.data.getPosts).toBeDefined();
        });
    }, 15000);
});
