// @nestjs/apollo/dist/tests/tests/e2e/graphql-gateway-async-class.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const graphql_1 = require("@nestjs/graphql");
const testing_1 = require("@nestjs/testing");
const supertest_1 = tslib_1.__importDefault(require("supertest"));
const drivers_1 = require("../../lib/drivers");
const supergraph_sdl_1 = require("../graphql-federation/gateway/supergraph-sdl");
const federation_posts_module_1 = require("../graphql-federation/posts-service/federation-posts.module");
const federation_users_module_1 = require("../graphql-federation/users-service/federation-users.module");
let usersPort;
let postsPort;
let GatewayConfigService = class GatewayConfigService {
    createGqlOptions() {
        return {
            gateway: {
                supergraphSdl: (0, supergraph_sdl_1.getSupergraphSdl)(usersPort, postsPort),
            },
        };
    }
};
GatewayConfigService = tslib_1.__decorate([
    (0, common_1.Injectable)()
], GatewayConfigService);
let GatewayConfigModule = class GatewayConfigModule {
};
GatewayConfigModule = tslib_1.__decorate([
    (0, common_1.Module)({
        providers: [GatewayConfigService],
        exports: [GatewayConfigService],
    })
], GatewayConfigModule);
describe('GraphQL Gateway async-class', () => {
    let postsApp;
    let usersApp;
    let gatewayApp;
    beforeEach(async () => {
        const usersModule = await testing_1.Test.createTestingModule({
            imports: [federation_users_module_1.AppModule],
        }).compile();
        usersApp = usersModule.createNestApplication();
        await usersApp.listen(0);
        usersPort = usersApp.getHttpServer().address().port;
        const postsModule = await testing_1.Test.createTestingModule({
            imports: [federation_posts_module_1.AppModule],
        }).compile();
        postsApp = postsModule.createNestApplication();
        await postsApp.listen(0);
        postsPort = postsApp.getHttpServer().address().port;
        const gatewayModule = await testing_1.Test.createTestingModule({
            imports: [
                graphql_1.GraphQLModule.forRootAsync({
                    driver: drivers_1.ApolloGatewayDriver,
                    useClass: GatewayConfigService,
                    imports: [GatewayConfigModule],
                }),
            ],
        }).compile();
        gatewayApp = gatewayModule.createNestApplication();
        await gatewayApp.init();
    });
    it(`should run lookup across boundaries`, () => {
        return (0, supertest_1.default)(gatewayApp.getHttpServer())
            .post('/graphql')
            .send({
            operationName: null,
            variables: {},
            query: `
        {
          getPosts {
            id,
            title,
            body,
            user {
              id,
              name,
            }
          }
        }`,
        })
            .expect(200, {
            data: {
                getPosts: [
                    {
                        id: '1',
                        title: 'HELLO WORLD',
                        body: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
                        user: {
                            id: '5',
                            name: 'GraphQL',
                        },
                    },
                ],
            },
        });
    });
    afterEach(async () => {
        await postsApp.close();
        await usersApp.close();
        await gatewayApp.close();
    });
});
