// @nestjs/apollo/dist/tests/tests/e2e/graphql-request-scoped.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const graphql_1 = require("@nestjs/graphql");
const testing_1 = require("@nestjs/testing");
const path_1 = require("path");
const supertest_1 = tslib_1.__importDefault(require("supertest"));
const drivers_1 = require("../../lib/drivers");
const cats_request_scoped_service_1 = require("../graphql/cats/cats-request-scoped.service");
const cats_module_1 = require("../graphql/cats/cats.module");
describe('GraphQL request scoped', () => {
    let app;
    beforeEach(async () => {
        const module = await testing_1.Test.createTestingModule({
            imports: [
                cats_module_1.CatsModule.enableRequestScope(),
                graphql_1.GraphQLModule.forRoot({
                    driver: drivers_1.ApolloDriver,
                    typePaths: [(0, path_1.join)(__dirname, '..', 'graphql', '**', '*.graphql')],
                }),
            ],
        }).compile();
        app = module.createNestApplication();
        await app.init();
        const performHttpCall = (end) => (0, supertest_1.default)(app.getHttpServer())
            .post('/graphql')
            .send({
            operationName: null,
            variables: {},
            query: '{\n  getCats {\n    id\n  }\n}\n',
        })
            .expect(200, {
            data: {
                getCats: [
                    {
                        id: 1,
                    },
                ],
            },
        })
            .end((err, res) => {
            if (err)
                return end(err);
            end();
        });
        await new Promise((resolve) => performHttpCall(resolve));
        await new Promise((resolve) => performHttpCall(resolve));
        await new Promise((resolve) => performHttpCall(resolve));
    });
    it(`should create resolver for each incoming request`, () => {
        expect(cats_request_scoped_service_1.CatsRequestScopedService.COUNTER).toEqual(3);
    });
    afterEach(async () => {
        await app.close();
    });
});
