// @nestjs/apollo/dist/tests/tests/e2e/graphql-async-existing.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@nestjs/core");
const supertest_1 = tslib_1.__importDefault(require("supertest"));
const async_options_existing_module_1 = require("../graphql/async-options-existing.module");
describe('GraphQL (async existing)', () => {
    let app;
    beforeEach(async () => {
        app = await core_1.NestFactory.create(async_options_existing_module_1.AsyncExistingApplicationModule, {
            logger: false,
        });
        await app.init();
    });
    it(`should return query result`, () => {
        return (0, supertest_1.default)(app.getHttpServer())
            .post('/graphql')
            .send({
            operationName: null,
            variables: {},
            query: '{\n  getCats {\n    id\n  }\n}\n',
        })
            .expect(200, {
            data: {
                getCats: [
                    {
                        id: 1,
                    },
                ],
            },
        });
    });
    afterEach(async () => {
        await app.close();
    });
});
