// @nestjs/apollo/dist/tests/tests/e2e/graphql-sort-schema.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("@nestjs/graphql");
const graphql_constants_1 = require("@nestjs/graphql/graphql.constants");
const testing_1 = require("@nestjs/testing");
const graphql_2 = require("graphql");
const sort_schema_module_1 = require("../graphql/sort-schema.module");
describe('GraphQL sort schema', () => {
    let app;
    let schema;
    beforeEach(async () => {
        const module = await testing_1.Test.createTestingModule({
            imports: [sort_schema_module_1.SortSchemaModule],
        }).compile();
        app = module.createNestApplication();
        await app.init();
        const graphQLSchemaHost = app.get(graphql_1.GraphQLSchemaHost);
        schema = graphQLSchemaHost.schema;
    });
    it('should match schema snapshot', () => {
        expect(graphql_constants_1.GRAPHQL_SDL_FILE_HEADER + (0, graphql_2.printSchema)(schema)).toEqual(expectedSchema);
    });
    afterEach(async () => {
        await app.close();
    });
});
const expectedSchema = `# ------------------------------------------------------
# THIS FILE WAS AUTOMATICALLY GENERATED (DO NOT MODIFY)
# ------------------------------------------------------

type Cat {
  age: Int
  color: String
  id: Int
  name: String
  weight: Int
}

type Mutation {
  createCat(name: String): Cat
}

type Query {
  cat(id: ID!): Cat
  getCats: [Cat]
}

type Subscription {
  catCreated: Cat
}`;
