// @nestjs/apollo/dist/tests/tests/e2e/graphql.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const testing_1 = require("@nestjs/testing");
const supertest_1 = tslib_1.__importDefault(require("supertest"));
const app_module_1 = require("../graphql/app.module");
const graphql_1 = require("@nestjs/graphql");
describe('GraphQL', () => {
    let app;
    beforeEach(async () => {
        const module = await testing_1.Test.createTestingModule({
            imports: [app_module_1.ApplicationModule],
        }).compile();
        app = module.createNestApplication();
        await app.init();
    });
    it(`should return query result`, () => {
        return (0, supertest_1.default)(app.getHttpServer())
            .post('/graphql')
            .send({
            operationName: null,
            variables: {},
            query: `
        {
          getCats {
            id,
            color,
            weight
          }
        }`,
        })
            .expect(200, {
            data: {
                getCats: [
                    {
                        id: 1,
                        color: 'black',
                        weight: 5,
                    },
                ],
            },
        });
    });
    it(`should return complete options`, () => {
        const graphql = app.get(graphql_1.GraphQLModule);
        expect(graphql.completeOptions).toHaveProperty('path', '/graphql');
        expect(graphql.completeOptions).toHaveProperty('schema');
    });
    afterEach(async () => {
        await app.close();
    });
});
