// @nestjs/apollo/dist/tests/tests/e2e/graphql-federation-fastify.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const testing_1 = require("@nestjs/testing");
const supertest_1 = tslib_1.__importDefault(require("supertest"));
const federation_posts_module_1 = require("../graphql-federation/posts-service/federation-posts.module");
const platform_fastify_1 = require("@nestjs/platform-fastify");
describe('GraphQL federation with fastify', () => {
    let app;
    beforeEach(async () => {
        const module = await testing_1.Test.createTestingModule({
            imports: [federation_posts_module_1.AppModule],
        }).compile();
        app = module.createNestApplication(new platform_fastify_1.FastifyAdapter());
        await app.init();
        await app.getHttpAdapter().getInstance().ready();
    });
    it(`should return query result`, () => {
        return (0, supertest_1.default)(app.getHttpServer())
            .post('/graphql')
            .send({
            operationName: null,
            variables: {},
            query: `
        {
          getPosts {
            id,
            title,
            body,
          }
        }`,
        })
            .expect(200, {
            data: {
                getPosts: [
                    {
                        id: '1',
                        title: 'HELLO WORLD',
                        body: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
                    },
                ],
            },
        });
    });
    afterEach(async () => {
        await app.close();
    });
});
