// @nestjs/apollo/dist/tests/tests/e2e/graphql-federation.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const testing_1 = require("@nestjs/testing");
const supertest_1 = tslib_1.__importDefault(require("supertest"));
const federation_posts_module_1 = require("../graphql-federation/posts-service/federation-posts.module");
const federation_users_module_1 = require("../graphql-federation/users-service/federation-users.module");
describe('GraphQL Federation', () => {
    let app;
    describe('UsersService', () => {
        beforeEach(async () => {
            const module = await testing_1.Test.createTestingModule({
                imports: [federation_users_module_1.AppModule],
            }).compile();
            app = module.createNestApplication();
            await app.init();
        });
        it(`should return query result`, () => {
            return (0, supertest_1.default)(app.getHttpServer())
                .post('/graphql')
                .send({
                operationName: null,
                variables: {},
                query: `
        {
          getUser(id: "5") {
            id,
            name,
          }
        }`,
            })
                .expect(200, {
                data: {
                    getUser: {
                        id: '5',
                        name: 'GraphQL',
                    },
                },
            });
        });
        it('should resolve references', () => {
            return (0, supertest_1.default)(app.getHttpServer())
                .post('/graphql')
                .send({
                variables: {
                    representations: [
                        {
                            __typename: 'User',
                            id: '5',
                        },
                    ],
                },
                query: `
            query ($representations: [_Any!]!) {
              _entities(representations: $representations) {
                __typename
                ... on User {
                  id
                  name
                }
              }
            }`,
            })
                .expect(200, {
                data: {
                    _entities: [
                        {
                            __typename: 'User',
                            id: '5',
                            name: 'GraphQL',
                        },
                    ],
                },
            });
        });
    });
    describe('PostsService', () => {
        beforeEach(async () => {
            const module = await testing_1.Test.createTestingModule({
                imports: [federation_posts_module_1.AppModule],
            }).compile();
            app = module.createNestApplication();
            await app.init();
        });
        it(`should return query result`, () => {
            return (0, supertest_1.default)(app.getHttpServer())
                .post('/graphql')
                .send({
                operationName: null,
                variables: {},
                query: `
        {
          getPosts {
            id,
            title,
            body,
          }
        }`,
            })
                .expect(200, {
                data: {
                    getPosts: [
                        {
                            id: '1',
                            title: 'HELLO WORLD',
                            body: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
                        },
                    ],
                },
            });
        });
        it('should return a stripped reference', () => {
            return (0, supertest_1.default)(app.getHttpServer())
                .post('/graphql')
                .send({
                operationName: null,
                variables: {},
                query: `
        {
          getPosts {
            id,
            title,
            body,
            user {
              id
            }
          }
        }`,
            })
                .expect(200, {
                data: {
                    getPosts: [
                        {
                            id: '1',
                            title: 'HELLO WORLD',
                            body: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
                            user: {
                                id: '5',
                            },
                        },
                    ],
                },
            });
        });
        it(`should handle scalars`, () => {
            return (0, supertest_1.default)(app.getHttpServer())
                .post('/graphql')
                .send({
                operationName: null,
                variables: {},
                query: `
        mutation {
          publishPost(id: "1", publishDate: 500) {
            id,
            title,
            body,
            publishDate
          }
        }`,
            })
                .expect(200, {
                data: {
                    publishPost: {
                        id: '1',
                        title: 'HELLO WORLD',
                        body: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
                        publishDate: 500,
                    },
                },
            });
        });
        it('should accept enum as query input', () => {
            return (0, supertest_1.default)(app.getHttpServer())
                .post('/graphql')
                .send({
                variables: {
                    postType: 'TEXT',
                },
                query: `
            query ($postType: PostType!) {
              getPosts(type: $postType) {
                id
                type
              }
            }`,
            })
                .expect(200, {
                data: {
                    getPosts: [
                        {
                            id: '1',
                            type: 'TEXT',
                        },
                    ],
                },
            });
        });
    });
    afterEach(async () => {
        await app.close();
    });
});
