// @nestjs/apollo/dist/tests/tests/e2e/graphql-federation-async-existing.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const testing_1 = require("@nestjs/testing");
const supertest_1 = tslib_1.__importDefault(require("supertest"));
const federation_users_async_existing_module_1 = require("../graphql-federation/users-service/federation-users.async-existing.module");
describe('GraphQL Federation Async', () => {
    let app;
    beforeEach(async () => {
        const module = await testing_1.Test.createTestingModule({
            imports: [federation_users_async_existing_module_1.AppModule],
        }).compile();
        app = module.createNestApplication();
        await app.init();
    });
    it(`should return query result`, () => {
        return (0, supertest_1.default)(app.getHttpServer())
            .post('/graphql')
            .send({
            operationName: null,
            variables: {},
            query: `
        {
          getUser(id: "5") {
            id,
            name,
          }
        }`,
        })
            .expect(200, {
            data: {
                getUser: {
                    id: '5',
                    name: 'GraphQL',
                },
            },
        });
    });
    afterEach(async () => {
        await app.close();
    });
});
