// @nestjs/apollo/dist/tests/tests/e2e/issue-2940-http-error-status.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const graphql_1 = require("@nestjs/graphql");
const testing_1 = require("@nestjs/testing");
const graphql_2 = require("graphql");
const supertest_1 = tslib_1.__importDefault(require("supertest"));
const lib_1 = require("../../lib");
let IssueResolver = class IssueResolver {
    throwHttpException() {
        throw new common_1.BadRequestException('foo');
    }
    throwGraphqlErrorWithHttpStatus() {
        throw new graphql_2.GraphQLError('boom', {
            extensions: {
                code: 'BAD_REQUEST',
                http: { status: 400 },
            },
        });
    }
};
tslib_1.__decorate([
    (0, graphql_1.Query)(() => String),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", String)
], IssueResolver.prototype, "throwHttpException", null);
tslib_1.__decorate([
    (0, graphql_1.Query)(() => String),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", String)
], IssueResolver.prototype, "throwGraphqlErrorWithHttpStatus", null);
IssueResolver = tslib_1.__decorate([
    (0, graphql_1.Resolver)()
], IssueResolver);
function buildModule(autoTransformHttpErrors) {
    let Issue2940Module = class Issue2940Module {
    };
    Issue2940Module = tslib_1.__decorate([
        (0, common_1.Module)({
            imports: [
                graphql_1.GraphQLModule.forRoot({
                    driver: lib_1.ApolloDriver,
                    autoSchemaFile: true,
                    includeStacktraceInErrorResponses: false,
                    autoTransformHttpErrors,
                }),
            ],
            providers: [IssueResolver],
        })
    ], Issue2940Module);
    return Issue2940Module;
}
async function bootstrap(autoTransformHttpErrors) {
    const moduleRef = await testing_1.Test.createTestingModule({
        imports: [buildModule(autoTransformHttpErrors)],
    }).compile();
    const app = moduleRef.createNestApplication();
    await app.init();
    return app;
}
describe('Issue #2940 - GraphQL error formatting with HTTP status', () => {
    describe('default (autoTransformHttpErrors enabled)', () => {
        let app;
        beforeEach(async () => {
            app = await bootstrap();
        });
        afterEach(async () => {
            await app.close();
        });
        it('should return the standard {data, errors} response shape (HTTP 200) when a NestJS HttpException is thrown from a resolver', async () => {
            const res = await (0, supertest_1.default)(app.getHttpServer())
                .post('/graphql')
                .send({ query: '{ throwHttpException }' });
            expect(res.status).toBe(200);
            expect(res.body).toHaveProperty('data', null);
            expect(res.body).toHaveProperty('errors');
            expect(res.body).not.toHaveProperty('error');
            expect(res.body.errors[0].extensions).toMatchObject({
                code: 'BAD_REQUEST',
                originalError: {
                    statusCode: 400,
                    message: 'foo',
                },
            });
            expect(res.body.errors[0].extensions).not.toHaveProperty('http');
        });
        it('should keep the response at HTTP 200 when a resolver throws a GraphQLError carrying extensions.http.status', async () => {
            const res = await (0, supertest_1.default)(app.getHttpServer())
                .post('/graphql')
                .send({ query: '{ throwGraphqlErrorWithHttpStatus }' });
            expect(res.status).toBe(200);
            expect(res.body).toHaveProperty('data', null);
            expect(res.body).toHaveProperty('errors');
            expect(res.body).not.toHaveProperty('error');
            expect(res.body.errors[0].extensions.code).toBe('BAD_REQUEST');
            expect(res.body.errors[0].extensions).not.toHaveProperty('http');
        });
        it('should still return a non-200 status for request-level errors (e.g. parse failures)', async () => {
            const res = await (0, supertest_1.default)(app.getHttpServer())
                .post('/graphql')
                .send({ query: '{ throwHttpException ' });
            expect(res.status).toBe(400);
            expect(res.body).not.toHaveProperty('data');
            expect(res.body).toHaveProperty('errors');
        });
    });
    describe('with autoTransformHttpErrors disabled', () => {
        let app;
        beforeEach(async () => {
            app = await bootstrap(false);
        });
        afterEach(async () => {
            await app.close();
        });
        it('should preserve Apollo Server`s extensions.http.status response status when the user opts out', async () => {
            const res = await (0, supertest_1.default)(app.getHttpServer())
                .post('/graphql')
                .send({ query: '{ throwGraphqlErrorWithHttpStatus }' });
            expect(res.status).toBe(400);
            expect(res.body).toHaveProperty('data', null);
            expect(res.body).toHaveProperty('errors');
        });
    });
});
