// @nestjs/apollo/dist/tests/tests/e2e/pipes.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const testing_1 = require("@nestjs/testing");
const supertest_1 = tslib_1.__importDefault(require("supertest"));
const app_module_1 = require("../code-first/app.module");
describe('GraphQL - Pipes', () => {
    let app;
    beforeEach(async () => {
        const module = await testing_1.Test.createTestingModule({
            imports: [app_module_1.ApplicationModule],
        }).compile();
        app = module.createNestApplication();
        app.useGlobalPipes(new common_1.ValidationPipe());
        await app.init();
    });
    it(`should throw an error`, () => {
        return (0, supertest_1.default)(app.getHttpServer())
            .post('/graphql')
            .send({
            operationName: null,
            variables: {},
            query: 'mutation {\n  addRecipe(newRecipeData: {title: "test", ingredients: []}) {\n    id\n  }\n}\n',
        })
            .expect(200, {
            data: null,
            errors: [
                {
                    message: 'Bad Request Exception',
                    locations: [{ line: 2, column: 3 }],
                    path: ['addRecipe'],
                    extensions: {
                        code: 'BAD_REQUEST',
                        originalError: {
                            statusCode: 400,
                            message: [
                                'description must be longer than or equal to 30 characters',
                            ],
                            error: 'Bad Request',
                        },
                    },
                },
            ],
        });
    });
    afterEach(async () => {
        await app.close();
    });
});
