// @nestjs/apollo/dist/tests/tests/e2e/request-scoped.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const testing_1 = require("@nestjs/testing");
const supertest_1 = tslib_1.__importDefault(require("supertest"));
const request_scoped_guard_1 = require("../graphql/hello/guards/request-scoped.guard");
const hello_module_1 = require("../graphql/hello/hello.module");
const hello_resolver_1 = require("../graphql/hello/hello.resolver");
const logging_interceptor_1 = require("../graphql/hello/interceptors/logging.interceptor");
const users_service_1 = require("../graphql/hello/users/users.service");
class Meta {
    constructor() {
        Meta.COUNTER++;
    }
}
Meta.COUNTER = 0;
describe('Request scope', () => {
    let server;
    let app;
    beforeAll(async () => {
        const module = await testing_1.Test.createTestingModule({
            imports: [
                hello_module_1.HelloModule.forRoot({
                    provide: 'META',
                    useClass: Meta,
                }),
            ],
        }).compile();
        app = module.createNestApplication();
        server = app.getHttpServer();
        await app.init();
    });
    describe('when one service is request scoped', () => {
        beforeAll(async () => {
            const performHttpCall = (end) => (0, supertest_1.default)(server)
                .post('/graphql')
                .send({
                operationName: null,
                variables: {},
                query: `
        {
          getCats {
            id,
            color,
            weight
          }
        }`,
            })
                .end((err, res) => {
                if (err)
                    return end(err);
                end();
            });
            await new Promise((resolve) => performHttpCall(resolve));
            await new Promise((resolve) => performHttpCall(resolve));
            await new Promise((resolve) => performHttpCall(resolve));
        });
        it(`should create resolver for each request`, async () => {
            expect(hello_resolver_1.HelloResolver.COUNTER).toEqual(3);
        });
        it(`should create service for each request`, async () => {
            expect(users_service_1.UsersService.COUNTER).toEqual(3);
        });
        it(`should share static provider across requests`, async () => {
            expect(Meta.COUNTER).toEqual(1);
        });
        it(`should create request scoped interceptor for each request`, async () => {
            expect(logging_interceptor_1.Interceptor.COUNTER).toEqual(3);
            expect(logging_interceptor_1.Interceptor.REQUEST_SCOPED_DATA).toEqual([1, 1, 1]);
        });
        it(`should create request scoped guard for each request`, async () => {
            expect(request_scoped_guard_1.Guard.COUNTER).toEqual(3);
            expect(request_scoped_guard_1.Guard.REQUEST_SCOPED_DATA).toEqual([1, 1, 1]);
        });
        it(`should inject REQUEST context into request-scoped services`, async () => {
            expect(users_service_1.UsersService.REQUEST_CONTEXTS).toHaveLength(3);
            users_service_1.UsersService.REQUEST_CONTEXTS.forEach((context) => {
                expect(context).toBeDefined();
                expect(context.req).toBeDefined();
            });
        });
    });
    afterAll(async () => {
        await app.close();
    });
});
