// @nestjs/apollo/dist/tests/tests/e2e/resolver-registration-methods.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const testing_1 = require("@nestjs/testing");
const supertest_1 = tslib_1.__importDefault(require("supertest"));
const app_module_1 = require("../code-first/app.module");
const cats_module_1 = require("../code-first/cats/cats.module");
describe('GraphQL - Resolver registration methods', () => {
    let app;
    describe('useClass', () => {
        beforeEach(async () => {
            const module = await testing_1.Test.createTestingModule({
                imports: [app_module_1.ApplicationModule, cats_module_1.CatsModule.register('useClass')],
            }).compile();
            app = module.createNestApplication();
            await app.init();
        });
        it('should return the cats result', async () => {
            return (0, supertest_1.default)(app.getHttpServer())
                .post('/graphql')
                .send({
                operationName: null,
                variables: {},
                query: 'query {\n  getAnimalName \n}\n',
            })
                .expect(200, {
                data: {
                    getAnimalName: 'cat',
                },
            });
        });
        afterEach(async () => {
            await app.close();
        });
    });
    describe('useValue', () => {
        beforeEach(async () => {
            const module = await testing_1.Test.createTestingModule({
                imports: [app_module_1.ApplicationModule, cats_module_1.CatsModule.register('useValue')],
            }).compile();
            app = module.createNestApplication();
            await app.init();
        });
        it('should return the cats result', async () => {
            return (0, supertest_1.default)(app.getHttpServer())
                .post('/graphql')
                .send({
                operationName: null,
                variables: {},
                query: 'query {\n  getAnimalName \n}\n',
            })
                .expect(200, {
                data: {
                    getAnimalName: 'cat',
                },
            });
        });
        afterEach(async () => {
            await app.close();
        });
    });
    describe('useFactory', () => {
        beforeEach(async () => {
            const module = await testing_1.Test.createTestingModule({
                imports: [app_module_1.ApplicationModule, cats_module_1.CatsModule.register('useFactory')],
            }).compile();
            app = module.createNestApplication();
            await app.init();
        });
        it('should return the cats result', async () => {
            return (0, supertest_1.default)(app.getHttpServer())
                .post('/graphql')
                .send({
                operationName: null,
                variables: {},
                query: 'query {\n  getAnimalName \n}\n',
            })
                .expect(200, {
                data: {
                    getAnimalName: 'cat',
                },
            });
        });
        afterEach(async () => {
            await app.close();
        });
    });
});
