// @nestjs/apollo/dist/tests/tests/utils/introspection-schema.utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getQuery = getQuery;
exports.getMutation = getMutation;
exports.getSubscription = getSubscription;
exports.getQueryByName = getQueryByName;
exports.getMutationByName = getMutationByName;
exports.getSubscriptionByName = getSubscriptionByName;
function getQuery(introspectionSchema) {
    return introspectionSchema.types.find((item) => item.name === introspectionSchema.queryType.name);
}
function getMutation(introspectionSchema) {
    return introspectionSchema.types.find((item) => item.name === introspectionSchema.mutationType?.name);
}
function getSubscription(introspectionSchema) {
    return introspectionSchema.types.find((item) => item.name === introspectionSchema.subscriptionType?.name);
}
function getQueryByName(introspectionSchema, name) {
    const queryType = getQuery(introspectionSchema);
    return queryType.fields.find((item) => item.name === name);
}
function getMutationByName(introspectionSchema, name) {
    const mutationType = getMutation(introspectionSchema);
    return mutationType.fields.find((item) => item.name === name);
}
function getSubscriptionByName(introspectionSchema, name) {
    const subscriptionType = getSubscription(introspectionSchema);
    return subscriptionType.fields.find((item) => item.name === name);
}
