// @nestjs/graphql/dist/tests/tests/schema-builder/storages/type-metadata-storage.register-in.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const lib_1 = require("../../../lib");
const lazy_metadata_storage_1 = require("../../../lib/schema-builder/storages/lazy-metadata.storage");
// Mock modules for testing
class ModuleA {
}
class ModuleB {
}
class ModuleC {
}
// Test types for ObjectType
let ObjectTypeInModuleA = class ObjectTypeInModuleA {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], ObjectTypeInModuleA.prototype, "name", void 0);
ObjectTypeInModuleA = tslib_1.__decorate([
    (0, lib_1.ObjectType)({ registerIn: () => ModuleA })
], ObjectTypeInModuleA);
let ObjectTypeInModuleB = class ObjectTypeInModuleB {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], ObjectTypeInModuleB.prototype, "name", void 0);
ObjectTypeInModuleB = tslib_1.__decorate([
    (0, lib_1.ObjectType)({ registerIn: () => ModuleB })
], ObjectTypeInModuleB);
let ObjectTypeInModuleADirect = class ObjectTypeInModuleADirect {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], ObjectTypeInModuleADirect.prototype, "name", void 0);
ObjectTypeInModuleADirect = tslib_1.__decorate([
    (0, lib_1.ObjectType)({ registerIn: ModuleA }) // Direct reference (not arrow function)
], ObjectTypeInModuleADirect);
let ObjectTypeGlobal = class ObjectTypeGlobal {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], ObjectTypeGlobal.prototype, "name", void 0);
ObjectTypeGlobal = tslib_1.__decorate([
    (0, lib_1.ObjectType)() // No registerIn - should be included everywhere
], ObjectTypeGlobal);
// Test types for InputType
let InputTypeInModuleA = class InputTypeInModuleA {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], InputTypeInModuleA.prototype, "name", void 0);
InputTypeInModuleA = tslib_1.__decorate([
    (0, lib_1.InputType)({ registerIn: () => ModuleA })
], InputTypeInModuleA);
let InputTypeInModuleB = class InputTypeInModuleB {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], InputTypeInModuleB.prototype, "name", void 0);
InputTypeInModuleB = tslib_1.__decorate([
    (0, lib_1.InputType)({ registerIn: () => ModuleB })
], InputTypeInModuleB);
let InputTypeGlobal = class InputTypeGlobal {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], InputTypeGlobal.prototype, "name", void 0);
InputTypeGlobal = tslib_1.__decorate([
    (0, lib_1.InputType)() // No registerIn
], InputTypeGlobal);
// Test types for InterfaceType
let InterfaceTypeInModuleA = class InterfaceTypeInModuleA {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], InterfaceTypeInModuleA.prototype, "name", void 0);
InterfaceTypeInModuleA = tslib_1.__decorate([
    (0, lib_1.InterfaceType)({ registerIn: () => ModuleA })
], InterfaceTypeInModuleA);
let InterfaceTypeInModuleB = class InterfaceTypeInModuleB {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], InterfaceTypeInModuleB.prototype, "name", void 0);
InterfaceTypeInModuleB = tslib_1.__decorate([
    (0, lib_1.InterfaceType)({ registerIn: () => ModuleB })
], InterfaceTypeInModuleB);
let InterfaceTypeGlobal = class InterfaceTypeGlobal {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], InterfaceTypeGlobal.prototype, "name", void 0);
InterfaceTypeGlobal = tslib_1.__decorate([
    (0, lib_1.InterfaceType)() // No registerIn
], InterfaceTypeGlobal);
// Test types for ArgsType
let ArgsTypeInModuleA = class ArgsTypeInModuleA {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], ArgsTypeInModuleA.prototype, "name", void 0);
ArgsTypeInModuleA = tslib_1.__decorate([
    (0, lib_1.ArgsType)({ registerIn: () => ModuleA })
], ArgsTypeInModuleA);
let ArgsTypeInModuleB = class ArgsTypeInModuleB {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], ArgsTypeInModuleB.prototype, "name", void 0);
ArgsTypeInModuleB = tslib_1.__decorate([
    (0, lib_1.ArgsType)({ registerIn: () => ModuleB })
], ArgsTypeInModuleB);
let ArgsTypeGlobal = class ArgsTypeGlobal {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    tslib_1.__metadata("design:type", String)
], ArgsTypeGlobal.prototype, "name", void 0);
ArgsTypeGlobal = tslib_1.__decorate([
    (0, lib_1.ArgsType)() // No registerIn
], ArgsTypeGlobal);
// Test enums
var StatusEnumA;
(function (StatusEnumA) {
    StatusEnumA["ACTIVE"] = "ACTIVE";
    StatusEnumA["INACTIVE"] = "INACTIVE";
})(StatusEnumA || (StatusEnumA = {}));
var StatusEnumB;
(function (StatusEnumB) {
    StatusEnumB["PENDING"] = "PENDING";
    StatusEnumB["COMPLETED"] = "COMPLETED";
})(StatusEnumB || (StatusEnumB = {}));
var StatusEnumGlobal;
(function (StatusEnumGlobal) {
    StatusEnumGlobal["YES"] = "YES";
    StatusEnumGlobal["NO"] = "NO";
})(StatusEnumGlobal || (StatusEnumGlobal = {}));
(0, lib_1.registerEnumType)(StatusEnumA, {
    name: 'StatusEnumA',
    registerIn: () => ModuleA,
});
(0, lib_1.registerEnumType)(StatusEnumB, {
    name: 'StatusEnumB',
    registerIn: () => ModuleB,
});
(0, lib_1.registerEnumType)(StatusEnumGlobal, {
    name: 'StatusEnumGlobal',
});
// Test unions - created after loading metadata to ensure ObjectTypes are registered
let UnionInModuleA;
let UnionInModuleB;
let UnionGlobal;
describe('TypeMetadataStorage registerIn filtering', () => {
    beforeAll(() => {
        // Load all decorated class metadata
        lazy_metadata_storage_1.LazyMetadataStorage.load([
            ObjectTypeInModuleA,
            ObjectTypeInModuleB,
            ObjectTypeInModuleADirect,
            ObjectTypeGlobal,
            InputTypeInModuleA,
            InputTypeInModuleB,
            InputTypeGlobal,
            InterfaceTypeInModuleA,
            InterfaceTypeInModuleB,
            InterfaceTypeGlobal,
            ArgsTypeInModuleA,
            ArgsTypeInModuleB,
            ArgsTypeGlobal,
        ]);
        // Create unions after ObjectTypes are registered
        UnionInModuleA = (0, lib_1.createUnionType)({
            name: 'UnionInModuleA',
            types: () => [ObjectTypeInModuleA, ObjectTypeGlobal],
            registerIn: () => ModuleA,
        });
        UnionInModuleB = (0, lib_1.createUnionType)({
            name: 'UnionInModuleB',
            types: () => [ObjectTypeInModuleB, ObjectTypeGlobal],
            registerIn: () => ModuleB,
        });
        UnionGlobal = (0, lib_1.createUnionType)({
            name: 'UnionGlobal',
            types: () => [ObjectTypeGlobal],
        });
        // Load enum and union metadata
        lazy_metadata_storage_1.LazyMetadataStorage.load([]);
        lib_1.TypeMetadataStorage.compile();
    });
    afterAll(() => {
        lib_1.TypeMetadataStorage.clear();
    });
    describe('getObjectTypesMetadataByModules', () => {
        it('should return types registered in specified module', () => {
            const result = lib_1.TypeMetadataStorage.getObjectTypesMetadataByModules([
                ModuleA,
            ]);
            const targets = result.map((m) => m.target);
            expect(targets).toContain(ObjectTypeInModuleA);
            expect(targets).toContain(ObjectTypeInModuleADirect);
            expect(targets).not.toContain(ObjectTypeInModuleB);
        });
        it('should include types with no registerIn (global types)', () => {
            const result = lib_1.TypeMetadataStorage.getObjectTypesMetadataByModules([
                ModuleA,
            ]);
            const targets = result.map((m) => m.target);
            expect(targets).toContain(ObjectTypeGlobal);
        });
        it('should handle multiple modules', () => {
            const result = lib_1.TypeMetadataStorage.getObjectTypesMetadataByModules([
                ModuleA,
                ModuleB,
            ]);
            const targets = result.map((m) => m.target);
            expect(targets).toContain(ObjectTypeInModuleA);
            expect(targets).toContain(ObjectTypeInModuleB);
            expect(targets).toContain(ObjectTypeGlobal);
        });
        it('should return only global types when module has no registered types', () => {
            const result = lib_1.TypeMetadataStorage.getObjectTypesMetadataByModules([
                ModuleC,
            ]);
            const targets = result.map((m) => m.target);
            expect(targets).toContain(ObjectTypeGlobal);
            expect(targets).not.toContain(ObjectTypeInModuleA);
            expect(targets).not.toContain(ObjectTypeInModuleB);
        });
        it('should handle direct module reference (not arrow function)', () => {
            const result = lib_1.TypeMetadataStorage.getObjectTypesMetadataByModules([
                ModuleA,
            ]);
            const targets = result.map((m) => m.target);
            expect(targets).toContain(ObjectTypeInModuleADirect);
        });
    });
    describe('getInputTypesMetadataByModules', () => {
        it('should return input types registered in specified module', () => {
            const result = lib_1.TypeMetadataStorage.getInputTypesMetadataByModules([
                ModuleA,
            ]);
            const targets = result.map((m) => m.target);
            expect(targets).toContain(InputTypeInModuleA);
            expect(targets).not.toContain(InputTypeInModuleB);
        });
        it('should include input types with no registerIn', () => {
            const result = lib_1.TypeMetadataStorage.getInputTypesMetadataByModules([
                ModuleB,
            ]);
            const targets = result.map((m) => m.target);
            expect(targets).toContain(InputTypeGlobal);
            expect(targets).toContain(InputTypeInModuleB);
        });
    });
    describe('getInterfacesMetadataByModules', () => {
        it('should return interface types registered in specified module', () => {
            const result = lib_1.TypeMetadataStorage.getInterfacesMetadataByModules([
                ModuleA,
            ]);
            const targets = result.map((m) => m.target);
            expect(targets).toContain(InterfaceTypeInModuleA);
            expect(targets).not.toContain(InterfaceTypeInModuleB);
        });
        it('should include interface types with no registerIn', () => {
            const result = lib_1.TypeMetadataStorage.getInterfacesMetadataByModules([
                ModuleA,
            ]);
            const targets = result.map((m) => m.target);
            expect(targets).toContain(InterfaceTypeGlobal);
        });
    });
    describe('getArgumentsMetadataByModules', () => {
        it('should return args types registered in specified module', () => {
            const result = lib_1.TypeMetadataStorage.getArgumentsMetadataByModules([
                ModuleA,
            ]);
            const targets = result.map((m) => m.target);
            expect(targets).toContain(ArgsTypeInModuleA);
            expect(targets).not.toContain(ArgsTypeInModuleB);
        });
        it('should include args types with no registerIn', () => {
            const result = lib_1.TypeMetadataStorage.getArgumentsMetadataByModules([
                ModuleB,
            ]);
            const targets = result.map((m) => m.target);
            expect(targets).toContain(ArgsTypeGlobal);
        });
    });
    describe('getEnumsMetadataByModules', () => {
        it('should return enums registered in specified module', () => {
            const result = lib_1.TypeMetadataStorage.getEnumsMetadataByModules([ModuleA]);
            const refs = result.map((m) => m.ref);
            expect(refs).toContain(StatusEnumA);
            expect(refs).not.toContain(StatusEnumB);
        });
        it('should include enums with no registerIn', () => {
            const result = lib_1.TypeMetadataStorage.getEnumsMetadataByModules([ModuleA]);
            const refs = result.map((m) => m.ref);
            expect(refs).toContain(StatusEnumGlobal);
        });
        it('should filter correctly for ModuleB', () => {
            const result = lib_1.TypeMetadataStorage.getEnumsMetadataByModules([ModuleB]);
            const refs = result.map((m) => m.ref);
            expect(refs).toContain(StatusEnumB);
            expect(refs).toContain(StatusEnumGlobal);
            expect(refs).not.toContain(StatusEnumA);
        });
    });
    describe('getUnionsMetadataByModules', () => {
        it('should return unions registered in specified module', () => {
            const result = lib_1.TypeMetadataStorage.getUnionsMetadataByModules([ModuleA]);
            const names = result.map((m) => m.name);
            expect(names).toContain('UnionInModuleA');
            expect(names).not.toContain('UnionInModuleB');
        });
        it('should include unions with no registerIn', () => {
            const result = lib_1.TypeMetadataStorage.getUnionsMetadataByModules([ModuleA]);
            const names = result.map((m) => m.name);
            expect(names).toContain('UnionGlobal');
        });
        it('should filter correctly for ModuleB', () => {
            const result = lib_1.TypeMetadataStorage.getUnionsMetadataByModules([ModuleB]);
            const names = result.map((m) => m.name);
            expect(names).toContain('UnionInModuleB');
            expect(names).toContain('UnionGlobal');
            expect(names).not.toContain('UnionInModuleA');
        });
    });
    describe('default behavior (all types included when no module filter)', () => {
        it('should return all types when using getObjectTypesMetadata (no module filter)', () => {
            const result = lib_1.TypeMetadataStorage.getObjectTypesMetadata();
            const targets = result.map((m) => m.target);
            expect(targets).toContain(ObjectTypeInModuleA);
            expect(targets).toContain(ObjectTypeInModuleB);
            expect(targets).toContain(ObjectTypeGlobal);
        });
        it('should return all input types when using getInputTypesMetadata (no module filter)', () => {
            const result = lib_1.TypeMetadataStorage.getInputTypesMetadata();
            const targets = result.map((m) => m.target);
            expect(targets).toContain(InputTypeInModuleA);
            expect(targets).toContain(InputTypeInModuleB);
            expect(targets).toContain(InputTypeGlobal);
        });
        it('should return all enums when using getEnumsMetadata (no module filter)', () => {
            const result = lib_1.TypeMetadataStorage.getEnumsMetadata();
            const refs = result.map((m) => m.ref);
            expect(refs).toContain(StatusEnumA);
            expect(refs).toContain(StatusEnumB);
            expect(refs).toContain(StatusEnumGlobal);
        });
    });
});
