// @nestjs/graphql/dist/tests/tests/plugin/model-class-visitor.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const fs = tslib_1.__importStar(require("fs"));
const ts = tslib_1.__importStar(require("typescript"));
const lib_1 = require("../../lib");
const compiler_plugin_1 = require("../../lib/plugin/compiler-plugin");
const create_cat_alt_dto_1 = require("./fixtures/create-cat-alt.dto");
const create_cat_dto_1 = require("./fixtures/create-cat.dto");
const deprecation_dto_1 = require("./fixtures/deprecation.dto");
const nullable_dto_1 = require("./fixtures/nullable.dto");
const defaultCompilerOptions = {
    module: ts.ModuleKind.ES2020,
    target: ts.ScriptTarget.ES2020,
    newLine: ts.NewLineKind.LineFeed,
    noEmitHelpers: true,
    experimentalDecorators: true,
    strict: true,
};
function transpile(source, pluginOptions, compilerOptions = defaultCompilerOptions) {
    const filename = 'create-cat.input.ts';
    const fakeProgram = ts.createProgram([filename], compilerOptions);
    const result = ts.transpileModule(source, {
        compilerOptions: compilerOptions,
        fileName: 'test.input.ts',
        transformers: {
            before: [(0, compiler_plugin_1.before)(pluginOptions, fakeProgram)],
        },
    });
    return result.outputText;
}
function compileFiles(rootDir, file, pluginOptions = {}, compilerOptions = defaultCompilerOptions) {
    const caseDir = __dirname + rootDir;
    const options = {
        ...compilerOptions,
        rootDir: caseDir,
        outDir: caseDir + '/actual',
    };
    const program = ts.createProgram([caseDir + '/' + file], options);
    program.emit(undefined, undefined, undefined, undefined, {
        before: [(0, compiler_plugin_1.before)(pluginOptions, program)],
    });
    const jsFile = file.replace(/\.ts$/, '.js');
    const actual = fs.readFileSync(caseDir + '/actual/' + jsFile).toString();
    const expected = fs.readFileSync(caseDir + '/expected/' + jsFile).toString();
    return { actual, expected };
}
describe('API model properties', () => {
    it('should add the metadata factory when no decorators exist', () => {
        const options = {
            module: ts.ModuleKind.ES2020,
            target: ts.ScriptTarget.ES2020,
            newLine: ts.NewLineKind.LineFeed,
            noEmitHelpers: true,
            strict: true,
            experimentalDecorators: true,
        };
        const filename = 'create-cat.input.ts';
        const fakeProgram = ts.createProgram([filename], options);
        const result = ts.transpileModule(create_cat_dto_1.createCatDtoText, {
            compilerOptions: options,
            fileName: filename,
            transformers: {
                before: [(0, compiler_plugin_1.before)({}, fakeProgram)],
            },
        });
        expect(result.outputText).toEqual(create_cat_dto_1.createCatDtoTextTranspiled);
    });
    it('should add partial metadata factory when some decorators exist', () => {
        const options = {
            module: ts.ModuleKind.ES2020,
            target: ts.ScriptTarget.ES2020,
            newLine: ts.NewLineKind.LineFeed,
            noEmitHelpers: true,
            strict: true,
            removeComments: true,
            experimentalDecorators: true,
        };
        const filename = 'create-cat.input.ts';
        const fakeProgram = ts.createProgram([filename], options);
        const result = ts.transpileModule(create_cat_alt_dto_1.createCatDtoAltText, {
            compilerOptions: options,
            fileName: filename,
            transformers: {
                before: [
                    (0, compiler_plugin_1.before)({
                        introspectComments: true,
                    }, fakeProgram),
                ],
            },
        });
        expect(result.outputText).toEqual(create_cat_alt_dto_1.createCatDtoTextAltTranspiled);
    });
    it('should manage imports statements when code "downleveled"', () => {
        const compilerOptions = {
            ...defaultCompilerOptions,
            module: ts.ModuleKind.CommonJS,
        };
        const { actual, expected } = compileFiles('/cases/es5-eager-imports', 'post.model.ts', {}, compilerOptions);
        expect(actual).toEqual(expected);
    });
    it('should support & understand nullable type unions', () => {
        const options = {
            module: ts.ModuleKind.ES2020,
            target: ts.ScriptTarget.ES2020,
            newLine: ts.NewLineKind.LineFeed,
            noEmitHelpers: true,
            strict: true,
            experimentalDecorators: true,
        };
        const filename = 'nullable.input.ts';
        const fakeProgram = ts.createProgram([filename], options);
        const result = ts.transpileModule(nullable_dto_1.nullableDtoText, {
            compilerOptions: options,
            fileName: filename,
            transformers: {
                before: [(0, compiler_plugin_1.before)({}, fakeProgram)],
            },
        });
        expect(result.outputText).toEqual(nullable_dto_1.nullableDtoTextTranspiled);
    });
    it('should respect @deprecation tag from JsDoc', () => {
        const options = {
            module: ts.ModuleKind.ES2020,
            target: ts.ScriptTarget.ES2020,
            newLine: ts.NewLineKind.LineFeed,
            noEmitHelpers: true,
            strict: true,
            experimentalDecorators: true,
        };
        const filename = 'deprecation.input.ts';
        const fakeProgram = ts.createProgram([filename], options);
        const result = ts.transpileModule(deprecation_dto_1.deprecationDtoText, {
            compilerOptions: options,
            fileName: filename,
            transformers: {
                before: [(0, compiler_plugin_1.before)({ introspectComments: true }, fakeProgram)],
            },
        });
        expect(result.outputText).toEqual(deprecation_dto_1.deprecationDtoTranspiled);
    });
    it('should process only classes decorated with one of supported decorators', () => {
        const source = `
@${lib_1.ObjectType.name}()
class ObjectTypeModel {
  prop: string;
}

@${lib_1.InputType.name}()
class InputTypeModel {
  prop: string;
}

@${lib_1.InterfaceType.name}()
class InterfaceTypeModel {
  prop: string;
}

class NotAModel {
  prop: string;
}
`;
        const actual = transpile(source, {});
        expect(actual).toMatchInlineSnapshot(`
      "\"use strict\";
      let ObjectTypeModel = class ObjectTypeModel {
          static _GRAPHQL_METADATA_FACTORY() {
              return { prop: { type: () => String } };
          }
      };
      ObjectTypeModel = __decorate([
          ObjectType()
      ], ObjectTypeModel);
      let InputTypeModel = class InputTypeModel {
          static _GRAPHQL_METADATA_FACTORY() {
              return { prop: { type: () => String } };
          }
      };
      InputTypeModel = __decorate([
          InputType()
      ], InputTypeModel);
      let InterfaceTypeModel = class InterfaceTypeModel {
          static _GRAPHQL_METADATA_FACTORY() {
              return { prop: { type: () => String } };
          }
      };
      InterfaceTypeModel = __decorate([
          InterfaceType()
      ], InterfaceTypeModel);
      class NotAModel {
      }
      "
    `);
    });
    it('should process fields declared as getters ', () => {
        const source = `
@ObjectType()
class ObjectTypeModel {
  get prop(): string {}
}
`;
        const actual = transpile(source, {});
        expect(actual).toMatchInlineSnapshot(`
      "\"use strict\";
      let ObjectTypeModel = class ObjectTypeModel {
          get prop() { }
          static _GRAPHQL_METADATA_FACTORY() {
              return { prop: { type: () => String } };
          }
      };
      ObjectTypeModel = __decorate([
          ObjectType()
      ], ObjectTypeModel);
      "
    `);
    });
    describe('should add description from JSDoc to class decorators argument', () => {
        it('when there are no arguments on decorator', () => {
            const source = `
/** Test Description */
@ObjectType()
class ObjectTypeModel {}
`;
            const actual = transpile(source, { introspectComments: true });
            expect(actual).toMatchInlineSnapshot(`
        "\"use strict\";
        /** Test Description */
        let ObjectTypeModel = class ObjectTypeModel {
            static _GRAPHQL_METADATA_FACTORY() {
                return {};
            }
        };
        ObjectTypeModel = __decorate([
            ObjectType({ description: \"Test Description\" })
        ], ObjectTypeModel);
        "
      `);
        });
        it('when there are arguments on decorator', () => {
            const source = `
/** Test1 Description */
@ObjectType({isAbstract: true})
class Test1Model {}

/** Test2 Description */
@ObjectType('name', {isAbstract: true})
class Test2Model {}
`;
            const actual = transpile(source, { introspectComments: true });
            expect(actual).toMatchInlineSnapshot(`
        "\"use strict\";
        /** Test1 Description */
        let Test1Model = class Test1Model {
            static _GRAPHQL_METADATA_FACTORY() {
                return {};
            }
        };
        Test1Model = __decorate([
            ObjectType({ description: \"Test1 Description\", isAbstract: true })
        ], Test1Model);
        /** Test2 Description */
        let Test2Model = class Test2Model {
            static _GRAPHQL_METADATA_FACTORY() {
                return {};
            }
        };
        Test2Model = __decorate([
            ObjectType('name', { description: \"Test2 Description\", isAbstract: true })
        ], Test2Model);
        "
      `);
        });
        it('should work if parameters passed as variable reference or function', () => {
            const source = `
/** test1 description */
@ObjectType('name', getOptions())
class Test1 {}

/** test2 description */
@ObjectType('name', options)
class Test2 {}
`;
            const actual = transpile(source, { introspectComments: true });
            expect(actual).toMatchInlineSnapshot(`
        "\"use strict\";
        /** test1 description */
        let Test1 = class Test1 {
            static _GRAPHQL_METADATA_FACTORY() {
                return {};
            }
        };
        Test1 = __decorate([
            ObjectType('name', { ...{ description: \"test1 description\" }, ...getOptions() })
        ], Test1);
        /** test2 description */
        let Test2 = class Test2 {
            static _GRAPHQL_METADATA_FACTORY() {
                return {};
            }
        };
        Test2 = __decorate([
            ObjectType('name', { ...{ description: \"test2 description\" }, ...options })
        ], Test2);
        "
      `);
        });
    });
    it('should amend Field decorator with introspection data', () => {
        const source = `
import { ID } from '@nestjs/graphql';

const field3Options = {nullable: true};

@ObjectType()
class Model {
   /**
   * Description
   */
   @Field(() => ID)
   field: string;

   /**
   * Description
   */
   @Field(() => ID, {nullable: true})
   field1: string;

   /**
   * Description
   */
   @Field({nullable: false})
   field2?: string;
   
   /**
   * Description
   */
   @Field(field3Options)
   field3: string;
   
   name: string;
}
`;
        const actual = transpile(source, { introspectComments: true });
        expect(actual).toMatchInlineSnapshot(`
      "import { ID } from '@nestjs/graphql';
      const field3Options = { nullable: true };
      let Model = class Model {
          static _GRAPHQL_METADATA_FACTORY() {
              return { name: { type: () => String } };
          }
      };
      __decorate([
          Field(() => ID, { description: \"Description\" })
      ], Model.prototype, \"field\", void 0);
      __decorate([
          Field(() => ID, { description: \"Description\", nullable: true })
      ], Model.prototype, \"field1\", void 0);
      __decorate([
          Field(() => String, { nullable: false, description: \"Description\" })
      ], Model.prototype, \"field2\", void 0);
      __decorate([
          Field(() => String, { ...{ description: \"Description\" }, ...field3Options })
      ], Model.prototype, \"field3\", void 0);
      Model = __decorate([
          ObjectType()
      ], Model);
      "
    `);
    });
    it('should treat a single-key string enum as an enum, not a string', () => {
        const source = `
enum MultipleEnum {
    One = 'One',
    Two = 'Two',
}

enum SingleEnum {
    Single = 'Single',
}

@ObjectType()
export class ClassA {
    multipleEnum: MultipleEnum;
    singleEnum: SingleEnum;
}
`;
        const actual = transpile(source, {});
        expect(actual).toContain('multipleEnum: { type: () => MultipleEnum }');
        expect(actual).toContain('singleEnum: { type: () => SingleEnum }');
    });
});
