// @nestjs/apollo/dist/tests/tests/graphql-federation/posts-service/federation-posts.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppModule = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const graphql_1 = require("@nestjs/graphql");
const disabled_1 = require("@apollo/server/plugin/disabled");
const path_1 = require("path");
const drivers_1 = require("../../../lib/drivers");
const posts_module_1 = require("./posts/posts.module");
const upper_directive_1 = require("./posts/upper.directive");
let AppModule = class AppModule {
};
exports.AppModule = AppModule;
exports.AppModule = AppModule = tslib_1.__decorate([
    (0, common_1.Module)({
        imports: [
            graphql_1.GraphQLModule.forRoot({
                driver: drivers_1.ApolloFederationDriver,
                typePaths: [(0, path_1.join)(__dirname, '**/*.graphql')],
                transformSchema: (schema) => (0, upper_directive_1.upperDirectiveTransformer)(schema, 'upper'),
                plugins: [(0, disabled_1.ApolloServerPluginInlineTraceDisabled)()],
            }),
            posts_module_1.PostsModule,
        ],
    })
], AppModule);
