// @nestjs/graphql/dist/tests/tests/schema-builder/factories/directive-inheritance.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const testing_1 = require("@nestjs/testing");
const lib_1 = require("../../../lib");
let AbstractBase = class AbstractBase {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => lib_1.ID),
    tslib_1.__metadata("design:type", Number)
], AbstractBase.prototype, "id", void 0);
AbstractBase = tslib_1.__decorate([
    (0, lib_1.ObjectType)({ isAbstract: true }),
    (0, lib_1.Directive)('@key(fields: "id")')
], AbstractBase);
let DerivedUser = class DerivedUser extends AbstractBase {
};
tslib_1.__decorate([
    (0, lib_1.Field)(),
    tslib_1.__metadata("design:type", String)
], DerivedUser.prototype, "name", void 0);
DerivedUser = tslib_1.__decorate([
    (0, lib_1.ObjectType)()
], DerivedUser);
let DerivedUserResolver = class DerivedUserResolver {
    user() {
        return null;
    }
};
tslib_1.__decorate([
    (0, lib_1.Query)(() => DerivedUser),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", DerivedUser)
], DerivedUserResolver.prototype, "user", null);
DerivedUserResolver = tslib_1.__decorate([
    (0, lib_1.Resolver)(() => DerivedUser)
], DerivedUserResolver);
describe('Directive inheritance from abstract @ObjectType', () => {
    let schemaFactory;
    beforeAll(async () => {
        const moduleRef = await testing_1.Test.createTestingModule({
            imports: [lib_1.GraphQLSchemaBuilderModule],
        }).compile();
        schemaFactory = moduleRef.get(lib_1.GraphQLSchemaFactory);
    });
    afterAll(() => {
        lib_1.TypeMetadataStorage.clear();
    });
    it('should propagate class-level directives from an abstract parent to the derived type', async () => {
        const schema = await schemaFactory.create([DerivedUserResolver], {
            skipCheck: true,
        });
        const derivedType = schema.getType('DerivedUser');
        expect(derivedType).toBeDefined();
        const directives = derivedType.astNode?.directives ?? [];
        const directiveNames = directives.map((d) => d.name.value);
        expect(directiveNames).toContain('key');
    });
});
