// @nestjs/graphql/dist/tests/tests/schema-builder/factories/enum-union-directives.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const testing_1 = require("@nestjs/testing");
const lib_1 = require("../../../lib");
var MovieGenre;
(function (MovieGenre) {
    MovieGenre["ACTION"] = "ACTION";
    MovieGenre["COMEDY"] = "COMEDY";
})(MovieGenre || (MovieGenre = {}));
(0, lib_1.registerEnumType)(MovieGenre, {
    name: 'MovieGenre',
    directives: ['@tag(name: "public")'],
    valuesMap: {
        ACTION: {
            directives: ['@tag(name: "featured")'],
        },
    },
});
let Book = class Book {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => lib_1.ID),
    tslib_1.__metadata("design:type", String)
], Book.prototype, "id", void 0);
Book = tslib_1.__decorate([
    (0, lib_1.ObjectType)()
], Book);
let Movie = class Movie {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => lib_1.ID),
    tslib_1.__metadata("design:type", String)
], Movie.prototype, "id", void 0);
Movie = tslib_1.__decorate([
    (0, lib_1.ObjectType)()
], Movie);
const MediaUnion = (0, lib_1.createUnionType)({
    name: 'Media',
    types: () => [Book, Movie],
    directives: ['@tag(name: "public")'],
});
let MediaResolver = class MediaResolver {
    genre() {
        return MovieGenre.ACTION;
    }
    media() {
        return null;
    }
};
tslib_1.__decorate([
    (0, lib_1.Query)(() => MovieGenre),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", String)
], MediaResolver.prototype, "genre", null);
tslib_1.__decorate([
    (0, lib_1.Query)(() => MediaUnion),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", Object)
], MediaResolver.prototype, "media", null);
MediaResolver = tslib_1.__decorate([
    (0, lib_1.Resolver)()
], MediaResolver);
describe('Directives on enums and unions (issue #2920)', () => {
    let schemaFactory;
    beforeAll(async () => {
        const moduleRef = await testing_1.Test.createTestingModule({
            imports: [lib_1.GraphQLSchemaBuilderModule],
        }).compile();
        schemaFactory = moduleRef.get(lib_1.GraphQLSchemaFactory);
    });
    afterAll(() => {
        lib_1.TypeMetadataStorage.clear();
    });
    it('should attach directives to the generated enum type', async () => {
        const schema = await schemaFactory.create([MediaResolver]);
        const enumType = schema.getType('MovieGenre');
        expect(enumType).toBeDefined();
        const directiveNames = enumType.astNode?.directives?.map((d) => d.name.value) ?? [];
        expect(directiveNames).toContain('tag');
    });
    it('should attach directives to individual enum values', async () => {
        const schema = await schemaFactory.create([MediaResolver]);
        const enumType = schema.getType('MovieGenre');
        const actionValue = enumType.getValue('ACTION');
        expect(actionValue).toBeDefined();
        const directiveNames = actionValue.astNode?.directives?.map((d) => d.name.value) ?? [];
        expect(directiveNames).toContain('tag');
        const comedyValue = enumType.getValue('COMEDY');
        expect(comedyValue.astNode).toBeUndefined();
    });
    it('should attach directives to the generated union type', async () => {
        const schema = await schemaFactory.create([MediaResolver]);
        const unionType = schema.getType('Media');
        expect(unionType).toBeDefined();
        const directiveNames = unionType.astNode?.directives?.map((d) => d.name.value) ?? [];
        expect(directiveNames).toContain('tag');
    });
});
