// @nestjs/graphql/dist/tests/tests/schema-builder/factories/enum-default-value.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const testing_1 = require("@nestjs/testing");
const graphql_1 = require("graphql");
const lib_1 = require("../../../lib");
var Category;
(function (Category) {
    Category["BOOKS"] = "books";
    Category["MOVIES"] = "movies";
})(Category || (Category = {}));
(0, lib_1.registerEnumType)(Category, { name: 'Category' });
let CategoryFilterInput = class CategoryFilterInput {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => Category, { defaultValue: 'BOOKS' }),
    tslib_1.__metadata("design:type", String)
], CategoryFilterInput.prototype, "category", void 0);
CategoryFilterInput = tslib_1.__decorate([
    (0, lib_1.InputType)()
], CategoryFilterInput);
let CategoryResolver = class CategoryResolver {
    getCategory(category) {
        return category;
    }
    getCategoryAlreadyValue(category) {
        return category;
    }
    filterCategories(filter) {
        return filter.category;
    }
};
tslib_1.__decorate([
    (0, lib_1.Query)(() => String),
    tslib_1.__param(0, (0, lib_1.Args)('category', { type: () => Category, defaultValue: 'BOOKS' })),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [String]),
    tslib_1.__metadata("design:returntype", String)
], CategoryResolver.prototype, "getCategory", null);
tslib_1.__decorate([
    (0, lib_1.Query)(() => String),
    tslib_1.__param(0, (0, lib_1.Args)('category', { type: () => Category, defaultValue: Category.MOVIES })),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [String]),
    tslib_1.__metadata("design:returntype", String)
], CategoryResolver.prototype, "getCategoryAlreadyValue", null);
tslib_1.__decorate([
    (0, lib_1.Query)(() => String),
    tslib_1.__param(0, (0, lib_1.Args)('filter', { type: () => CategoryFilterInput })),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [CategoryFilterInput]),
    tslib_1.__metadata("design:returntype", String)
], CategoryResolver.prototype, "filterCategories", null);
CategoryResolver = tslib_1.__decorate([
    (0, lib_1.Resolver)()
], CategoryResolver);
describe('Enum default value translation (issue #3618)', () => {
    let schema;
    beforeAll(async () => {
        const moduleRef = await testing_1.Test.createTestingModule({
            imports: [lib_1.GraphQLSchemaBuilderModule],
        }).compile();
        const schemaFactory = moduleRef.get(lib_1.GraphQLSchemaFactory);
        schema = await schemaFactory.create([CategoryResolver]);
    });
    afterAll(() => {
        lib_1.TypeMetadataStorage.clear();
    });
    it('should emit the enum key when defaultValue is passed as the enum key string', () => {
        const printed = (0, graphql_1.printSchema)(schema);
        expect(printed).toContain('getCategory(category: Category! = BOOKS): String!');
    });
    it('should emit the enum key when defaultValue is passed as the actual enum value', () => {
        const printed = (0, graphql_1.printSchema)(schema);
        expect(printed).toContain('getCategoryAlreadyValue(category: Category! = MOVIES): String!');
    });
    it('should translate enum key default values on input object fields', () => {
        const printed = (0, graphql_1.printSchema)(schema);
        expect(printed).toContain('category: Category! = BOOKS');
    });
    it('should store the resolved enum value on the generated GraphQLEnumType', () => {
        const queryType = schema.getQueryType();
        const fields = queryType.getFields();
        const categoryArg = fields.getCategory.args.find((a) => a.name === 'category');
        expect(categoryArg).toBeDefined();
        expect(categoryArg.defaultValue).toBe(Category.BOOKS);
        const innerType = 'ofType' in categoryArg.type
            ? categoryArg.type.ofType
            : categoryArg.type;
        expect(innerType.toString()).toBe('Category');
    });
});
