// @nestjs/graphql/dist/tests/tests/schema-builder/factories/extensions-on-input-type.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const testing_1 = require("@nestjs/testing");
const lib_1 = require("../../../lib");
let OType = class OType {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    (0, lib_1.Extensions)({ key: 'value' }),
    tslib_1.__metadata("design:type", String)
], OType.prototype, "OtypeMember", void 0);
OType = tslib_1.__decorate([
    (0, lib_1.ObjectType)()
], OType);
let IType = class IType {
};
tslib_1.__decorate([
    (0, lib_1.Field)(() => String),
    (0, lib_1.Extensions)({ key: 'value' }),
    tslib_1.__metadata("design:type", String)
], IType.prototype, "ItypeMember", void 0);
IType = tslib_1.__decorate([
    (0, lib_1.InputType)()
], IType);
let ExtensionsResolver = class ExtensionsResolver {
    getOType(input) {
        return null;
    }
};
tslib_1.__decorate([
    (0, lib_1.Query)(() => OType),
    tslib_1.__param(0, (0, lib_1.Args)('input', { type: () => IType })),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [IType]),
    tslib_1.__metadata("design:returntype", OType)
], ExtensionsResolver.prototype, "getOType", null);
ExtensionsResolver = tslib_1.__decorate([
    (0, lib_1.Resolver)()
], ExtensionsResolver);
describe('@Extensions on @Field in @InputType (issue #3328)', () => {
    let schemaFactory;
    beforeAll(async () => {
        const moduleRef = await testing_1.Test.createTestingModule({
            imports: [lib_1.GraphQLSchemaBuilderModule],
        }).compile();
        schemaFactory = moduleRef.get(lib_1.GraphQLSchemaFactory);
    });
    afterAll(() => {
        lib_1.TypeMetadataStorage.clear();
    });
    it('exposes @Extensions on @Field for both @ObjectType and @InputType (parity)', async () => {
        const schema = await schemaFactory.create([ExtensionsResolver], {
            skipCheck: true,
        });
        const objectType = schema.getType('OType');
        expect(objectType).toBeDefined();
        const objectFields = objectType.getFields();
        expect(objectFields.OtypeMember.extensions).toBeDefined();
        expect(objectFields.OtypeMember.extensions.key).toBe('value');
        const inputType = schema.getType('IType');
        expect(inputType).toBeDefined();
        const inputFields = inputType.getFields();
        // Regression: pre-fix this map was empty for input fields.
        expect(inputFields.ItypeMember.extensions).toBeDefined();
        expect(inputFields.ItypeMember.extensions.key).toBe('value');
    });
});
