// @nestjs/graphql/dist/tests/tests/schema-builder/factories/root-type.factory.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const testing_1 = require("@nestjs/testing");
const lib_1 = require("../../../lib");
const multiple_fields_with_same_name_error_1 = require("../../../lib/schema-builder/errors/multiple-fields-with-same-name.error");
const root_type_factory_1 = require("../../../lib/schema-builder/factories/root-type.factory");
const lazy_metadata_storage_1 = require("../../../lib/schema-builder/storages/lazy-metadata.storage");
function booleanResolverFactory(Decorator) {
    let BooleanResolver = class BooleanResolver {
        bool() {
            return true;
        }
    };
    tslib_1.__decorate([
        Decorator(() => Boolean),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], BooleanResolver.prototype, "bool", null);
    BooleanResolver = tslib_1.__decorate([
        (0, lib_1.Resolver)()
    ], BooleanResolver);
    return BooleanResolver;
}
let StringResolver = class StringResolver {
    str() {
        return '';
    }
};
tslib_1.__decorate([
    (0, lib_1.Query)(() => String),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", void 0)
], StringResolver.prototype, "str", null);
StringResolver = tslib_1.__decorate([
    (0, lib_1.Resolver)()
], StringResolver);
describe('RootTypeFactory', () => {
    let rootTypeFactory;
    beforeEach(async () => {
        const module = await testing_1.Test.createTestingModule({
            imports: [lib_1.GraphQLSchemaBuilderModule],
        }).compile();
        rootTypeFactory = module.get(root_type_factory_1.RootTypeFactory);
    });
    afterEach(() => {
        lib_1.TypeMetadataStorage.clear();
    });
    describe('generateFields', () => {
        describe.each([
            ['queries', lib_1.Query, 'Query', lib_1.TypeMetadataStorage.getQueriesMetadata],
            [
                'mutations',
                lib_1.Mutation,
                'Mutation',
                lib_1.TypeMetadataStorage.getMutationsMetadata,
            ],
        ])('when duplicate %s are defined', (_, Decorator, objectTypeName, metadataFn) => {
            let metadata;
            beforeEach(() => {
                lazy_metadata_storage_1.LazyMetadataStorage.load([
                    booleanResolverFactory(Decorator),
                    booleanResolverFactory(Decorator),
                ]);
                lib_1.TypeMetadataStorage.compile();
                metadata = metadataFn.apply(lib_1.TypeMetadataStorage);
            });
            it('should throw an error with noDuplicateFields: true', () => {
                expect(() => rootTypeFactory.generateFields(metadata, { noDuplicatedFields: true }, objectTypeName)).toThrow(multiple_fields_with_same_name_error_1.MultipleFieldsWithSameNameError);
            });
            it('should create GraphQL fields with noDuplicateFields: false', () => {
                const fields = rootTypeFactory.generateFields(metadata, { noDuplicatedFields: false }, objectTypeName);
                expect(fields).toHaveProperty('bool');
            });
            it('should create GraphQL fields with noDuplicateFields undefined', () => {
                const fields = rootTypeFactory.generateFields(metadata, {}, objectTypeName);
                expect(fields).toHaveProperty('bool');
            });
        });
        describe('when no duplicate queries are found', () => {
            beforeEach(() => {
                lazy_metadata_storage_1.LazyMetadataStorage.load([
                    booleanResolverFactory(lib_1.Query),
                    StringResolver,
                ]);
                lib_1.TypeMetadataStorage.compile();
            });
            it('should correctly create GraphQL fields', () => {
                const queriesMetadata = lib_1.TypeMetadataStorage.getQueriesMetadata();
                const fields = rootTypeFactory.generateFields(queriesMetadata, { noDuplicatedFields: true }, 'Query');
                expect(fields).toHaveProperty('bool');
                expect(fields).toHaveProperty('str');
            });
        });
    });
});
