// @nestjs/apollo/dist/tests/tests/subscriptions/federation-graphql-ws.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const testing_1 = require("@nestjs/testing");
const apollo_cache_inmemory_1 = require("apollo-cache-inmemory");
const apollo_client_1 = tslib_1.__importDefault(require("apollo-client"));
const graphql_tag_1 = require("graphql-tag");
const graphql_ws_1 = require("graphql-ws");
const supertest_1 = tslib_1.__importDefault(require("supertest"));
const ws_1 = tslib_1.__importDefault(require("ws"));
const federation_app_module_1 = require("./federation-app/federation-app.module");
const notification_resolver_1 = require("./federation-app/notification.resolver");
const graphql_ws_link_1 = require("./utils/graphql-ws.link");
const subscriptionQuery = (0, graphql_tag_1.gql) `
  subscription FederatedSubscription($id: String!, $recipient: String!) {
    newFederatedNotification(id: $id, recipient: $recipient) {
      id
      recipient
      message
    }
  }
`;
describe('GraphQL Federation - graphql-ws subscriptions', () => {
    let app;
    let wsClient;
    let port;
    beforeEach(async () => {
        const module = await testing_1.Test.createTestingModule({
            imports: [
                federation_app_module_1.FederationAppModule.forRoot({
                    subscriptions: { 'graphql-ws': {} },
                }),
            ],
        }).compile();
        app = module.createNestApplication();
        await app.init();
        await app.listen(0);
        port = app.getHttpServer().address().port;
    });
    afterEach(async () => {
        try {
            await wsClient?.dispose();
        }
        catch { }
        await app.close();
    });
    it('should connect and receive a filtered notification', async () => {
        wsClient = (0, graphql_ws_1.createClient)({
            url: `ws://localhost:${port}/graphql`,
            webSocketImpl: ws_1.default,
            retryAttempts: 0,
        });
        wsClient.on('connected', () => {
            setTimeout(() => {
                notification_resolver_1.pubSub.publish('newFederatedNotification', {
                    newFederatedNotification: {
                        id: '99',
                        recipient: 'alice',
                        message: 'wrong id',
                    },
                });
                notification_resolver_1.pubSub.publish('newFederatedNotification', {
                    newFederatedNotification: {
                        id: '1',
                        recipient: 'bob',
                        message: 'wrong recipient',
                    },
                });
                notification_resolver_1.pubSub.publish('newFederatedNotification', {
                    newFederatedNotification: {
                        id: '1',
                        recipient: 'alice',
                        message: 'Hello from federation',
                    },
                });
            }, 100);
        });
        const apolloClient = new apollo_client_1.default({
            link: new graphql_ws_link_1.GraphQLWsLink(wsClient),
            cache: new apollo_cache_inmemory_1.InMemoryCache(),
        });
        await new Promise((resolve, reject) => {
            apolloClient
                .subscribe({
                query: subscriptionQuery,
                variables: { id: '1', recipient: 'alice' },
            })
                .subscribe({
                next(value) {
                    try {
                        expect(value.data.newFederatedNotification.id).toEqual('1');
                        expect(value.data.newFederatedNotification.recipient).toEqual('alice');
                        expect(value.data.newFederatedNotification.message).toEqual('Hello from federation');
                        resolve();
                    }
                    catch (e) {
                        reject(e);
                    }
                },
                complete() { },
                error: reject,
            });
        });
    });
    it('should expose the Subscription type in the federation SDL', async () => {
        const response = await (0, supertest_1.default)(app.getHttpServer())
            .post('/graphql')
            .send({ query: '{ _service { sdl } }' })
            .expect(200);
        const sdl = response.body.data._service.sdl;
        expect(sdl).toContain('type Subscription');
        expect(sdl).toContain('newFederatedNotification');
    });
});
