// @nestjs/apollo/dist/tests/tests/subscriptions/graphql-ws.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const testing_1 = require("@nestjs/testing");
const apollo_cache_inmemory_1 = require("apollo-cache-inmemory");
const apollo_client_1 = tslib_1.__importStar(require("apollo-client"));
const graphql_tag_1 = require("graphql-tag");
const graphql_ws_1 = require("graphql-ws");
const ws_1 = tslib_1.__importDefault(require("ws"));
const app_module_1 = require("./app/app.module");
const notification_resolver_1 = require("./app/notification.resolver");
const graphql_ws_link_1 = require("./utils/graphql-ws.link");
const malformed_token_exception_1 = require("./utils/malformed-token.exception");
const subscriptionQuery = (0, graphql_tag_1.gql) `
  subscription TestSubscription($id: String!) {
    newNotification(id: $id) {
      id
      message
    }
  }
`;
describe('graphql-ws protocol', () => {
    let app;
    let wsClient;
    let port;
    beforeEach(async () => {
        const module = await testing_1.Test.createTestingModule({
            imports: [
                app_module_1.AppModule.forRoot({
                    context: (context) => {
                        const { authorization } = context?.connectionParams ?? {};
                        if (authorization) {
                            return { user: authorization.split('Bearer ')[1] };
                        }
                        else {
                            return {};
                        }
                    },
                    subscriptions: {
                        'graphql-ws': {
                            onConnect: (context) => {
                                if (!context.connectionParams.authorization) {
                                    return context.extra.socket.close(4000, 'Missing authorization');
                                }
                                const authorization = context.connectionParams
                                    .authorization;
                                if (!authorization.startsWith('Bearer ')) {
                                    throw new malformed_token_exception_1.MalformedTokenException();
                                }
                                return true;
                            },
                        },
                    },
                }),
            ],
        }).compile();
        app = module.createNestApplication();
        await app.init();
        await app.listen(0);
        port = app.getHttpServer().address().port;
    });
    it('should receive an error if missing token', async () => {
        wsClient = (0, graphql_ws_1.createClient)({
            url: `ws://localhost:${port}/graphql`,
            webSocketImpl: ws_1.default,
            connectionParams: {},
            retryAttempts: 0,
        });
        const closedPromise = new Promise((resolve) => {
            wsClient.on('closed', (ev) => {
                expect(ev.code).toEqual(4000);
                expect(ev.reason).toEqual('Missing authorization');
                resolve();
            });
        });
        const apolloClient = new apollo_client_1.default({
            link: new graphql_ws_link_1.GraphQLWsLink(wsClient),
            cache: new apollo_cache_inmemory_1.InMemoryCache(),
        });
        apolloClient
            .subscribe({
            query: subscriptionQuery,
            variables: {
                id: '1',
            },
        })
            .subscribe({
            next() { },
            complete() { },
            error() { },
        });
        await closedPromise;
    });
    it('should receive an error if token is malformed', async () => {
        wsClient = (0, graphql_ws_1.createClient)({
            url: `ws://localhost:${port}/graphql`,
            webSocketImpl: ws_1.default,
            connectionParams: {
                authorization: 'wrong token',
            },
            retryAttempts: 0,
        });
        const closedPromise = new Promise((resolve) => {
            wsClient.on('closed', (ev) => {
                expect(ev.code).toEqual(4500);
                expect(ev.reason).toEqual('Malformed token');
                resolve();
            });
        });
        const apolloClient = new apollo_client_1.default({
            link: new graphql_ws_link_1.GraphQLWsLink(wsClient),
            cache: new apollo_cache_inmemory_1.InMemoryCache(),
        });
        apolloClient
            .subscribe({
            query: subscriptionQuery,
            variables: {
                id: '1',
            },
        })
            .subscribe({
            next() { },
            complete() { },
            error() { },
        });
        await closedPromise;
    });
    it('should receive error on subscription if guard fails', async () => {
        wsClient = (0, graphql_ws_1.createClient)({
            url: `ws://localhost:${port}/graphql`,
            webSocketImpl: ws_1.default,
            connectionParams: {
                authorization: 'Bearer notest',
            },
            retryAttempts: 0,
        });
        const apolloClient = new apollo_client_1.default({
            link: new graphql_ws_link_1.GraphQLWsLink(wsClient),
            cache: new apollo_cache_inmemory_1.InMemoryCache(),
        });
        await new Promise((resolve, reject) => {
            apolloClient
                .subscribe({
                query: subscriptionQuery,
                variables: {
                    id: '1',
                },
            })
                .subscribe({
                next() { },
                complete() { },
                error(error) {
                    try {
                        expect(error).toBeInstanceOf(apollo_client_1.ApolloError);
                        expect(error.graphQLErrors[0].message).toEqual('Forbidden resource');
                        expect(error.graphQLErrors[0].path[0]).toEqual('newNotification');
                        resolve();
                    }
                    catch (e) {
                        reject(e);
                    }
                },
            });
        });
    });
    it('should connect to subscriptions', async () => {
        wsClient = (0, graphql_ws_1.createClient)({
            url: `ws://localhost:${port}/graphql`,
            webSocketImpl: ws_1.default,
            connectionParams: {
                authorization: 'Bearer test',
            },
            retryAttempts: 0,
        });
        wsClient.on('connected', () => {
            // timeout needed to allow the subscription to be established
            setTimeout(() => {
                notification_resolver_1.pubSub.publish('newNotification', {
                    newNotification: {
                        id: '2',
                        recipient: 'test',
                        message: 'wrong message!',
                    },
                });
                notification_resolver_1.pubSub.publish('newNotification', {
                    newNotification: {
                        id: '1',
                        recipient: 'someone-else',
                        message: 'wrong message!',
                    },
                });
                notification_resolver_1.pubSub.publish('newNotification', {
                    newNotification: {
                        id: '1',
                        recipient: 'test',
                        message: 'Hello graphql-ws',
                    },
                });
            }, 100);
        });
        const apolloClient = new apollo_client_1.default({
            link: new graphql_ws_link_1.GraphQLWsLink(wsClient),
            cache: new apollo_cache_inmemory_1.InMemoryCache(),
        });
        await new Promise((resolve, reject) => {
            apolloClient
                .subscribe({
                query: subscriptionQuery,
                variables: {
                    id: '1',
                },
            })
                .subscribe({
                next(value) {
                    try {
                        expect(value.data.newNotification.id).toEqual('1');
                        expect(value.data.newNotification.message).toEqual('Hello graphql-ws');
                        resolve();
                    }
                    catch (e) {
                        reject(e);
                    }
                },
                complete() { },
                error(error) {
                    reject(error);
                },
            });
        });
    });
    afterEach(async () => {
        try {
            await wsClient.dispose();
        }
        catch { }
        await app.close();
    });
});
