// @nestjs/apollo/dist/tests/tests/code-first-federation/post/post.resolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostResolver = void 0;
const tslib_1 = require("tslib");
const graphql_1 = require("@nestjs/graphql");
const search_result_union_1 = require("../unions/search-result.union");
const user_entity_1 = require("../user/user.entity");
const post_entity_1 = require("./post.entity");
const post_service_1 = require("./post.service");
let PostResolver = class PostResolver {
    constructor(postService) {
        this.postService = postService;
    }
    findPost(id) {
        return this.postService.findOne(id);
    }
    getPosts() {
        return this.postService.all();
    }
    search() {
        return [
            new user_entity_1.User({ id: 1, posts: [] }),
            new post_entity_1.Post({ id: 2, title: 'lorem ipsum', authorId: 1 }),
        ];
    }
    resolveRef(reference) {
        return this.postService.findOne(reference.id);
    }
};
exports.PostResolver = PostResolver;
tslib_1.__decorate([
    (0, graphql_1.Query)((returns) => post_entity_1.Post),
    tslib_1.__param(0, (0, graphql_1.Args)('id')),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Number]),
    tslib_1.__metadata("design:returntype", void 0)
], PostResolver.prototype, "findPost", null);
tslib_1.__decorate([
    (0, graphql_1.Query)((returns) => [post_entity_1.Post]),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", void 0)
], PostResolver.prototype, "getPosts", null);
tslib_1.__decorate([
    (0, graphql_1.Query)((returns) => [search_result_union_1.FederationSearchResultUnion], {
        deprecationReason: 'test',
    }),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", Array)
], PostResolver.prototype, "search", null);
tslib_1.__decorate([
    (0, graphql_1.ResolveReference)(),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", void 0)
], PostResolver.prototype, "resolveRef", null);
exports.PostResolver = PostResolver = tslib_1.__decorate([
    (0, graphql_1.Resolver)((of) => post_entity_1.Post),
    tslib_1.__metadata("design:paramtypes", [post_service_1.PostService])
], PostResolver);
