// @nestjs/apollo/dist/tests/tests/code-first/cats/cats.module.js
"use strict";
var CatsModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CatsModule = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const cats_resolver_1 = require("./cats.resolver");
let CatsModule = CatsModule_1 = class CatsModule {
    static register(resolverRegistrationMethod) {
        switch (resolverRegistrationMethod) {
            case 'useClass':
                return {
                    module: CatsModule_1,
                    providers: [
                        {
                            provide: cats_resolver_1.CatsResolver,
                            useClass: cats_resolver_1.CatsResolver,
                        },
                    ],
                };
            case 'useValue':
                return {
                    module: CatsModule_1,
                    providers: [
                        {
                            provide: cats_resolver_1.CatsResolver,
                            useValue: new cats_resolver_1.CatsResolver(),
                        },
                    ],
                };
            case 'useFactory':
            default:
                return {
                    module: CatsModule_1,
                    providers: [
                        {
                            provide: cats_resolver_1.CatsResolver,
                            useFactory() {
                                return new cats_resolver_1.CatsResolver();
                            },
                        },
                    ],
                };
        }
    }
};
exports.CatsModule = CatsModule;
exports.CatsModule = CatsModule = CatsModule_1 = tslib_1.__decorate([
    (0, common_1.Module)({})
], CatsModule);
