// @nestjs/graphql/dist/tests/tests/type-helpers/apply-field-decorators.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const type_helpers_utils_1 = require("../../lib/type-helpers/type-helpers.utils");
const lib_1 = require("../../lib");
const lazy_metadata_storage_1 = require("../../lib/schema-builder/storages/lazy-metadata.storage");
class Holder {
}
describe('applyFieldDecorators', () => {
    afterEach(() => {
        lib_1.TypeMetadataStorage.clear();
    });
    it('should not register an Extensions metadata entry when item.extensions is empty', () => {
        (0, type_helpers_utils_1.applyFieldDecorators)(Holder, {
            name: 'field',
            schemaName: 'field',
            typeFn: () => String,
            target: Holder,
            options: {},
            extensions: {},
        });
        lazy_metadata_storage_1.LazyMetadataStorage.load([Holder]);
        const storage = lib_1.TypeMetadataStorage;
        const entries = storage.metadataByTargetCollection
            .get(Holder)
            .fieldExtensions.getByName('field');
        expect(entries).toEqual([]);
    });
    it('should register an Extensions metadata entry when item.extensions has keys', () => {
        (0, type_helpers_utils_1.applyFieldDecorators)(Holder, {
            name: 'field',
            schemaName: 'field',
            typeFn: () => String,
            target: Holder,
            options: {},
            extensions: { audit: true },
        });
        lazy_metadata_storage_1.LazyMetadataStorage.load([Holder]);
        const storage = lib_1.TypeMetadataStorage;
        const entries = storage.metadataByTargetCollection
            .get(Holder)
            .fieldExtensions.getByName('field');
        expect(entries).toHaveLength(1);
        expect(entries[0].value).toEqual({ audit: true });
    });
});
