// @nestjs/apollo/dist/tests/tests/graphql/cats/cats.resolvers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CatsResolvers = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const graphql_1 = require("@nestjs/graphql");
const graphql_subscriptions_1 = require("graphql-subscriptions");
const cats_guard_1 = require("./cats.guard");
const cats_service_1 = require("./cats.service");
const pubSub = new graphql_subscriptions_1.PubSub();
let CatsResolvers = class CatsResolvers {
    constructor(catsService) {
        this.catsService = catsService;
    }
    async getCats() {
        return await this.catsService.findAll();
    }
    getColor() {
        return 'black';
    }
    weight() {
        return 5;
    }
    async findOneById(id) {
        return await this.catsService.findOneById(id);
    }
    async create(args) {
        const createdCat = await this.catsService.create(args);
        pubSub.publish('catCreated', { catCreated: createdCat });
        return createdCat;
    }
    catCreated() {
        return pubSub.asyncIterableIterator('catCreated');
    }
};
exports.CatsResolvers = CatsResolvers;
tslib_1.__decorate([
    (0, graphql_1.Query)(),
    (0, common_1.UseGuards)(cats_guard_1.CatsGuard),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", Promise)
], CatsResolvers.prototype, "getCats", null);
tslib_1.__decorate([
    (0, graphql_1.ResolveField)('color'),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", void 0)
], CatsResolvers.prototype, "getColor", null);
tslib_1.__decorate([
    (0, graphql_1.ResolveField)(),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", void 0)
], CatsResolvers.prototype, "weight", null);
tslib_1.__decorate([
    (0, graphql_1.Query)('cat'),
    tslib_1.__param(0, (0, graphql_1.Args)('id', common_1.ParseIntPipe)),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Number]),
    tslib_1.__metadata("design:returntype", Promise)
], CatsResolvers.prototype, "findOneById", null);
tslib_1.__decorate([
    (0, graphql_1.Mutation)('createCat'),
    tslib_1.__param(0, (0, graphql_1.Args)()),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", Promise)
], CatsResolvers.prototype, "create", null);
tslib_1.__decorate([
    (0, graphql_1.Subscription)('catCreated'),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", void 0)
], CatsResolvers.prototype, "catCreated", null);
exports.CatsResolvers = CatsResolvers = tslib_1.__decorate([
    (0, graphql_1.Resolver)('Cat'),
    tslib_1.__metadata("design:paramtypes", [cats_service_1.CatsService])
], CatsResolvers);
