// @nestjs/apollo/dist/tests/tests/graphql/cats/cats.module.js
"use strict";
var CatsModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CatsModule = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const cats_request_scoped_service_1 = require("./cats-request-scoped.service");
const cats_resolvers_1 = require("./cats.resolvers");
const cats_service_1 = require("./cats.service");
let CatsModule = CatsModule_1 = class CatsModule {
    static enableRequestScope() {
        return {
            module: CatsModule_1,
            providers: [
                {
                    provide: cats_service_1.CatsService,
                    useClass: cats_request_scoped_service_1.CatsRequestScopedService,
                    scope: common_1.Scope.REQUEST,
                },
            ],
        };
    }
};
exports.CatsModule = CatsModule;
exports.CatsModule = CatsModule = CatsModule_1 = tslib_1.__decorate([
    (0, common_1.Module)({
        providers: [cats_service_1.CatsService, cats_resolvers_1.CatsResolvers],
    })
], CatsModule);
