// @nestjs/apollo/dist/tests/tests/graphql/hello/hello.module.js
"use strict";
var HelloModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelloModule = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const graphql_1 = require("@nestjs/graphql");
const path_1 = require("path");
const drivers_1 = require("../../../lib/drivers");
const hello_resolver_1 = require("./hello.resolver");
const hello_service_1 = require("./hello.service");
const users_service_1 = require("./users/users.service");
let HelloModule = HelloModule_1 = class HelloModule {
    constructor(meta) {
        this.meta = meta;
    }
    static forRoot(meta) {
        return {
            module: HelloModule_1,
            providers: [meta],
        };
    }
};
exports.HelloModule = HelloModule;
exports.HelloModule = HelloModule = HelloModule_1 = tslib_1.__decorate([
    (0, common_1.Module)({
        imports: [
            graphql_1.GraphQLModule.forRoot({
                driver: drivers_1.ApolloDriver,
                typePaths: [(0, path_1.join)(__dirname, '*.graphql')],
            }),
        ],
        providers: [
            hello_resolver_1.HelloResolver,
            hello_service_1.HelloService,
            users_service_1.UsersService,
            {
                provide: 'REQUEST_ID',
                useFactory: () => 1,
                scope: common_1.Scope.REQUEST,
            },
        ],
    }),
    tslib_1.__param(0, (0, common_1.Inject)('META')),
    tslib_1.__metadata("design:paramtypes", [Object])
], HelloModule);
