// @nestjs/apollo/dist/tests/tests/graphql-federation/posts-service/posts/posts.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostsService = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const post_type_enum_1 = require("./post-type.enum");
let PostsService = class PostsService {
    constructor() {
        this.posts = [
            {
                id: '1',
                title: 'HELLO WORLD',
                body: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
                userId: '5',
                publishDate: new Date(0),
                type: post_type_enum_1.PostType.TEXT,
            },
        ];
    }
    findAll() {
        return Promise.resolve(this.posts);
    }
    findById(id) {
        return Promise.resolve(this.posts.find((p) => p.id === id));
    }
    findByUserId(id) {
        return Promise.resolve(this.posts.filter((p) => p.userId === id));
    }
    findByType(type) {
        return Promise.resolve(this.posts.filter((p) => p.type === type));
    }
    async publish(id, publishDate) {
        const post = await this.findById(id);
        post.publishDate = publishDate;
        return post;
    }
};
exports.PostsService = PostsService;
exports.PostsService = PostsService = tslib_1.__decorate([
    (0, common_1.Injectable)()
], PostsService);
