// @nestjs/graphql/dist/tests/tests/plugin/type-helpers/intersection-type.helper.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const lib_1 = require("../../../lib");
const plugin_constants_1 = require("../../../lib/plugin/plugin-constants");
const get_fields_and_decorator_util_1 = require("../../../lib/schema-builder/utils/get-fields-and-decorator.util");
describe('IntersectionType', () => {
    let ClassA = class ClassA {
    };
    tslib_1.__decorate([
        (0, lib_1.Field)({ nullable: true }),
        (0, lib_1.Directive)('@upper'),
        (0, lib_1.Extensions)({ extension: true }),
        tslib_1.__metadata("design:type", String)
    ], ClassA.prototype, "login", void 0);
    tslib_1.__decorate([
        (0, lib_1.Field)(),
        tslib_1.__metadata("design:type", String)
    ], ClassA.prototype, "password", void 0);
    ClassA = tslib_1.__decorate([
        (0, lib_1.ObjectType)()
    ], ClassA);
    let ClassB = class ClassB {
        static [plugin_constants_1.METADATA_FACTORY_NAME]() {
            return {
                firstName: { nullable: true, type: () => String },
            };
        }
    };
    tslib_1.__decorate([
        (0, lib_1.Field)(),
        (0, lib_1.Directive)('@upper'),
        (0, lib_1.Extensions)({ extension: true }),
        tslib_1.__metadata("design:type", String)
    ], ClassB.prototype, "lastName", void 0);
    tslib_1.__decorate([
        (0, lib_1.Field)(() => [String]),
        tslib_1.__metadata("design:type", Array)
    ], ClassB.prototype, "hobbies", void 0);
    ClassB = tslib_1.__decorate([
        (0, lib_1.ObjectType)()
    ], ClassB);
    class UpdateUserDto extends (0, lib_1.IntersectionType)(ClassA, ClassB) {
    }
    it('should inherit all fields from two types', () => {
        const prototype = Object.getPrototypeOf(UpdateUserDto);
        const { fields } = (0, get_fields_and_decorator_util_1.getFieldsAndDecoratorForType)(prototype);
        const [loginField, passwordField, lastNameField, hobbiesField, firstNameField,] = fields;
        expect(fields.length).toEqual(5);
        expect(loginField.name).toEqual('login');
        expect(passwordField.name).toEqual('password');
        expect(lastNameField.name).toEqual('lastName');
        expect(hobbiesField.name).toEqual('hobbies');
        expect(hobbiesField.options).toEqual({ isArray: true, arrayDepth: 1 });
        expect(firstNameField.name).toEqual('firstName');
        expect(loginField.directives.length).toEqual(1);
        expect(loginField.directives).toContainEqual({
            fieldName: 'login',
            sdl: '@upper',
            target: prototype,
        });
        expect(loginField.extensions).toEqual({
            extension: true,
        });
        expect(lastNameField.directives.length).toEqual(1);
        expect(lastNameField.directives).toContainEqual({
            fieldName: 'lastName',
            sdl: '@upper',
            target: prototype,
        });
        expect(lastNameField.extensions).toEqual({
            extension: true,
        });
    });
});
