// @nestjs/graphql/dist/tests/tests/plugin/type-helpers/partial-type.helper.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const decorators_1 = require("../../../lib/decorators");
const metadata_loader_1 = require("../../../lib/plugin/metadata-loader");
const get_fields_and_decorator_util_1 = require("../../../lib/schema-builder/utils/get-fields-and-decorator.util");
const type_helpers_1 = require("../../../lib/type-helpers");
const base_type_fixture_1 = require("./fixtures/base-type.fixture");
const serialized_metadata_fixture_1 = require("./fixtures/serialized-metadata.fixture");
const type_helpers_test_utils_1 = require("./type-helpers.test-utils");
let CreateUserDto = class CreateUserDto extends base_type_fixture_1.BaseType {
};
tslib_1.__decorate([
    (0, class_validator_1.IsString)(),
    (0, decorators_1.Field)(),
    tslib_1.__metadata("design:type", String)
], CreateUserDto.prototype, "firstName", void 0);
tslib_1.__decorate([
    (0, decorators_1.Field)({ nullable: true }),
    tslib_1.__metadata("design:type", String)
], CreateUserDto.prototype, "login", void 0);
tslib_1.__decorate([
    (0, class_transformer_1.Expose)(),
    (0, class_transformer_1.Transform)((str) => str + '_transformed'),
    (0, class_validator_1.IsString)(),
    (0, decorators_1.Field)(),
    tslib_1.__metadata("design:type", String)
], CreateUserDto.prototype, "password", void 0);
CreateUserDto = tslib_1.__decorate([
    (0, decorators_1.ObjectType)()
], CreateUserDto);
class UpdateUserDto extends (0, type_helpers_1.PartialType)(CreateUserDto) {
}
describe('PartialType', () => {
    const metadataLoader = new metadata_loader_1.MetadataLoader();
    it('should inherit all fields and set "nullable" to true', async () => {
        await metadataLoader.load(serialized_metadata_fixture_1.SERIALIZED_METADATA);
        const prototype = Object.getPrototypeOf(UpdateUserDto);
        const { fields } = (0, get_fields_and_decorator_util_1.getFieldsAndDecoratorForType)(prototype);
        expect(fields.length).toEqual(7);
        expect(fields).toEqual(expect.arrayContaining([
            expect.objectContaining({
                name: 'id',
                options: { nullable: true },
            }),
            expect.objectContaining({
                name: 'createdAt',
                options: { nullable: true },
            }),
            expect.objectContaining({
                name: 'updatedAt',
                options: { nullable: true },
            }),
            expect.objectContaining({
                name: 'login',
                options: { nullable: true },
            }),
            expect.objectContaining({
                name: 'password',
                options: { nullable: true },
            }),
            expect.objectContaining({
                name: 'firstName',
                options: { nullable: true },
            }),
            expect.objectContaining({
                name: 'meta',
                options: { nullable: true },
            }),
        ]));
        expect(fields[0].directives.length).toEqual(1);
        expect(fields[0].directives).toContainEqual({
            fieldName: 'id',
            sdl: '@upper',
            target: prototype,
        });
        expect(fields[0].extensions).toEqual({
            extension: true,
        });
    });
    describe('Validation metadata', () => {
        it('should inherit metadata', () => {
            const validationKeys = new Set((0, type_helpers_test_utils_1.getValidationMetadataByTarget)(UpdateUserDto).map((item) => item.propertyName));
            expect(Array.from(validationKeys)).toEqual([
                'firstName',
                'password',
                'id',
                'createdAt',
                'updatedAt',
                'login',
                'meta',
            ]);
        });
        it('should apply @IsOptional to properties reflected by the plugin', async () => {
            const updateDto = new UpdateUserDto();
            updateDto.firstName = null;
            const validationErrors = await (0, class_validator_1.validate)(updateDto);
            expect(validationErrors).toHaveLength(0);
        });
        describe('when object does not fulfil validation rules', () => {
            it('"validate" should return validation errors', async () => {
                const updateDto = new UpdateUserDto();
                updateDto.password = 1234567;
                const validationErrors = await (0, class_validator_1.validate)(updateDto);
                expect(validationErrors.length).toEqual(1);
                expect(validationErrors[0].constraints).toEqual({
                    isString: 'password must be a string',
                });
            });
        });
    });
    describe('omitDefaultValues', () => {
        let CreateFooDto = class CreateFooDto {
        };
        tslib_1.__decorate([
            (0, decorators_1.Field)({ defaultValue: 'foo' }),
            tslib_1.__metadata("design:type", String)
        ], CreateFooDto.prototype, "name", void 0);
        CreateFooDto = tslib_1.__decorate([
            (0, decorators_1.ObjectType)()
        ], CreateFooDto);
        class UpdateFooDto extends (0, type_helpers_1.PartialType)(CreateFooDto, {
            omitDefaultValues: true,
        }) {
        }
        it('should skip default values of inherited class', async () => {
            await metadataLoader.load(serialized_metadata_fixture_1.SERIALIZED_METADATA);
            const prototype = Object.getPrototypeOf(UpdateFooDto);
            const { fields } = (0, get_fields_and_decorator_util_1.getFieldsAndDecoratorForType)(prototype);
            expect(fields.length).toEqual(1);
            expect(fields[0].name).toEqual('name');
            expect(fields[0].options.nullable).toEqual(true);
            expect(fields[0].options.defaultValue).toEqual(undefined);
        });
    });
    describe('skipNullProperties', () => {
        it('should apply @IsOptional to properties reflected by the plugin if option `skipNullProperties` is true', async () => {
            class UpdateUserWithNullableDto extends (0, type_helpers_1.PartialType)(CreateUserDto, {
                skipNullProperties: true,
            }) {
            }
            const updateDto = new UpdateUserWithNullableDto();
            updateDto.firstName = null;
            const validationErrors = await (0, class_validator_1.validate)(updateDto);
            expect(validationErrors).toHaveLength(0);
        });
        it('should apply @IsOptional to properties reflected by the plugin if option `skipNullProperties` is undefined', async () => {
            class UpdateUserWithoutNullableDto extends (0, type_helpers_1.PartialType)(CreateUserDto, {
                skipNullProperties: undefined,
            }) {
            }
            const updateDto = new UpdateUserWithoutNullableDto();
            updateDto.firstName = null;
            const validationErrors = await (0, class_validator_1.validate)(updateDto);
            expect(validationErrors).toHaveLength(0);
        });
        it('should apply @ValidateIf to properties reflected by the plugin if option `skipNullProperties` is false', async () => {
            class UpdateUserWithoutNullableDto extends (0, type_helpers_1.PartialType)(CreateUserDto, {
                skipNullProperties: false,
            }) {
            }
            const updateDto = new UpdateUserWithoutNullableDto();
            updateDto.firstName = null;
            const validationErrors = await (0, class_validator_1.validate)(updateDto);
            expect(validationErrors).toHaveLength(1);
            expect(validationErrors[0].constraints).toEqual({
                isString: 'firstName must be a string',
            });
        });
    });
});
