// @nestjs/graphql/dist/tests/tests/plugin/type-helpers/pick-type.helper.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const class_validator_1 = require("class-validator");
const decorators_1 = require("../../../lib/decorators");
const metadata_loader_1 = require("../../../lib/plugin/metadata-loader");
const get_fields_and_decorator_util_1 = require("../../../lib/schema-builder/utils/get-fields-and-decorator.util");
const type_helpers_1 = require("../../../lib/type-helpers");
const create_user_dto_fixture_1 = require("./fixtures/create-user-dto.fixture");
const serialized_metadata_fixture_1 = require("./fixtures/serialized-metadata.fixture");
const type_helpers_test_utils_1 = require("./type-helpers.test-utils");
class UpdateUserDto extends (0, type_helpers_1.PickType)(create_user_dto_fixture_1.CreateUserDto, ['login']) {
}
class UpdateUserWithIdDto extends (0, type_helpers_1.PickType)(create_user_dto_fixture_1.CreateUserDto, ['_id']) {
}
class IsActiveUserDto extends (0, type_helpers_1.PickType)(create_user_dto_fixture_1.CreateUserDto, ['active']) {
}
describe('PickType', () => {
    const metadataLoader = new metadata_loader_1.MetadataLoader();
    it('should inherit "login" field', () => {
        const prototype = Object.getPrototypeOf(UpdateUserDto);
        const { fields } = (0, get_fields_and_decorator_util_1.getFieldsAndDecoratorForType)(prototype);
        expect(fields.length).toEqual(1);
        expect(fields[0].name).toEqual('login');
        expect(fields[0].directives.length).toEqual(1);
        expect(fields[0].directives).toContainEqual({
            fieldName: 'login',
            sdl: '@upper',
            target: prototype,
        });
    });
    it('should inherit "active" field', async () => {
        await metadataLoader.load(serialized_metadata_fixture_1.SERIALIZED_METADATA);
        const prototype = Object.getPrototypeOf(IsActiveUserDto);
        const { fields } = (0, get_fields_and_decorator_util_1.getFieldsAndDecoratorForType)(prototype);
        expect(fields.length).toEqual(1);
        expect(fields[0].name).toEqual('active');
    });
    it('should inherit renamed "_id" field', () => {
        const { fields } = (0, get_fields_and_decorator_util_1.getFieldsAndDecoratorForType)(Object.getPrototypeOf(UpdateUserWithIdDto));
        expect(fields.length).toEqual(1);
        expect(fields[0].name).toEqual('_id');
        expect(fields[0].extensions).toEqual({ extension: true });
    });
    describe('Validation metadata', () => {
        it('should inherit metadata with "password" property excluded', () => {
            const validationKeys = (0, type_helpers_test_utils_1.getValidationMetadataByTarget)(UpdateUserDto).map((item) => item.propertyName);
            expect(validationKeys).toEqual(['login']);
        });
        describe('when object does not fulfil validation rules', () => {
            it('"validate" should return validation errors', async () => {
                await metadataLoader.load(serialized_metadata_fixture_1.SERIALIZED_METADATA);
                const updateDto = new UpdateUserDto();
                updateDto.login = '1234567';
                let validationErrors = await (0, class_validator_1.validate)(updateDto);
                expect(validationErrors.length).toEqual(1);
                expect(validationErrors[0].constraints).toEqual({
                    minLength: 'login must be longer than or equal to 10 characters',
                });
                const isActiveDto = new IsActiveUserDto();
                isActiveDto.active = '1234567';
                validationErrors = await (0, class_validator_1.validate)(isActiveDto);
                expect(validationErrors.length).toEqual(1);
                expect(validationErrors[0].constraints).toEqual({
                    isBoolean: 'active must be a boolean value',
                });
            });
        });
        describe('otherwise', () => {
            it('"validate" should return an empty array', async () => {
                const updateDto = new UpdateUserDto();
                updateDto.login = '1234567891011';
                // @ts-expect-error
                updateDto.password;
                const validationErrors = await (0, class_validator_1.validate)(updateDto);
                expect(validationErrors.length).toEqual(0);
            });
        });
    });
    let ArgsType1 = class ArgsType1 {
    };
    tslib_1.__decorate([
        (0, decorators_1.Field)(),
        tslib_1.__metadata("design:type", String)
    ], ArgsType1.prototype, "a", void 0);
    tslib_1.__decorate([
        (0, decorators_1.Field)(),
        tslib_1.__metadata("design:type", String)
    ], ArgsType1.prototype, "b", void 0);
    tslib_1.__decorate([
        (0, decorators_1.Field)(),
        tslib_1.__metadata("design:type", String)
    ], ArgsType1.prototype, "c", void 0);
    tslib_1.__decorate([
        (0, decorators_1.Field)(),
        tslib_1.__metadata("design:type", String)
    ], ArgsType1.prototype, "d", void 0);
    tslib_1.__decorate([
        (0, decorators_1.Field)(),
        tslib_1.__metadata("design:type", String)
    ], ArgsType1.prototype, "e", void 0);
    tslib_1.__decorate([
        (0, decorators_1.Field)(),
        tslib_1.__metadata("design:type", String)
    ], ArgsType1.prototype, "f", void 0);
    tslib_1.__decorate([
        (0, decorators_1.Field)(),
        tslib_1.__metadata("design:type", String)
    ], ArgsType1.prototype, "g", void 0);
    tslib_1.__decorate([
        (0, decorators_1.Field)(),
        tslib_1.__metadata("design:type", String)
    ], ArgsType1.prototype, "h", void 0);
    ArgsType1 = tslib_1.__decorate([
        (0, decorators_1.ArgsType)()
    ], ArgsType1);
    let PickArgsTest1 = class PickArgsTest1 extends (0, type_helpers_1.PickType)(ArgsType1, ['a', 'c', 'f', 'h']) {
    };
    PickArgsTest1 = tslib_1.__decorate([
        (0, decorators_1.ObjectType)()
    ], PickArgsTest1);
    it('should inherit: a, c, f, h fields', () => {
        const { fields } = (0, get_fields_and_decorator_util_1.getFieldsAndDecoratorForType)(Object.getPrototypeOf(PickArgsTest1));
        expect(fields.length).toEqual(4);
        expect(fields[0].name).toEqual('a');
        expect(fields[1].name).toEqual('c');
        expect(fields[2].name).toEqual('f');
        expect(fields[3].name).toEqual('h');
    });
});
