// @nestjs/apollo/dist/tests/tests/code-first/app.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplicationModule = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const graphql_1 = require("@nestjs/graphql");
const drivers_1 = require("../../lib/drivers");
const directions_module_1 = require("./directions/directions.module");
const recipes_module_1 = require("./recipes/recipes.module");
let ApplicationModule = class ApplicationModule {
};
exports.ApplicationModule = ApplicationModule;
exports.ApplicationModule = ApplicationModule = tslib_1.__decorate([
    (0, common_1.Module)({
        imports: [
            recipes_module_1.RecipesModule,
            directions_module_1.DirectionsModule,
            graphql_1.GraphQLModule.forRoot({
                driver: drivers_1.ApolloDriver,
                includeStacktraceInErrorResponses: false,
                installSubscriptionHandlers: true,
                autoSchemaFile: true,
            }),
        ],
    })
], ApplicationModule);
