// @nestjs/apollo/dist/tests/tests/code-first-register-in/users/users.resolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UsersResolver = void 0;
const tslib_1 = require("tslib");
const graphql_1 = require("@nestjs/graphql");
const create_user_input_1 = require("./create-user.input");
const user_model_1 = require("./user.model");
let UsersResolver = class UsersResolver {
    constructor() {
        this._users = [
            { id: '1', name: 'Ada', info: { version: '1.0.0' } },
        ];
    }
    users() {
        return this._users;
    }
    createUser(input) {
        const user = {
            id: String(this._users.length + 1),
            name: input.name,
            info: { version: '1.0.0' },
        };
        this._users.push(user);
        return user;
    }
};
exports.UsersResolver = UsersResolver;
tslib_1.__decorate([
    (0, graphql_1.Query)((returns) => [user_model_1.User]),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", Array)
], UsersResolver.prototype, "users", null);
tslib_1.__decorate([
    (0, graphql_1.Mutation)((returns) => user_model_1.User),
    tslib_1.__param(0, (0, graphql_1.Args)('input')),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [create_user_input_1.CreateUserInput]),
    tslib_1.__metadata("design:returntype", user_model_1.User)
], UsersResolver.prototype, "createUser", null);
exports.UsersResolver = UsersResolver = tslib_1.__decorate([
    (0, graphql_1.Resolver)((of) => user_model_1.User)
], UsersResolver);
