// @nestjs/apollo/dist/tests/tests/subscriptions/federation-app/federation-app.module.js
"use strict";
var FederationAppModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FederationAppModule = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const graphql_1 = require("@nestjs/graphql");
const drivers_1 = require("../../../lib/drivers");
const notification_module_1 = require("./notification.module");
let FederationAppModule = FederationAppModule_1 = class FederationAppModule {
    static forRoot(options) {
        return {
            module: FederationAppModule_1,
            imports: [
                notification_module_1.NotificationModule,
                graphql_1.GraphQLModule.forRoot({
                    driver: drivers_1.ApolloFederationDriver,
                    autoSchemaFile: true,
                    includeStacktraceInErrorResponses: false,
                    subscriptions: options?.subscriptions,
                }),
            ],
        };
    }
};
exports.FederationAppModule = FederationAppModule;
exports.FederationAppModule = FederationAppModule = FederationAppModule_1 = tslib_1.__decorate([
    (0, common_1.Module)({})
], FederationAppModule);
