// @nestjs/graphql/dist/tests/tests/services/resolvers-explorer.service.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("reflect-metadata");
const graphql_constants_1 = require("../../lib/graphql.constants");
const resolvers_explorer_service_1 = require("../../lib/services/resolvers-explorer.service");
describe('ResolversExplorerService', () => {
    describe('canUseFastFieldResolver', () => {
        let service;
        let mockInstance;
        beforeEach(() => {
            // Create a minimal mock of the service to test private method
            service = Object.create(resolvers_explorer_service_1.ResolversExplorerService.prototype);
            service.gqlOptions = {};
            // Reset caches for each test
            service.hasGlobalFieldMiddleware = null;
            service.fieldResolverEnhancersLookup = null;
            mockInstance = {
                constructor: class TestResolver {
                },
                testMethod: vi.fn(),
            };
        });
        // Helper to access private method
        const canUseFastFieldResolver = (instance, methodKey, contextOptions) => {
            return service.canUseFastFieldResolver(instance, methodKey, contextOptions);
        };
        describe('when all conditions are met for fast-path', () => {
            it('should return true when no enhancers, no middleware, no param decorators', () => {
                const result = canUseFastFieldResolver(mockInstance, 'testMethod', {
                    guards: false,
                    filters: false,
                    interceptors: false,
                });
                expect(result).toBe(true);
            });
            it('should return true when contextOptions is undefined', () => {
                const result = canUseFastFieldResolver(mockInstance, 'testMethod', undefined);
                expect(result).toBe(true);
            });
        });
        describe('when guards are enabled', () => {
            it('should return false', () => {
                const result = canUseFastFieldResolver(mockInstance, 'testMethod', {
                    guards: true,
                    filters: false,
                    interceptors: false,
                });
                expect(result).toBe(false);
            });
        });
        describe('when filters are enabled', () => {
            it('should return false', () => {
                const result = canUseFastFieldResolver(mockInstance, 'testMethod', {
                    guards: false,
                    filters: true,
                    interceptors: false,
                });
                expect(result).toBe(false);
            });
        });
        describe('when interceptors are enabled', () => {
            it('should return false', () => {
                const result = canUseFastFieldResolver(mockInstance, 'testMethod', {
                    guards: false,
                    filters: false,
                    interceptors: true,
                });
                expect(result).toBe(false);
            });
        });
        describe('when method-level field middleware exists', () => {
            it('should return false', () => {
                Reflect.defineMetadata(graphql_constants_1.FIELD_RESOLVER_MIDDLEWARE_METADATA, [() => { }], mockInstance.testMethod);
                const result = canUseFastFieldResolver(mockInstance, 'testMethod', {
                    guards: false,
                    filters: false,
                    interceptors: false,
                });
                expect(result).toBe(false);
            });
            afterEach(() => {
                Reflect.deleteMetadata(graphql_constants_1.FIELD_RESOLVER_MIDDLEWARE_METADATA, mockInstance.testMethod);
            });
        });
        describe('when global field middleware exists', () => {
            it('should return false', () => {
                service.gqlOptions = {
                    buildSchemaOptions: {
                        fieldMiddleware: [() => { }],
                    },
                };
                const result = canUseFastFieldResolver(mockInstance, 'testMethod', {
                    guards: false,
                    filters: false,
                    interceptors: false,
                });
                expect(result).toBe(false);
            });
        });
        describe('when parameter decorators are used', () => {
            it('should return false when @Parent is used', () => {
                Reflect.defineMetadata(graphql_constants_1.PARAM_ARGS_METADATA, { '0:root': { index: 0, data: undefined } }, mockInstance.constructor, 'testMethod');
                const result = canUseFastFieldResolver(mockInstance, 'testMethod', {
                    guards: false,
                    filters: false,
                    interceptors: false,
                });
                expect(result).toBe(false);
            });
            it('should return false when @Args is used', () => {
                Reflect.defineMetadata(graphql_constants_1.PARAM_ARGS_METADATA, { '1:args': { index: 0, data: 'id' } }, mockInstance.constructor, 'testMethod');
                const result = canUseFastFieldResolver(mockInstance, 'testMethod', {
                    guards: false,
                    filters: false,
                    interceptors: false,
                });
                expect(result).toBe(false);
            });
            afterEach(() => {
                Reflect.deleteMetadata(graphql_constants_1.PARAM_ARGS_METADATA, mockInstance.constructor, 'testMethod');
            });
        });
        describe('when empty param metadata exists', () => {
            it('should return true', () => {
                Reflect.defineMetadata(graphql_constants_1.PARAM_ARGS_METADATA, {}, mockInstance.constructor, 'testMethod');
                const result = canUseFastFieldResolver(mockInstance, 'testMethod', {
                    guards: false,
                    filters: false,
                    interceptors: false,
                });
                expect(result).toBe(true);
            });
            afterEach(() => {
                Reflect.deleteMetadata(graphql_constants_1.PARAM_ARGS_METADATA, mockInstance.constructor, 'testMethod');
            });
        });
    });
    describe('fast-path field resolver behavior', () => {
        it('should correctly bind this context when using fast-path', () => {
            const mockService = { getData: vi.fn().mockReturnValue('test-data') };
            class TestResolver {
                constructor() {
                    this.service = mockService;
                }
                getField() {
                    return this.service.getData();
                }
            }
            const instance = new TestResolver();
            const resolverFn = instance.getField.bind(instance);
            // Simulate what graphql-js passes
            const result = resolverFn({ id: '1' }, {}, { user: {} }, { fieldName: 'getField' });
            expect(result).toBe('test-data');
            expect(mockService.getData).toHaveBeenCalled();
        });
        it('should correctly handle async resolvers in fast-path', async () => {
            class TestResolver {
                async getAsyncField() {
                    return Promise.resolve('async-data');
                }
            }
            const instance = new TestResolver();
            const resolverFn = instance.getAsyncField.bind(instance);
            const result = resolverFn({}, {}, {}, {});
            expect(result).toBeInstanceOf(Promise);
            expect(await result).toBe('async-data');
        });
        it('should propagate errors correctly in fast-path', () => {
            class TestResolver {
                getErrorField() {
                    throw new Error('Test error');
                }
            }
            const instance = new TestResolver();
            const resolverFn = instance.getErrorField.bind(instance);
            expect(() => resolverFn({}, {}, {}, {})).toThrow('Test error');
        });
        it('should receive all four graphql arguments', () => {
            const receivedArgs = [];
            class TestResolver {
                getField(...args) {
                    receivedArgs.push(...args);
                    return 'result';
                }
            }
            const instance = new TestResolver();
            const resolverFn = instance.getField.bind(instance);
            const parent = { id: '1' };
            const args = { limit: 10 };
            const context = { user: { id: 'user1' } };
            const info = { fieldName: 'getField' };
            resolverFn(parent, args, context, info);
            expect(receivedArgs).toEqual([parent, args, context, info]);
        });
    });
});
